# ClonalOmics

Python notebooks containing analysis of clonal gene expression, with a bit of proteomics and ATAC for good measure.

This repository was created and is maintained by [Marty Weissman](http://martyweissman.com), and he takes responsibility/blame for its contents.  The Python notebooks here analyze the data created by Jeff Mold, Michael Ratz, Carl-Johan Eriksson, and others from the [lab of Jonas Frisén](https://ki.se/en/cmb/jonas-frisens-group) at the Karolinska Institute.  A paper will be submitted soon, we hope.

# Paper abstract (draft)

Cells of the same type often differ in their expression of mRNAs and proteins. This has been attributed to the stochastic nature of transcription, resulting in fluctuations around the mean over short time scales (1-3). Such transcriptional differences have been implicated in the selection of cells during chemotherapy, where cells of a certain gene expression profile may escape therapy (4,5). Less is known about whether, beneath short-term stochastic fluctuations, there exist stable, heritable differences between cells of the same type. To study this, we took advantage of the intrinsic clonal barcode that T cell receptors constitute to track genome-wide transcriptome and chromatin accessibility profiles of single cells descended from individual naïve CD8+ T cells in response to an antigen in humans. Using this strategy, we found that clonal transcriptional profiles are stable over years in vivo in humans and can affect hundreds of expressed genes. Clone-specific gene expression profiles were retained during reactivation and differentiation of clonally related cells in separate environments. We provide evidence that clonal gene expression mirrors heritable chromatin accessibility, indicating that stable epigenetic states dictate clonal gene expression profiles. Clonal gene expression is not specific to T cells, or humans, as we also find evidence of this in genetically barcoded cells in the mouse brain. Long term clonal gene expression profiles add a layer of diversity within cell types, which may influence developmental cell selection, plasticity, regeneration and response to therapies.

1.	Elowitz, M. B., Levine, A. J., Siggia, E. D. & Swain, P. S. Stochastic gene expression in a single cell. Science 297, 1183-1186, doi:10.1126/science.1070919 (2002).
2.	Chang, H. H., Hemberg, M., Barahona, M., Ingber, D. E. & Huang, S. Transcriptome-wide noise controls lineage choice in mammalian progenitor cells. Nature 453, 544-547, doi:10.1038/nature06965 (2008).
3.	Raj, A. & van Oudenaarden, A. Nature, nurture, or chance: stochastic gene expression and its consequences. Cell 135, 216-226, doi:10.1016/j.cell.2008.09.050 (2008).
4.	Shaffer, S. M. et al. Rare cell variability and drug-induced reprogramming as a mode of cancer drug resistance. Nature 546, 431-435, doi:10.1038/nature22794 (2017).
5.	Emert, B. L. et al. Variability within rare cell states enables multiple paths toward drug resistance. Nat Biotechnol 39, 865-876, doi:10.1038/s41587-021-00837-3 (2021).

