{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 3:  RNA and ATAC, P9855\n",
    "\n",
    "Here we study the chromatin accessibility (ATAC peak heights), gene expression (mRNA counts), and a bit of protein abundance for experiment P9855."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.7.1 anndata==0.7.6 umap==0.5.1 numpy==1.20.1 scipy==1.6.1 pandas==1.2.3 scikit-learn==0.24.1 statsmodels==0.12.2 python-igraph==0.8.3 louvain==0.7.0 leidenalg==0.8.3\n"
     ]
    }
   ],
   "source": [
    "#Import packages.  Put plots \"inline\" in the notebook.  \n",
    "\n",
    "import numpy as np  # For numerical computations.\n",
    "import pandas as pd  # Pandas for data analysis.\n",
    "from IPython.display import display # Pretty display of data frames.\n",
    "\n",
    "# Plotting\n",
    "import matplotlib.pyplot as plt  # For basic plotting.\n",
    "import matplotlib.transforms as transforms\n",
    "import matplotlib.colors as pcolors\n",
    "from matplotlib.patches import Rectangle\n",
    "plt.rcParams['svg.fonttype'] = 'none' # Export fonts as text, not as curves.\n",
    "import seaborn as sns # For pretty visualization in Seaborn.  See https://seaborn.pydata.org/\n",
    "\n",
    "# Stats and scipy\n",
    "import statsmodels.api as sm\n",
    "import scipy.stats as ss\n",
    "import scipy.cluster.hierarchy as sch # For hierarchical clustering.\n",
    "\n",
    "\n",
    "from bisect import bisect # See bisect.bisect.\n",
    "from itertools import compress # Nice way to apply masks to lists, etc.\n",
    "\n",
    "# Import from scikit, for machine learning.\n",
    "\n",
    "from sklearn import base\n",
    "from sklearn.preprocessing import Binarizer, MinMaxScaler\n",
    "from sklearn.model_selection import ShuffleSplit, train_test_split, GridSearchCV\n",
    "from sklearn.feature_selection import SelectKBest, chi2, f_classif\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.naive_bayes import BernoulliNB, GaussianNB # Naive Bayes Classifiers.\n",
    "from sklearn.ensemble import VotingClassifier # Voting classifier.\n",
    "from sklearn.svm import LinearSVC, NuSVC, SVC # Support Vector Machine \n",
    "from sklearn.pipeline import make_pipeline, Pipeline \n",
    "from sklearn.cluster import MiniBatchKMeans, KMeans\n",
    "from sklearn.metrics.pairwise import pairwise_distances_argmin\n",
    "\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.decomposition import PCA\n",
    "# UMAPs are all the rage... no more tSNE. Sad :(\n",
    "import umap\n",
    "\n",
    "import scanpy as sc\n",
    "sc.settings.verbosity = 1             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')\n",
    "\n",
    "# Put plots inline rather than in a pop-up.\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hrule(repchar = '=', length=80):\n",
    "    '''\n",
    "    A quick function to print a horizontal line.\n",
    "    '''\n",
    "    if len(repchar) == 1:\n",
    "        print(repchar*length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def setfontsize():\n",
    "    SMALL_SIZE = 6\n",
    "    MEDIUM_SIZE = 6\n",
    "    BIGGER_SIZE = 6\n",
    "    \n",
    "    plt.rcParams['font.sans-serif'] = 'Helvetica'\n",
    "    font = findfont(FontProperties(family=['sans-serif']))\n",
    "\n",
    "    plt.rc('font', size=SMALL_SIZE)          # controls default text sizes\n",
    "    plt.rc('axes', titlesize=SMALL_SIZE)     # fontsize of the axes title\n",
    "    plt.rc('axes', labelsize=MEDIUM_SIZE)    # fontsize of the x and y labels\n",
    "    plt.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "    plt.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "    plt.rc('legend', fontsize=SMALL_SIZE)    # legend fontsize\n",
    "    plt.rc('figure', titlesize=BIGGER_SIZE)  # fontsize of the figure title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def desetfontsize():\n",
    "    SMALL_SIZE = 10\n",
    "    MEDIUM_SIZE = 12\n",
    "    BIGGER_SIZE = 14\n",
    "\n",
    "    plt.rc('font', size=SMALL_SIZE)          # controls default text sizes\n",
    "    plt.rc('axes', titlesize=SMALL_SIZE)     # fontsize of the axes title\n",
    "    plt.rc('axes', labelsize=MEDIUM_SIZE)    # fontsize of the x and y labels\n",
    "    plt.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "    plt.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "    plt.rc('legend', fontsize=SMALL_SIZE)    # legend fontsize\n",
    "    plt.rc('figure', titlesize=BIGGER_SIZE)  # fontsize of the figure title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.image import imread\n",
    "from tempfile import NamedTemporaryFile\n",
    "\n",
    "def get_size(fig, dpi=300):\n",
    "    with NamedTemporaryFile(suffix='.png') as f:\n",
    "        fig.savefig(f.name, bbox_inches='tight', dpi=dpi)\n",
    "        height, width, _channels = imread(f.name).shape\n",
    "        return width / dpi, height / dpi\n",
    "\n",
    "def set_size(fig, size, dpi=300, eps=1e-2, give_up=2, min_size_px=10):\n",
    "    target_width, target_height = size\n",
    "    set_width, set_height = target_width, target_height # reasonable starting point\n",
    "    deltas = [] # how far we have\n",
    "    while True:\n",
    "        fig.set_size_inches([set_width, set_height])\n",
    "        actual_width, actual_height = get_size(fig, dpi=dpi)\n",
    "        set_width *= target_width / actual_width\n",
    "        set_height *= target_height / actual_height\n",
    "        deltas.append(abs(actual_width - target_width) + abs(actual_height - target_height))\n",
    "        if deltas[-1] < eps:\n",
    "            return True\n",
    "        if len(deltas) > give_up and sorted(deltas[-give_up:]) == deltas[-give_up:]:\n",
    "            return False\n",
    "        if set_width * dpi < min_size_px or set_height * dpi < min_size_px:\n",
    "            return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.font_manager as fm\n",
    "from matplotlib.font_manager import findfont, FontProperties\n",
    "\n",
    "font = fm.FontProperties(\n",
    "        family = 'Helvetica', fname = '/System/Library/Fonts/Helvetica.ttc')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/System/Library/Fonts/Helvetica.ttc'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt.rcParams['font.sans-serif'] = 'Helvetica'\n",
    "\n",
    "font = findfont(FontProperties(family=['sans-serif']))\n",
    "font"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "matplotlib.use('cairo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-9-d576feec86de>:5: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "setfontsize()\n",
    "fig,ax = plt.subplots(1,1,figsize=(1,1), dpi=300)\n",
    "ax.text(0.1,0.5,'ABCDEFGHIJKLM')\n",
    "ax.text(0.1,0.2, 'abcdefghijklm')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.8/site-packages/anndata/_core/anndata.py:120: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "import anndata\n",
    "\n",
    "adata = anndata.read_loom('../Data/Processed/P9855.loom')\n",
    "adata.var_names = adata.var['gene_name']\n",
    "adata.obs_names = adata.obs['obs_names']\n",
    "\n",
    "prot_data = anndata.read_loom('../Data/Processed/P9855_protein.loom')\n",
    "prot_data.var_names = prot_data.var['var_names']\n",
    "prot_data.obs_names = prot_data.obs['obs_names']\n",
    "\n",
    "ATAC_data = anndata.read_loom('../Data/Processed/P9855_ATAC.loom')\n",
    "ATAC_data.var_names = ATAC_data.var['PeakID']\n",
    "ATAC_data.obs_names = ATAC_data.obs['Sample_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 70 × 11881\n",
       "    obs: 'Clone_ID', 'Good clone', 'Number_Of_Cells', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'log1p_total_counts_mt', 'n_genes_by_counts', 'obs_names', 'pct_counts_mt', 'total_counts', 'total_counts_mt'\n",
       "    var: 'ENSG', 'F_pval', 'GC_content', 'TSS', 'chr', 'dir', 'end', 'gene_name', 'highly_variable', 'highly_variable_rank', 'log1p_mean_counts', 'log1p_total_counts', 'mean_counts', 'means', 'mt', 'n_cells', 'n_cells_by_counts', 'n_counts', 'pct_dropout_by_counts', 'start', 'total_counts', 'transcript_length', 'variances', 'variances_norm'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 24213 × 8\n",
       "    obs: 'Clone_ID', 'Clone_Name_Well_ID', 'FSC-A', 'FSC-H', 'FSC-W', 'Numerical_ID', 'Plate', 'SSC-A', 'SSC-H', 'SSC-W', 'SampleID', 'obs_names'\n",
       "    var: 'gene', 'var_names'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 29 × 80599\n",
       "    obs: 'ATAC_clone', 'ATAC_parent', 'Batch', 'Cell Number', 'Clone ID', 'Clone_ID', 'Clone_ID_InVitro', 'Condition', 'D', 'D.1', 'Donor', 'HLA Type', 'In_Vivo_Clone_Number', 'Is_replicate', 'Ja', 'Ja2', 'Jb', 'Jb2', 'Nuc_Seq_TRA', 'Nuc_Seq_TRA_2', 'Nuc_Seq_TRB', 'Read_Count', 'Read_Count.1', 'Sample_name', 'TCR Alpha', 'TCR Alpha2', 'TCR Beta', 'TRB', 'Unnamed: 21', 'Va', 'Va2', 'Vb', 'Vb2', 'count', 'founder_pheno', 'notes', 'protein_pheno', 'sequence'\n",
       "    var: 'PeakID', 'chr', 'end', 'in_ensg', 'in_gene', 'itype', 'mid', 'start'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATAC_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 12 × 80599\n",
       "    obs: 'ATAC_clone', 'ATAC_parent', 'Batch', 'Cell Number', 'Clone ID', 'Clone_ID', 'Clone_ID_InVitro', 'Condition', 'D', 'D.1', 'Donor', 'HLA Type', 'In_Vivo_Clone_Number', 'Is_replicate', 'Ja', 'Ja2', 'Jb', 'Jb2', 'Nuc_Seq_TRA', 'Nuc_Seq_TRA_2', 'Nuc_Seq_TRB', 'Read_Count', 'Read_Count.1', 'Sample_name', 'TCR Alpha', 'TCR Alpha2', 'TCR Beta', 'TRB', 'Unnamed: 21', 'Va', 'Va2', 'Vb', 'Vb2', 'count', 'founder_pheno', 'notes', 'protein_pheno', 'sequence'\n",
       "    var: 'PeakID', 'chr', 'end', 'in_ensg', 'in_gene', 'itype', 'mid', 'start'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATACR_data = ATAC_data[ATAC_data.obs['Is_replicate'] == True, :]\n",
    "ATACR_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quick ATAC plotting tool"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We create a quick exploratory tool for plotting ATAC peaks, focusing on clones with replicates, and the sister clones 5a/5b.  We assign a modified clone label, for coloring purposes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "clones_focus = ['1', '15', '22', '23', '8', '9', '5a', '5b']\n",
    "\n",
    "def clone_rename(cl):\n",
    "    cl_strip = cl.split('_')[0]\n",
    "    if cl_strip in clones_focus:\n",
    "        return cl_strip #'Clone '+cl_strip[5:]\n",
    "    else:\n",
    "        return 'Other'\n",
    "\n",
    "ATAC_data.obs['colorclone'] = ATAC_data.obs['ATAC_clone'].map(lambda cl : clone_rename(cl))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ATAC_plot(peak):\n",
    "    fig,ax = plt.subplots(1,1,figsize=(12,3))\n",
    "    sns.set_style('whitegrid')\n",
    "    peak_heights = pd.DataFrame(ATAC_data.to_df()[peak])\n",
    "    peak_heights.columns = ['height']\n",
    "    peak_heights['parent'] = ATAC_data.obs.ATAC_clone.apply(lambda c : c.split('_')[0])\n",
    "    peak_heights['hue'] = ATAC_data.obs.colorclone\n",
    "    sns.scatterplot(x='parent', y='height', color='gray', data=peak_heights[peak_heights['hue'] == 'Other'], \n",
    "                    legend=False, ax=ax,s=100)\n",
    "    \n",
    "    sns.scatterplot(x='parent', y='height', hue='hue', edgecolor='k',\n",
    "                    data=peak_heights[peak_heights['hue'] != 'Other'], \n",
    "                    legend=False, ax=ax, s=100)\n",
    "    \n",
    "    ax.set_ylabel('Peak height')\n",
    "    ax.set_xlabel('Clone')\n",
    "    plt.suptitle('Heights (pseudolog-normalized) for {}'.format(peak))\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-16-afdeac0540c6>:18: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "ATAC_plot('Interval_1111')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quick figure, IL2RG\n",
    "\n",
    "By popular demand, here's a little figure related to IL2RG:  population and clonal expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "obs_names\n",
       "P9855_2001    7.654436\n",
       "P9855_2002    7.852678\n",
       "P9855_2003    8.214191\n",
       "P9855_2004    7.992360\n",
       "P9855_2005    8.090677\n",
       "                ...   \n",
       "P9855_2078    7.712408\n",
       "P9855_2079    7.539834\n",
       "P9855_2080    7.655730\n",
       "P9855_2081    7.998237\n",
       "P9855_2082    8.375481\n",
       "Name: IL2RG, Length: 70, dtype: float32"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IL2RG = adata.to_df()['IL2RG']\n",
    "IL2RG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-19-e81f9b3a308e>:31: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "fig,ax = plt.subplots(1,2, sharey = True, figsize=(10,5), \n",
    "                      gridspec_kw = {'width_ratios':[1,10]})\n",
    "\n",
    "allplot_df = pd.DataFrame(np.exp(IL2RG) - 1) # Remove log-normalization.\n",
    "#allplot_df = pd.DataFrame(IL2RG) # Use log-normalized data?\n",
    "allplot_df.columns=['IL2RG Expression']\n",
    "allplot_df['Clone'] = adata.obs['Clone_ID']\n",
    "allmean = allplot_df['IL2RG Expression'].median()\n",
    "\n",
    "randclones = allplot_df['Clone'].value_counts()[:8].index\n",
    "allplot_df = allplot_df[allplot_df['Clone'].isin(randclones)]\n",
    "\n",
    "sns.boxplot(y='IL2RG Expression', color=(0.9,0.9,0.9,1), data = allplot_df, ax=ax[0], orient='v', whis=np.inf)\n",
    "sns.stripplot(y='IL2RG Expression', color='black', data = allplot_df, \n",
    "              jitter=0.05, ax=ax[0], orient='v',s=7.0)\n",
    "\n",
    "sns.boxplot(y='IL2RG Expression', x='Clone', color=(0.9,0.9,0.9,1), data = allplot_df, \n",
    "            ax=ax[1], orient='v', whis=np.inf)\n",
    "\n",
    "sns.stripplot(y='IL2RG Expression', x='Clone', hue='Clone', data = allplot_df, \n",
    "              jitter=0.05, ax=ax[1], orient='v',s=7.0)\n",
    "\n",
    "ax[0].set_xlabel('All clones')\n",
    "ax[1].set_xticklabels(ax[1].get_xticklabels(), rotation = 45, ha=\"right\")\n",
    "ax[1].axhline(allmean, color=(1,0,0,0.4))\n",
    "ax[1].set_ylabel('')\n",
    "ax[1].legend_.remove()\n",
    "\n",
    "plt.suptitle('Expression of IL2RG, in {} bulks,'\n",
    "             '\\n overall (left) and separated into {} clones (right).'.format(len(allplot_df), len(allplot_df['Clone'].unique())))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Panel:  UMAP Clustering based on ATAC peak heights\n",
    "\n",
    "We cluster bulks by ATAC peaks using UMAP here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-20-f4a4864619ac>:46: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "n_components = 20\n",
    "n_neighbors = 10\n",
    "min_dist = 0.25\n",
    "\n",
    "mm = 1/25.4\n",
    "figsize = (60*mm, 40*mm)\n",
    "fig, ax = plt.subplots(1,1,figsize=figsize, dpi=300)\n",
    "setfontsize()\n",
    "\n",
    "X = ATAC_data.to_df().copy()\n",
    "y = ATAC_data.obs['colorclone']\n",
    "\n",
    "hue_order = ['1', '15', '22', '23', '8', '9', '5a', '5b'] # ['Clone '+a for a in clr_list]\n",
    "hue_order = ['  ' + hue for hue in hue_order]\n",
    "pca = PCA(n_components=n_components)\n",
    "X_pca = pca.fit_transform(X) # Apply PCA with 10 components.\n",
    "X_embedded = umap.UMAP(n_neighbors=n_neighbors, min_dist=min_dist, \n",
    "                           random_state = 42, metric='cosine').fit_transform(X_pca) # Reduce from 10 components to 2-dimensions.\n",
    "\n",
    "cl_df = pd.DataFrame(X_embedded, columns = ['UMAP x', 'UMAP y'], index=X.index) # Just for graphing.\n",
    "cl_df['Clone'] = '  ' + y\n",
    "#display(cl_df)\n",
    "\n",
    "sns.scatterplot(x='UMAP x', y='UMAP y', data = cl_df[cl_df['Clone'] == '  Other'], s=12, \n",
    "                color='lightgray', edgecolor='gray', linewidth=0.4, ax=ax, label='  Other')\n",
    "\n",
    "sns.scatterplot(x='UMAP x', y='UMAP y', hue='Clone', hue_order = hue_order, palette='Dark2',\n",
    "                data = cl_df[cl_df['Clone'] != '  Other'], s=20, edgecolor=None, ax=ax)\n",
    "\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "\n",
    "lgnd = plt.legend(loc='upper left', bbox_to_anchor=(1.01, 1.1), ncol=1, framealpha=0.0, title='Clone')\n",
    "\n",
    "#change the marker size manually for both lines\n",
    "lgnd.legendHandles[0]._sizes = [20]\n",
    "for j in range(1,9):\n",
    "    lgnd.legendHandles[j]._sizes = [25]\n",
    "\n",
    "plt.suptitle(\"UMAP clustering, {} peaks.\".format(len(X.T)), \n",
    "            x=0.6, y=1.0)\n",
    "\n",
    "set_size(fig, figsize)\n",
    "plt.savefig(\"Panels/UMAP.pdf\", bbox_inches='tight')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplement? -- varying the UMAP parameters\n",
    "\n",
    "Now we vary the UMAP parameters through 27 different choices, and look at the resulting distance (in the UMAP plane) between clones, replicates, and the sister clones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_components</th>\n",
       "      <th>n_neighbors</th>\n",
       "      <th>min_dist</th>\n",
       "      <th>rep_dist</th>\n",
       "      <th>sis_dist</th>\n",
       "      <th>nonrep_dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.049175</td>\n",
       "      <td>0.170591</td>\n",
       "      <td>6.543134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.037747</td>\n",
       "      <td>0.106842</td>\n",
       "      <td>11.729912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.057874</td>\n",
       "      <td>0.142394</td>\n",
       "      <td>11.142283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.113509</td>\n",
       "      <td>0.279723</td>\n",
       "      <td>3.339342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.206475</td>\n",
       "      <td>0.431932</td>\n",
       "      <td>2.356382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.242284</td>\n",
       "      <td>0.534982</td>\n",
       "      <td>2.776036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.253111</td>\n",
       "      <td>0.378336</td>\n",
       "      <td>1.713337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.262609</td>\n",
       "      <td>0.382962</td>\n",
       "      <td>1.890784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.329633</td>\n",
       "      <td>0.454379</td>\n",
       "      <td>2.32306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.074423</td>\n",
       "      <td>0.095475</td>\n",
       "      <td>4.397453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.105703</td>\n",
       "      <td>0.123668</td>\n",
       "      <td>4.490133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.186123</td>\n",
       "      <td>0.205156</td>\n",
       "      <td>3.801679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.123611</td>\n",
       "      <td>0.216881</td>\n",
       "      <td>2.313449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.196943</td>\n",
       "      <td>0.243575</td>\n",
       "      <td>2.439017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.252807</td>\n",
       "      <td>0.314566</td>\n",
       "      <td>2.491022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.245934</td>\n",
       "      <td>0.273598</td>\n",
       "      <td>1.697266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.259419</td>\n",
       "      <td>0.41518</td>\n",
       "      <td>1.936227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.340481</td>\n",
       "      <td>0.392483</td>\n",
       "      <td>2.039331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>20.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.048818</td>\n",
       "      <td>0.05499</td>\n",
       "      <td>7.011836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.059781</td>\n",
       "      <td>0.042863</td>\n",
       "      <td>8.467725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.066367</td>\n",
       "      <td>0.059939</td>\n",
       "      <td>10.935038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>20.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.129034</td>\n",
       "      <td>0.210826</td>\n",
       "      <td>2.066247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>20.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.151329</td>\n",
       "      <td>0.165361</td>\n",
       "      <td>2.49899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>20.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.232518</td>\n",
       "      <td>0.250814</td>\n",
       "      <td>2.399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>20.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.273134</td>\n",
       "      <td>0.291181</td>\n",
       "      <td>1.625788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>20.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.250463</td>\n",
       "      <td>0.378422</td>\n",
       "      <td>1.821301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>20.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.317335</td>\n",
       "      <td>0.372107</td>\n",
       "      <td>2.181691</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   n_components n_neighbors min_dist  rep_dist  sis_dist nonrep_dist\n",
       "0           5.0         3.0      0.1  0.049175  0.170591    6.543134\n",
       "1           5.0         3.0     0.25  0.037747  0.106842   11.729912\n",
       "2           5.0         3.0      0.5  0.057874  0.142394   11.142283\n",
       "3           5.0         5.0      0.1  0.113509  0.279723    3.339342\n",
       "4           5.0         5.0     0.25  0.206475  0.431932    2.356382\n",
       "5           5.0         5.0      0.5  0.242284  0.534982    2.776036\n",
       "6           5.0        10.0      0.1  0.253111  0.378336    1.713337\n",
       "7           5.0        10.0     0.25  0.262609  0.382962    1.890784\n",
       "8           5.0        10.0      0.5  0.329633  0.454379     2.32306\n",
       "9          10.0         3.0      0.1  0.074423  0.095475    4.397453\n",
       "10         10.0         3.0     0.25  0.105703  0.123668    4.490133\n",
       "11         10.0         3.0      0.5  0.186123  0.205156    3.801679\n",
       "12         10.0         5.0      0.1  0.123611  0.216881    2.313449\n",
       "13         10.0         5.0     0.25  0.196943  0.243575    2.439017\n",
       "14         10.0         5.0      0.5  0.252807  0.314566    2.491022\n",
       "15         10.0        10.0      0.1  0.245934  0.273598    1.697266\n",
       "16         10.0        10.0     0.25  0.259419   0.41518    1.936227\n",
       "17         10.0        10.0      0.5  0.340481  0.392483    2.039331\n",
       "18         20.0         3.0      0.1  0.048818   0.05499    7.011836\n",
       "19         20.0         3.0     0.25  0.059781  0.042863    8.467725\n",
       "20         20.0         3.0      0.5  0.066367  0.059939   10.935038\n",
       "21         20.0         5.0      0.1  0.129034  0.210826    2.066247\n",
       "22         20.0         5.0     0.25  0.151329  0.165361     2.49899\n",
       "23         20.0         5.0      0.5  0.232518  0.250814       2.399\n",
       "24         20.0        10.0      0.1  0.273134  0.291181    1.625788\n",
       "25         20.0        10.0     0.25  0.250463  0.378422    1.821301\n",
       "26         20.0        10.0      0.5  0.317335  0.372107    2.181691"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from scipy.spatial import distance_matrix\n",
    "from scipy.spatial.distance import squareform, pdist\n",
    "\n",
    "n_comp_array = [5,10,20] # PCA components.  Can't be more than the number of samples!\n",
    "n_neigh_array = [3,5,10]\n",
    "min_dist_array = [0.1, 0.25, 0.5]\n",
    "\n",
    "\n",
    "X = ATAC_data.to_df().copy()\n",
    "y = ATAC_data.obs['colorclone']\n",
    "\n",
    "clr_list = ['1', '15', '22', '23', '8', '9', '5a', '5b'] # ['Clone '+a for a in clr_list]\n",
    "\n",
    "dist_df = pd.DataFrame(index=range(len(n_comp_array)*len(n_neigh_array)*len(min_dist_array)), \n",
    "                       columns=['n_components','n_neighbors','min_dist','rep_dist','sis_dist','nonrep_dist'])\n",
    "\n",
    "j = 0\n",
    "dm = {}\n",
    "for n_components in n_comp_array:\n",
    "    pca = PCA(n_components=n_components)\n",
    "    X_pca = pca.fit_transform(X) # Apply PCA with 10 components.\n",
    "    for n_neighbors in n_neigh_array:\n",
    "        for min_dist in min_dist_array:\n",
    "            X_embedded = umap.UMAP(n_neighbors=n_neighbors, min_dist=min_dist, \n",
    "                                   random_state = 42, metric='cosine').fit_transform(X_pca) # Reduce from 10 components to 2-dimensions.\n",
    "\n",
    "            cl_df = pd.DataFrame(X_embedded, columns = ['UMAP x', 'UMAP y'], index=X.index) # Just for graphing.\n",
    "            cl_df.index = 'clone' + ATAC_data.obs['ATAC_clone']\n",
    "            #cl_df['Clone'] = y\n",
    "            dm[j] = pd.DataFrame(distance_matrix(cl_df.values, cl_df.values), index=cl_df.index, columns=cl_df.index)\n",
    "            \n",
    "            #dm[j].index = 'clone' + ATAC_data.obs['ATAC_clone']\n",
    "            #dm[j].columns = 'clone' + ATAC_data.obs['ATAC_clone']\n",
    "            #display(dm[j])\n",
    "            rep_dist = np.array([dm[j].loc['clone'+num+'_R1', 'clone'+num+'_R2'] for num in ['1','15','22','23','8','9']])\n",
    "            nonreps = [cl for cl in cl_df.index if cl.split('_')[0] not in \n",
    "                      ['clone'+num for num in clr_list]]\n",
    "            dm_nonrep = dm[j].loc[nonreps][nonreps]\n",
    "            nonrep_dist = dm_nonrep.values.flatten().mean()\n",
    "            sis_dist = dm[j].loc['clone5a_R1','clone5b_R1']\n",
    "            dist_df.loc[j] = {'n_components':n_components,\n",
    "                             'n_neighbors':n_neighbors,\n",
    "                             'min_dist':min_dist,\n",
    "                             'rep_dist':rep_dist.mean() / nonrep_dist.mean(),\n",
    "                             'sis_dist':sis_dist / nonrep_dist.mean(),\n",
    "                             'nonrep_dist':nonrep_dist}\n",
    "            j += 1\n",
    "display(dist_df)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The mean UMAP distance between replicate clones is 18.0% +/- 9.6% of the mean non-replicate distance.\n",
      "The mean UMAP distance between sister clones 5a/5b is 25.9% +/- 13.6% of the mean non-replicate distance.\n"
     ]
    }
   ],
   "source": [
    "print('The mean UMAP distance between replicate clones is '\n",
    "      '{:0.1%} +/- {:0.1%} of the mean non-replicate distance.'.format(dist_df.rep_dist.mean(), dist_df.rep_dist.std()))\n",
    "\n",
    "print('The mean UMAP distance between sister clones 5a/5b is '\n",
    "      '{:0.1%} +/- {:0.1%} of the mean non-replicate distance.'.format(dist_df.sis_dist.mean(), dist_df.sis_dist.std()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_df = ATAC_data.to_df()\n",
    "ATAC_df.index = 'clone'+ATAC_data.obs['ATAC_clone']\n",
    "ATAC_df = ATAC_df.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rep_dist         0.059615\n",
       "sis_dist         0.985023\n",
       "nonrep_dist    194.908316\n",
       "dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "X = ATAC_df.T\n",
    "pca = PCA(n_components=20)\n",
    "X_PCA = pca.fit_transform(X)\n",
    "clr_list = ['1', '15', '22', '23', '8', '9', '5a', '5b']\n",
    "def clone_rename(cl):\n",
    "    cl_strip = cl.split('_')[0]\n",
    "    if cl_strip[5:] in clr_list:\n",
    "        return 'Clone '+cl_strip[5:]\n",
    "    else:\n",
    "        return 'Other'\n",
    "    \n",
    "y = [clone_rename(cl) for cl in ATAC_df.columns]\n",
    "hue_order = ['Clone '+a for a in clr_list]\n",
    "\n",
    "dm_raw = pd.DataFrame(distance_matrix(X_PCA, X_PCA), index=X.index, columns=X.index)\n",
    "rep_dist = np.array([dm_raw.loc['clone'+num+'_R1', 'clone'+num+'_R2'] for num in ['1','15','22','23','8','9']])\n",
    "nonreps = [cl for cl in X.index if cl.split('_')[0] not in ['clone'+num for num in clr_list]]\n",
    "dm_nonrep = dm_raw.loc[nonreps][nonreps]\n",
    "nonrep_dist = dm_nonrep.values.flatten().mean()\n",
    "sis_dist = dm_raw.loc['clone5a_R1','clone5b_R1']\n",
    "dist_raw_df = pd.Series({'rep_dist':rep_dist.mean() / nonrep_dist.mean(),\n",
    "                  'sis_dist':sis_dist / nonrep_dist.mean(),\n",
    "                  'nonrep_dist':nonrep_dist})\n",
    "display(dist_raw_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rep_dist         0.620445\n",
       "sis_dist         0.942482\n",
       "nonrep_dist    205.066003\n",
       "dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#X = ATAC_df.T\n",
    "clr_list = ['1', '15', '22', '23', '8', '9', '5a', '5b']\n",
    "def clone_rename(cl):\n",
    "    cl_strip = cl.split('_')[0]\n",
    "    if cl_strip[5:] in clr_list:\n",
    "        return 'Clone '+cl_strip[5:]\n",
    "    else:\n",
    "        return 'Other'\n",
    "    \n",
    "y = [clone_rename(cl) for cl in ATAC_df.columns]\n",
    "hue_order = ['Clone '+a for a in clr_list]\n",
    "\n",
    "dm_raw = pd.DataFrame(distance_matrix(X.values, X.values), index=X.index, columns=X.index)\n",
    "rep_dist = np.array([dm_raw.loc['clone'+num+'_R1', 'clone'+num+'_R2'] for num in ['1','15','22','23','8','9']])\n",
    "nonreps = [cl for cl in X.index if cl.split('_')[0] not in ['clone'+num for num in clr_list]]\n",
    "dm_nonrep = dm_raw.loc[nonreps][nonreps]\n",
    "nonrep_dist = dm_nonrep.values.flatten().mean()\n",
    "sis_dist = dm_raw.loc['clone5a_R1','clone5b_R1']\n",
    "dist_raw_df = pd.Series({'rep_dist':rep_dist.mean() / nonrep_dist.mean(),\n",
    "                  'sis_dist':sis_dist / nonrep_dist.mean(),\n",
    "                  'nonrep_dist':nonrep_dist})\n",
    "display(dist_raw_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def upper_values(sq_df):\n",
    "    n = len(sq_df)\n",
    "    if sq_df.shape != (n,n):\n",
    "        print('Not square!')\n",
    "        return None\n",
    "    index_list = [[i,j] for i in range(n) for j in range(n) if j > i]\n",
    "    return np.array([sq_df.values[il[0],il[1]] for il in index_list])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-27-60c6bee640b3>:87: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "dist_df['Distance between replicates'] = dist_df['rep_dist'] / dist_df['nonrep_dist']\n",
    "dist_df['Distance between 5a/5b'] = dist_df['sis_dist'] / dist_df['nonrep_dist']\n",
    "\n",
    "fig,ax = plt.subplots(1,2,figsize=(4, 3.0), dpi=300, sharey=True,\n",
    "                     gridspec_kw = {'width_ratios':[3,1]})\n",
    "fig.subplots_adjust(hspace=-2)\n",
    "sns.set_style('white')\n",
    "setfontsize()\n",
    "\n",
    "# First, the non-UMAP distances.\n",
    "dd = dist_raw_df\n",
    "nonreps = [cl for cl in cl_df.index if cl.split('_')[0] not in \n",
    "           ['clone'+num for num in clr_list]]\n",
    "dm_nonrep = dm_raw.loc[nonreps][nonreps]\n",
    "nonrep_dist = upper_values(dm_nonrep) # All distances between distinct non-replicates\n",
    "nonrep_dist = nonrep_dist / nonrep_dist.mean()\n",
    "    \n",
    "ax[0].scatter(nonrep_dist,[28]*len(nonrep_dist), s=0.8, color='gray', alpha=0.5) #, \n",
    "              # label = 'Distances between \\n non-replicate clones')\n",
    "ax[0].scatter(1.0, 28, s=2, color='k', marker='d') #, label='Mean dist between \\n non-replicate clones')\n",
    "rep_dist = np.array([dm_raw.loc['clone'+num+'_R1', 'clone'+num+'_R2'] \n",
    "                     for num in ['1','15','22','23','8','9']])\n",
    "rep_dist = rep_dist / dd.nonrep_dist\n",
    "ax[0].scatter(rep_dist,[28]*len(rep_dist), s=1, color='steelblue') #, \n",
    "              #label = 'Distances between \\n replicate clones')\n",
    "ax[0].scatter(dd.sis_dist,28, s=1, color='red', marker='o') #, \n",
    "              #label = 'Distance between \\n 5a/5b sister clones')\n",
    "\n",
    "for j in dist_df.index:\n",
    "    dd = dist_df.loc[j]\n",
    "    \n",
    "    # First, the nonreplicates.\n",
    "    dm_nonrep = dm[j].loc[nonreps][nonreps]\n",
    "    nonrep_dist = upper_values(dm_nonrep) # All distances between distinct non-replicates\n",
    "    nonrep_dist = nonrep_dist / nonrep_dist.mean()\n",
    "    \n",
    "    ax[0].scatter(nonrep_dist,[j]*len(nonrep_dist), s=0.8, color='gray', alpha=0.5, \n",
    "               label = 'Distances between \\n non-replicate clones')\n",
    "    \n",
    "    ax[0].scatter(1.0, j, s=2, color='k', marker='d', label='Mean dist between \\n non-replicate clones')\n",
    "    \n",
    "    # Then, the replicates.\n",
    "    rep_dist = np.array([dm[j].loc['clone'+num+'_R1', 'clone'+num+'_R2'] \n",
    "                         for num in ['1','15','22','23','8','9']])\n",
    "    rep_dist = rep_dist / dd.nonrep_dist\n",
    "    ax[0].scatter(rep_dist,[j]*len(rep_dist), s=1, color='steelblue', \n",
    "              label = 'Distances between \\n replicate clones')\n",
    "    \n",
    "    # Finally, the sisters 5a/5b.\n",
    "    ax[0].scatter(dd.sis_dist,j, s=1, color='red', marker='o', \n",
    "              label = 'Distance between \\n 5a/5b sister clones')\n",
    "    \n",
    "    if j == 0:\n",
    "        ax[0].legend(loc = 'upper right')\n",
    "    \n",
    "    ax[1].text(1.0,j,int(dist_df.loc[j,'n_components']),ha='left',va='center')\n",
    "    ax[1].text(0.5,j,int(dist_df.loc[j,'n_neighbors']),ha='left',va='center')\n",
    "    ax[1].text(0,j,dist_df.loc[j,'min_dist'],ha='left',va='center')\n",
    "\n",
    "ax[0].axhline(8.5, linewidth=0.5, alpha=0.5)\n",
    "ax[1].axhline(8.5, linewidth=0.5, alpha=0.5)\n",
    "ax[0].axhline(17.5, linewidth=0.5, alpha=0.5)\n",
    "ax[1].axhline(17.5, linewidth=0.5, alpha=0.5)\n",
    "\n",
    "for yoff in [0, 9, 18]:\n",
    "    for yoff2 in [2.5, 5.5]:\n",
    "        for j in [0,1]:\n",
    "            ax[j].axhline(yoff+yoff2, linewidth=0.5, alpha=0.5, color='gray')\n",
    "        \n",
    "ax[1].axis('off')\n",
    "\n",
    "ax[0].text(1.3,28, 'Raw distances without UMAP', ha='left', va='center')\n",
    "ax[1].text(0,27,'Min dist.', ha='left')\n",
    "ax[1].text(0.5,27,'# Nghbors', ha='left')\n",
    "ax[1].text(1,27,'#PCs', ha='left')\n",
    "ax[1].set_title('UMAP parameters')\n",
    "ax[0].set_yticks([])\n",
    "ax[0].tick_params(axis='x', which='major', pad=0)\n",
    "\n",
    "plt.suptitle('Raw distance based on {} peaks, and \\n'\n",
    "             'Normalized UMAP distances from '\n",
    "            '27 parameter choices'.format(len(X.T)), y=1)\n",
    "\n",
    "set_size(fig, (4, 3.0))\n",
    "plt.savefig(\"Panels/UMAP_dist.pdf\", bbox_inches='tight')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-28-4620c799c2d4>:27: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "figsize=(60*mm,30*mm)\n",
    "fig, ax = plt.subplots(1,1,figsize=figsize, dpi=300)\n",
    "setfontsize()\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "plot_df = dist_df.copy()\n",
    "plot_df['Parameter'] = dist_df['n_components'].astype(int).astype(str)+', '\n",
    "plot_df['Parameter'] += dist_df['n_neighbors'].astype(int).astype(str)+', '\n",
    "plot_df['Parameter'] +=  dist_df['min_dist'].astype(str).apply(lambda s : s[1:])\n",
    "sns.scatterplot(y = 'Distance between replicates', x='Parameter', data=plot_df, \n",
    "                s=4, edgecolor='k', color='steelblue', ax=ax, label='Mean distance between replicates')\n",
    "sns.scatterplot(y = 'Distance between 5a/5b', x='Parameter', data=plot_df, \n",
    "                s=4, edgecolor='k', color='orange', ax=ax, label='Mean distance between 5a/5b')\n",
    "\n",
    "ax.axhline(1.0, label='Mean distance between non-replicates', color='k', linewidth=1)\n",
    "ax.legend(bbox_to_anchor=(0.95, 0.9))\n",
    "ax.set_ylabel('UMAP distance')\n",
    "ax.set_xticklabels([])\n",
    "ax.set_xlabel('27 UMAP parameters')\n",
    "ax.set_yticks([0,0.2,1])\n",
    "ax.tick_params(axis='both', pad=-1)\n",
    "ax.tick_params(axis='y', labelright=True, labelleft=False)\n",
    "\n",
    "set_size(fig, figsize)\n",
    "plt.savefig(\"Panels/UMAP_params.pdf\", bbox_inches='tight')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Clonal peak variability\n",
    "\n",
    "Here we look at clonal peak variability, among the replicate clones.  The replicate clones offer a method for simultaneous quality control and variability assessment.  We put the replicate peak heights into a dataframe first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_77062</th>\n",
       "      <th>Interval_18293</th>\n",
       "      <th>Interval_51938</th>\n",
       "      <th>Interval_33832</th>\n",
       "      <th>Interval_13433</th>\n",
       "      <th>Interval_48454</th>\n",
       "      <th>Interval_27763</th>\n",
       "      <th>Interval_53540</th>\n",
       "      <th>Interval_46169</th>\n",
       "      <th>Interval_46787</th>\n",
       "      <th>...</th>\n",
       "      <th>Interval_58433</th>\n",
       "      <th>Interval_62113</th>\n",
       "      <th>Interval_45676</th>\n",
       "      <th>Interval_43639</th>\n",
       "      <th>Interval_35777</th>\n",
       "      <th>Interval_8387</th>\n",
       "      <th>Interval_17712</th>\n",
       "      <th>Interval_31505</th>\n",
       "      <th>Interval_18056</th>\n",
       "      <th>Interval_10427</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATAC_clone</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_R1</th>\n",
       "      <td>2.220971</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.315422</td>\n",
       "      <td>2.116660</td>\n",
       "      <td>5.671244</td>\n",
       "      <td>2.401717</td>\n",
       "      <td>3.377605</td>\n",
       "      <td>3.313264</td>\n",
       "      <td>0.648631</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000187</td>\n",
       "      <td>2.116660</td>\n",
       "      <td>3.090913</td>\n",
       "      <td>0.648631</td>\n",
       "      <td>1.318754</td>\n",
       "      <td>4.115008</td>\n",
       "      <td>2.401717</td>\n",
       "      <td>1.038805</td>\n",
       "      <td>2.116660</td>\n",
       "      <td>0.648631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_R2</th>\n",
       "      <td>2.574305</td>\n",
       "      <td>0.793996</td>\n",
       "      <td>2.743844</td>\n",
       "      <td>2.183737</td>\n",
       "      <td>5.564601</td>\n",
       "      <td>1.533994</td>\n",
       "      <td>3.179393</td>\n",
       "      <td>3.153849</td>\n",
       "      <td>1.230938</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.370034</td>\n",
       "      <td>1.954600</td>\n",
       "      <td>3.073051</td>\n",
       "      <td>0.473815</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.387019</td>\n",
       "      <td>1.230938</td>\n",
       "      <td>1.393903</td>\n",
       "      <td>1.864853</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15_R1</th>\n",
       "      <td>2.335525</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.683983</td>\n",
       "      <td>2.465641</td>\n",
       "      <td>5.688357</td>\n",
       "      <td>1.796163</td>\n",
       "      <td>3.665038</td>\n",
       "      <td>0.890413</td>\n",
       "      <td>1.523954</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.335525</td>\n",
       "      <td>1.353842</td>\n",
       "      <td>3.014939</td>\n",
       "      <td>0.541200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.979083</td>\n",
       "      <td>1.669292</td>\n",
       "      <td>1.908734</td>\n",
       "      <td>1.908734</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15_R2</th>\n",
       "      <td>2.405284</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.251001</td>\n",
       "      <td>2.472925</td>\n",
       "      <td>5.707511</td>\n",
       "      <td>2.536280</td>\n",
       "      <td>3.984053</td>\n",
       "      <td>0.574085</td>\n",
       "      <td>1.411480</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.332733</td>\n",
       "      <td>1.584639</td>\n",
       "      <td>2.893762</td>\n",
       "      <td>0.574085</td>\n",
       "      <td>0.574085</td>\n",
       "      <td>3.416717</td>\n",
       "      <td>1.411480</td>\n",
       "      <td>0.936490</td>\n",
       "      <td>1.584639</td>\n",
       "      <td>0.574085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_R1</th>\n",
       "      <td>2.244406</td>\n",
       "      <td>1.991444</td>\n",
       "      <td>1.836056</td>\n",
       "      <td>1.836056</td>\n",
       "      <td>5.708737</td>\n",
       "      <td>2.244406</td>\n",
       "      <td>3.383287</td>\n",
       "      <td>1.836056</td>\n",
       "      <td>1.134205</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.883102</td>\n",
       "      <td>1.652000</td>\n",
       "      <td>1.134205</td>\n",
       "      <td>0.719960</td>\n",
       "      <td>1.836056</td>\n",
       "      <td>3.577769</td>\n",
       "      <td>1.426249</td>\n",
       "      <td>1.426249</td>\n",
       "      <td>1.134205</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_R2</th>\n",
       "      <td>3.193007</td>\n",
       "      <td>1.524770</td>\n",
       "      <td>2.102691</td>\n",
       "      <td>2.843361</td>\n",
       "      <td>5.701477</td>\n",
       "      <td>1.986456</td>\n",
       "      <td>3.479647</td>\n",
       "      <td>1.307133</td>\n",
       "      <td>1.524770</td>\n",
       "      <td>0.641075</td>\n",
       "      <td>...</td>\n",
       "      <td>2.387270</td>\n",
       "      <td>1.986456</td>\n",
       "      <td>2.301105</td>\n",
       "      <td>0.641075</td>\n",
       "      <td>1.854911</td>\n",
       "      <td>4.069119</td>\n",
       "      <td>2.206811</td>\n",
       "      <td>1.028563</td>\n",
       "      <td>2.102691</td>\n",
       "      <td>0.641075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23_R1</th>\n",
       "      <td>1.966361</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.506567</td>\n",
       "      <td>1.684364</td>\n",
       "      <td>5.644652</td>\n",
       "      <td>1.290173</td>\n",
       "      <td>3.610647</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.684364</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.445269</td>\n",
       "      <td>1.290173</td>\n",
       "      <td>2.186078</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.966331</td>\n",
       "      <td>1.835270</td>\n",
       "      <td>1.684364</td>\n",
       "      <td>1.966361</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23_R2</th>\n",
       "      <td>2.089187</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.495113</td>\n",
       "      <td>2.207124</td>\n",
       "      <td>5.422726</td>\n",
       "      <td>1.618351</td>\n",
       "      <td>3.594341</td>\n",
       "      <td>1.106046</td>\n",
       "      <td>1.106046</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.649396</td>\n",
       "      <td>1.394653</td>\n",
       "      <td>1.955457</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.106046</td>\n",
       "      <td>3.839530</td>\n",
       "      <td>2.089187</td>\n",
       "      <td>1.801043</td>\n",
       "      <td>1.106046</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8_R1</th>\n",
       "      <td>2.349192</td>\n",
       "      <td>0.465191</td>\n",
       "      <td>1.845438</td>\n",
       "      <td>2.404186</td>\n",
       "      <td>5.449306</td>\n",
       "      <td>1.515987</td>\n",
       "      <td>3.879200</td>\n",
       "      <td>3.382059</td>\n",
       "      <td>0.781450</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.641270</td>\n",
       "      <td>1.376647</td>\n",
       "      <td>1.934876</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.214698</td>\n",
       "      <td>3.983718</td>\n",
       "      <td>2.722325</td>\n",
       "      <td>1.214698</td>\n",
       "      <td>2.092830</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8_R2</th>\n",
       "      <td>2.764788</td>\n",
       "      <td>0.783926</td>\n",
       "      <td>1.938780</td>\n",
       "      <td>2.645507</td>\n",
       "      <td>5.587162</td>\n",
       "      <td>1.938780</td>\n",
       "      <td>3.833457</td>\n",
       "      <td>3.569858</td>\n",
       "      <td>1.849280</td>\n",
       "      <td>0.466891</td>\n",
       "      <td>...</td>\n",
       "      <td>1.938780</td>\n",
       "      <td>1.641941</td>\n",
       "      <td>1.641941</td>\n",
       "      <td>1.519548</td>\n",
       "      <td>1.519548</td>\n",
       "      <td>4.178423</td>\n",
       "      <td>1.938780</td>\n",
       "      <td>1.750975</td>\n",
       "      <td>1.849280</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_R1</th>\n",
       "      <td>2.610464</td>\n",
       "      <td>1.078167</td>\n",
       "      <td>2.833161</td>\n",
       "      <td>2.774429</td>\n",
       "      <td>5.439472</td>\n",
       "      <td>2.414239</td>\n",
       "      <td>3.765365</td>\n",
       "      <td>1.078167</td>\n",
       "      <td>2.052520</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.061637</td>\n",
       "      <td>2.052520</td>\n",
       "      <td>2.574179</td>\n",
       "      <td>0.395295</td>\n",
       "      <td>2.112916</td>\n",
       "      <td>4.469330</td>\n",
       "      <td>1.845780</td>\n",
       "      <td>1.679597</td>\n",
       "      <td>1.766137</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_R2</th>\n",
       "      <td>1.911073</td>\n",
       "      <td>1.456665</td>\n",
       "      <td>2.905863</td>\n",
       "      <td>2.621405</td>\n",
       "      <td>5.426267</td>\n",
       "      <td>2.423591</td>\n",
       "      <td>3.834274</td>\n",
       "      <td>2.078779</td>\n",
       "      <td>2.025941</td>\n",
       "      <td>0.344620</td>\n",
       "      <td>...</td>\n",
       "      <td>2.949894</td>\n",
       "      <td>1.632085</td>\n",
       "      <td>2.527382</td>\n",
       "      <td>0.344620</td>\n",
       "      <td>1.970156</td>\n",
       "      <td>4.475250</td>\n",
       "      <td>1.911073</td>\n",
       "      <td>1.709461</td>\n",
       "      <td>1.848279</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 80599 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID      Interval_77062  Interval_18293  Interval_51938  Interval_33832  \\\n",
       "ATAC_clone                                                                   \n",
       "1_R1              2.220971        0.000000        2.315422        2.116660   \n",
       "1_R2              2.574305        0.793996        2.743844        2.183737   \n",
       "15_R1             2.335525        0.000000        2.683983        2.465641   \n",
       "15_R2             2.405284        0.000000        3.251001        2.472925   \n",
       "22_R1             2.244406        1.991444        1.836056        1.836056   \n",
       "22_R2             3.193007        1.524770        2.102691        2.843361   \n",
       "23_R1             1.966361        0.000000        1.506567        1.684364   \n",
       "23_R2             2.089187        0.000000        2.495113        2.207124   \n",
       "8_R1              2.349192        0.465191        1.845438        2.404186   \n",
       "8_R2              2.764788        0.783926        1.938780        2.645507   \n",
       "9_R1              2.610464        1.078167        2.833161        2.774429   \n",
       "9_R2              1.911073        1.456665        2.905863        2.621405   \n",
       "\n",
       "PeakID      Interval_13433  Interval_48454  Interval_27763  Interval_53540  \\\n",
       "ATAC_clone                                                                   \n",
       "1_R1              5.671244        2.401717        3.377605        3.313264   \n",
       "1_R2              5.564601        1.533994        3.179393        3.153849   \n",
       "15_R1             5.688357        1.796163        3.665038        0.890413   \n",
       "15_R2             5.707511        2.536280        3.984053        0.574085   \n",
       "22_R1             5.708737        2.244406        3.383287        1.836056   \n",
       "22_R2             5.701477        1.986456        3.479647        1.307133   \n",
       "23_R1             5.644652        1.290173        3.610647        0.000000   \n",
       "23_R2             5.422726        1.618351        3.594341        1.106046   \n",
       "8_R1              5.449306        1.515987        3.879200        3.382059   \n",
       "8_R2              5.587162        1.938780        3.833457        3.569858   \n",
       "9_R1              5.439472        2.414239        3.765365        1.078167   \n",
       "9_R2              5.426267        2.423591        3.834274        2.078779   \n",
       "\n",
       "PeakID      Interval_46169  Interval_46787  ...  Interval_58433  \\\n",
       "ATAC_clone                                  ...                   \n",
       "1_R1              0.648631        0.000000  ...        2.000187   \n",
       "1_R2              1.230938        0.000000  ...        2.370034   \n",
       "15_R1             1.523954        0.000000  ...        2.335525   \n",
       "15_R2             1.411480        0.000000  ...        2.332733   \n",
       "22_R1             1.134205        0.000000  ...        2.883102   \n",
       "22_R2             1.524770        0.641075  ...        2.387270   \n",
       "23_R1             1.684364        0.000000  ...        2.445269   \n",
       "23_R2             1.106046        0.000000  ...        2.649396   \n",
       "8_R1              0.781450        0.000000  ...        2.641270   \n",
       "8_R2              1.849280        0.466891  ...        1.938780   \n",
       "9_R1              2.052520        0.000000  ...        3.061637   \n",
       "9_R2              2.025941        0.344620  ...        2.949894   \n",
       "\n",
       "PeakID      Interval_62113  Interval_45676  Interval_43639  Interval_35777  \\\n",
       "ATAC_clone                                                                   \n",
       "1_R1              2.116660        3.090913        0.648631        1.318754   \n",
       "1_R2              1.954600        3.073051        0.473815        0.000000   \n",
       "15_R1             1.353842        3.014939        0.541200        0.000000   \n",
       "15_R2             1.584639        2.893762        0.574085        0.574085   \n",
       "22_R1             1.652000        1.134205        0.719960        1.836056   \n",
       "22_R2             1.986456        2.301105        0.641075        1.854911   \n",
       "23_R1             1.290173        2.186078        0.000000        0.000000   \n",
       "23_R2             1.394653        1.955457        0.000000        1.106046   \n",
       "8_R1              1.376647        1.934876        0.000000        1.214698   \n",
       "8_R2              1.641941        1.641941        1.519548        1.519548   \n",
       "9_R1              2.052520        2.574179        0.395295        2.112916   \n",
       "9_R2              1.632085        2.527382        0.344620        1.970156   \n",
       "\n",
       "PeakID      Interval_8387  Interval_17712  Interval_31505  Interval_18056  \\\n",
       "ATAC_clone                                                                  \n",
       "1_R1             4.115008        2.401717        1.038805        2.116660   \n",
       "1_R2             4.387019        1.230938        1.393903        1.864853   \n",
       "15_R1            2.979083        1.669292        1.908734        1.908734   \n",
       "15_R2            3.416717        1.411480        0.936490        1.584639   \n",
       "22_R1            3.577769        1.426249        1.426249        1.134205   \n",
       "22_R2            4.069119        2.206811        1.028563        2.102691   \n",
       "23_R1            3.966331        1.835270        1.684364        1.966361   \n",
       "23_R2            3.839530        2.089187        1.801043        1.106046   \n",
       "8_R1             3.983718        2.722325        1.214698        2.092830   \n",
       "8_R2             4.178423        1.938780        1.750975        1.849280   \n",
       "9_R1             4.469330        1.845780        1.679597        1.766137   \n",
       "9_R2             4.475250        1.911073        1.709461        1.848279   \n",
       "\n",
       "PeakID      Interval_10427  \n",
       "ATAC_clone                  \n",
       "1_R1              0.648631  \n",
       "1_R2              0.000000  \n",
       "15_R1             0.000000  \n",
       "15_R2             0.574085  \n",
       "22_R1             0.000000  \n",
       "22_R2             0.641075  \n",
       "23_R1             0.000000  \n",
       "23_R2             0.000000  \n",
       "8_R1              0.000000  \n",
       "8_R2              0.000000  \n",
       "9_R1              0.000000  \n",
       "9_R2              0.000000  \n",
       "\n",
       "[12 rows x 80599 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AR_df = ATACR_data.to_df()\n",
    "AR_df.index = ATACR_data.obs.ATAC_clone\n",
    "AR_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This data contains 12 samples, with 2 bulks from each clone.  These 2 bulks are considered replicates (R1 and R2)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1', '15', '22', '23', '8', '9'], dtype=object)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parents = ATACR_data.obs.ATAC_parent.unique()\n",
    "parents"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We examine the general distribution of replicate pairs, by \"melting\" the dataframe above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>min_value</th>\n",
       "      <th>Diff</th>\n",
       "      <th>Cut</th>\n",
       "      <th>bigdiff</th>\n",
       "      <th>badpeak</th>\n",
       "      <th>PeakID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.220971</td>\n",
       "      <td>0.353334</td>\n",
       "      <td>[1.976722536187036, 2.6322967147827145)</td>\n",
       "      <td>0.854268</td>\n",
       "      <td>False</td>\n",
       "      <td>Interval_77062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.335525</td>\n",
       "      <td>0.069759</td>\n",
       "      <td>[1.976722536187036, 2.6322967147827145)</td>\n",
       "      <td>0.854268</td>\n",
       "      <td>False</td>\n",
       "      <td>Interval_77062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.244406</td>\n",
       "      <td>0.948601</td>\n",
       "      <td>[1.976722536187036, 2.6322967147827145)</td>\n",
       "      <td>0.854268</td>\n",
       "      <td>True</td>\n",
       "      <td>Interval_77062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.966361</td>\n",
       "      <td>0.122825</td>\n",
       "      <td>[1.3211483574913574, 1.9767225360870362)</td>\n",
       "      <td>1.071294</td>\n",
       "      <td>False</td>\n",
       "      <td>Interval_77062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.349192</td>\n",
       "      <td>0.415596</td>\n",
       "      <td>[1.976722536187036, 2.6322967147827145)</td>\n",
       "      <td>0.854268</td>\n",
       "      <td>False</td>\n",
       "      <td>Interval_77062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483589</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.574085</td>\n",
       "      <td>[0.0, 0.01)</td>\n",
       "      <td>1.703403</td>\n",
       "      <td>False</td>\n",
       "      <td>Interval_10427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483590</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.641075</td>\n",
       "      <td>[0.0, 0.01)</td>\n",
       "      <td>1.703403</td>\n",
       "      <td>False</td>\n",
       "      <td>Interval_10427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483591</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[0.0, 0.01)</td>\n",
       "      <td>1.703403</td>\n",
       "      <td>False</td>\n",
       "      <td>Interval_10427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483592</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[0.0, 0.01)</td>\n",
       "      <td>1.703403</td>\n",
       "      <td>False</td>\n",
       "      <td>Interval_10427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483593</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[0.0, 0.01)</td>\n",
       "      <td>1.703403</td>\n",
       "      <td>False</td>\n",
       "      <td>Interval_10427</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>483594 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       min_value      Diff                                       Cut  \\\n",
       "0       2.220971  0.353334   [1.976722536187036, 2.6322967147827145)   \n",
       "1       2.335525  0.069759   [1.976722536187036, 2.6322967147827145)   \n",
       "2       2.244406  0.948601   [1.976722536187036, 2.6322967147827145)   \n",
       "3       1.966361  0.122825  [1.3211483574913574, 1.9767225360870362)   \n",
       "4       2.349192  0.415596   [1.976722536187036, 2.6322967147827145)   \n",
       "...          ...       ...                                       ...   \n",
       "483589       0.0  0.574085                               [0.0, 0.01)   \n",
       "483590       0.0  0.641075                               [0.0, 0.01)   \n",
       "483591       0.0  0.000000                               [0.0, 0.01)   \n",
       "483592       0.0  0.000000                               [0.0, 0.01)   \n",
       "483593       0.0  0.000000                               [0.0, 0.01)   \n",
       "\n",
       "         bigdiff  badpeak          PeakID  \n",
       "0       0.854268    False  Interval_77062  \n",
       "1       0.854268    False  Interval_77062  \n",
       "2       0.854268     True  Interval_77062  \n",
       "3       1.071294    False  Interval_77062  \n",
       "4       0.854268    False  Interval_77062  \n",
       "...          ...      ...             ...  \n",
       "483589  1.703403    False  Interval_10427  \n",
       "483590  1.703403    False  Interval_10427  \n",
       "483591  1.703403    False  Interval_10427  \n",
       "483592  1.703403    False  Interval_10427  \n",
       "483593  1.703403    False  Interval_10427  \n",
       "\n",
       "[483594 rows x 6 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AR_mins = pd.DataFrame(columns=AR_df.columns, index=parents)\n",
    "AR_maxs = pd.DataFrame(columns=AR_df.columns, index=parents)\n",
    "for cp in parents:\n",
    "    AR_both = AR_df.loc[[cp+'_R1',cp+'_R2']]\n",
    "    AR_mins.loc[cp] = AR_both.min()\n",
    "    AR_maxs.loc[cp] = AR_both.max()\n",
    "\n",
    "minmelt = AR_mins.melt()\n",
    "mindiff_df = pd.DataFrame(minmelt.value, index=minmelt.index)\n",
    "mindiff_df.columns = ['min_value']\n",
    "\n",
    "mindiff_df['Diff'] = (AR_maxs - AR_mins).melt().value.astype(float)\n",
    "\n",
    "val_max = mindiff_df.min_value.max() + 0.01\n",
    "int_size = (val_max - 0.01) / 10\n",
    "tuple_list = [(a+1e-10,a+int_size) for a in np.arange(0.01,val_max,int_size)]\n",
    "\n",
    "bins = pd.IntervalIndex.from_tuples([(0,0.01)] + tuple_list, closed='left')\n",
    "mindiff_df['Cut'] = pd.cut(mindiff_df.min_value, bins=bins)\n",
    "\n",
    "bigdiffs = mindiff_df.groupby('Cut').Diff.quantile(0.99).astype(float)\n",
    "mindiff_df['bigdiff'] = mindiff_df.Cut.map(bigdiffs).astype(float)\n",
    "\n",
    "mindiff_df['badpeak'] = (mindiff_df.Diff > mindiff_df.bigdiff)\n",
    "\n",
    "mindiff_df['PeakID'] = minmelt['PeakID']\n",
    "\n",
    "mindiff_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This produces about a half-million replicate-pair peak heights.  We organize each pair by the lowest peak height (among the two in the pair) and the difference in peak heights."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What we see is that replicate pairs (R1/R2) within a clone tend to have very close peak heights.  E.g., if the minimum peak height is 3, the maximum peak height is typically less than 3.5, and almost always less than 4.0.\n",
    "\n",
    "We can use this for quality control, throwing out peaks in which a replicate pair exhibits distant peak heights.  We give each peak a \"quality p-value\" based on the most distant of the replicate pairs.  If the most distant of the replicate pairs is in the 99th percentile (in peak-height difference), then the quality p-value would be 0.01.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PeakID</th>\n",
       "      <th>chr</th>\n",
       "      <th>end</th>\n",
       "      <th>in_ensg</th>\n",
       "      <th>in_gene</th>\n",
       "      <th>itype</th>\n",
       "      <th>mid</th>\n",
       "      <th>start</th>\n",
       "      <th>quality_p</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_77062</th>\n",
       "      <td>Interval_77062</td>\n",
       "      <td>9</td>\n",
       "      <td>115080094</td>\n",
       "      <td>ENSG00000119314</td>\n",
       "      <td>PTBP3</td>\n",
       "      <td>intron</td>\n",
       "      <td>115079731.5</td>\n",
       "      <td>115079369</td>\n",
       "      <td>0.104812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_18293</th>\n",
       "      <td>Interval_18293</td>\n",
       "      <td>12</td>\n",
       "      <td>117544007</td>\n",
       "      <td>ENSG00000258285</td>\n",
       "      <td>RP11-103B5.2</td>\n",
       "      <td>intron</td>\n",
       "      <td>117543889.5</td>\n",
       "      <td>117543772</td>\n",
       "      <td>0.521875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_51938</th>\n",
       "      <td>Interval_51938</td>\n",
       "      <td>3</td>\n",
       "      <td>156852899</td>\n",
       "      <td>None</td>\n",
       "      <td>nan</td>\n",
       "      <td>Intergen</td>\n",
       "      <td>156852750.0</td>\n",
       "      <td>156852601</td>\n",
       "      <td>0.028644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_33832</th>\n",
       "      <td>Interval_33832</td>\n",
       "      <td>19</td>\n",
       "      <td>1672563</td>\n",
       "      <td>None</td>\n",
       "      <td>nan</td>\n",
       "      <td>Intergen</td>\n",
       "      <td>1672247.0</td>\n",
       "      <td>1671931</td>\n",
       "      <td>0.016356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_13433</th>\n",
       "      <td>Interval_13433</td>\n",
       "      <td>11</td>\n",
       "      <td>82997741</td>\n",
       "      <td>ENSG00000137500</td>\n",
       "      <td>CCDC90B</td>\n",
       "      <td>promoter-TSS</td>\n",
       "      <td>82997180.5</td>\n",
       "      <td>82996620</td>\n",
       "      <td>0.010913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_8387</th>\n",
       "      <td>Interval_8387</td>\n",
       "      <td>10</td>\n",
       "      <td>33624658</td>\n",
       "      <td>ENSG00000099250</td>\n",
       "      <td>NRP1</td>\n",
       "      <td>promoter-TSS</td>\n",
       "      <td>33623619.5</td>\n",
       "      <td>33622581</td>\n",
       "      <td>0.067399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_17712</th>\n",
       "      <td>Interval_17712</td>\n",
       "      <td>12</td>\n",
       "      <td>96890243</td>\n",
       "      <td>ENSG00000188596</td>\n",
       "      <td>C12orf55</td>\n",
       "      <td>intron</td>\n",
       "      <td>96889859.5</td>\n",
       "      <td>96889476</td>\n",
       "      <td>0.029034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_31505</th>\n",
       "      <td>Interval_31505</td>\n",
       "      <td>17</td>\n",
       "      <td>73744122</td>\n",
       "      <td>ENSG00000132470</td>\n",
       "      <td>ITGB4</td>\n",
       "      <td>intron</td>\n",
       "      <td>73744009.0</td>\n",
       "      <td>73743896</td>\n",
       "      <td>0.076459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_18056</th>\n",
       "      <td>Interval_18056</td>\n",
       "      <td>12</td>\n",
       "      <td>110243675</td>\n",
       "      <td>ENSG00000111199</td>\n",
       "      <td>TRPV4</td>\n",
       "      <td>intron</td>\n",
       "      <td>110243346.5</td>\n",
       "      <td>110243018</td>\n",
       "      <td>0.081048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_10427</th>\n",
       "      <td>Interval_10427</td>\n",
       "      <td>10</td>\n",
       "      <td>114392835</td>\n",
       "      <td>ENSG00000151532</td>\n",
       "      <td>VTI1A</td>\n",
       "      <td>intron</td>\n",
       "      <td>114392701.5</td>\n",
       "      <td>114392568</td>\n",
       "      <td>0.273358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80599 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        PeakID chr        end          in_ensg       in_gene  \\\n",
       "PeakID                                                                         \n",
       "Interval_77062  Interval_77062   9  115080094  ENSG00000119314         PTBP3   \n",
       "Interval_18293  Interval_18293  12  117544007  ENSG00000258285  RP11-103B5.2   \n",
       "Interval_51938  Interval_51938   3  156852899             None           nan   \n",
       "Interval_33832  Interval_33832  19    1672563             None           nan   \n",
       "Interval_13433  Interval_13433  11   82997741  ENSG00000137500       CCDC90B   \n",
       "...                        ...  ..        ...              ...           ...   \n",
       "Interval_8387    Interval_8387  10   33624658  ENSG00000099250          NRP1   \n",
       "Interval_17712  Interval_17712  12   96890243  ENSG00000188596      C12orf55   \n",
       "Interval_31505  Interval_31505  17   73744122  ENSG00000132470         ITGB4   \n",
       "Interval_18056  Interval_18056  12  110243675  ENSG00000111199         TRPV4   \n",
       "Interval_10427  Interval_10427  10  114392835  ENSG00000151532         VTI1A   \n",
       "\n",
       "                       itype          mid      start  quality_p  \n",
       "PeakID                                                           \n",
       "Interval_77062        intron  115079731.5  115079369   0.104812  \n",
       "Interval_18293        intron  117543889.5  117543772   0.521875  \n",
       "Interval_51938      Intergen  156852750.0  156852601   0.028644  \n",
       "Interval_33832      Intergen    1672247.0    1671931   0.016356  \n",
       "Interval_13433  promoter-TSS   82997180.5   82996620   0.010913  \n",
       "...                      ...          ...        ...        ...  \n",
       "Interval_8387   promoter-TSS   33623619.5   33622581   0.067399  \n",
       "Interval_17712        intron   96889859.5   96889476   0.029034  \n",
       "Interval_31505        intron   73744009.0   73743896   0.076459  \n",
       "Interval_18056        intron  110243346.5  110243018   0.081048  \n",
       "Interval_10427        intron  114392701.5  114392568   0.273358  \n",
       "\n",
       "[80599 rows x 9 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_df = {}\n",
    "md_df = {}\n",
    "ATAC_data.var['quality_p'] = 1.0\n",
    "cuts = mindiff_df['Cut'].unique()\n",
    "\n",
    "for cut in cuts:\n",
    "    md_df[cut] = mindiff_df[mindiff_df['Cut'] == cut].copy()\n",
    "    diff_df[cut] = md_df[cut].Diff.sort_values()\n",
    "    md_df[cut]['ss'] = np.searchsorted(diff_df[cut], md_df[cut].Diff)\n",
    "    num_samps = len(diff_df[cut])\n",
    "    md_df[cut]['p_val'] = 1 - (md_df[cut]['ss'] / (1 + num_samps))\n",
    "    qp_series = md_df[cut].groupby('PeakID').p_val.min()\n",
    "    ATAC_data.var.loc[qp_series.index, 'quality_p'] = qp_series\n",
    "    \n",
    "ATAC_data.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-33-d76069ad77c5>:39: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "from matplotlib import collections  as mc\n",
    "\n",
    "sns.set_style('white')\n",
    "figsize=(60*mm, 45*mm)\n",
    "fig, ax = plt.subplots(figsize=figsize, dpi=300)\n",
    "setfontsize()\n",
    "qcols = {0.5:'steelblue', 0.9:'yellow', 0.99:'orange', 0.999:'red'}\n",
    "for quant in qcols.keys():\n",
    "    diffs = mindiff_df.groupby('Cut').quantile(quant).Diff\n",
    "    lines = [[(cut.left, diffs[cut]), (cut.right, diffs[cut])] for cut in cuts]\n",
    "    if quant == 0.5:\n",
    "        label = 'Median'\n",
    "        lw = 1.5\n",
    "    else:\n",
    "        if quant < 0.999:\n",
    "            label = '{:1.0f}th percentile'.format(quant*100)\n",
    "        else:\n",
    "            label = '{:1.1f}th percentile'.format(quant*100)\n",
    "        lw = 0.5\n",
    "    lc = mc.LineCollection(lines, linewidths=lw, color=qcols[quant], label=label)\n",
    "    ax.add_collection(lc)\n",
    "ax.legend()    \n",
    "\n",
    "sns.histplot(x='min_value', y='Diff', data=mindiff_df, bins=30, pmax=0.5, \n",
    "             cmap='Greens', cbar=False)\n",
    "\n",
    "\n",
    "plt.suptitle('{} Replicate peak-heights'.format(len(mindiff_df)))\n",
    "ax.set_xlabel('Minimum peak height within pair')\n",
    "ax.set_ylabel('$\\Delta$ of peak heights')\n",
    "ax.set_xticks([x + 0.25 for x in [0,1,2,3,4,5,6]])\n",
    "ax.set_xticklabels(['0','1','2','3','4','5','6'], y=0.05)\n",
    "ax.set_yticks([0,1,2,3])\n",
    "ax.set_yticklabels(['0','1','2','3'], x=0.035)\n",
    "\n",
    "set_size(fig, figsize)\n",
    "plt.savefig(\"Panels/Diff_dist.pdf\", bbox_inches='tight')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The \"quality_p\" column is very low (close to zero) when the peak height difference is very high when looking at one of the replicate pairs.  We might exclude those with quality_p < 0.05, for quality control purposes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-34-4e2d74c62c17>:3: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "desetfontsize()\n",
    "sns.histplot(ATAC_data.var.quality_p, bins=np.arange(0,1,0.01))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the distribution of quality p-values is not flat (as one would expect for p-values), since we are taking the minimum p-value among 6 (one for each of the six replicate pairs).\n",
    "\n",
    "We perform similar analysis to say whether a peak exhibits variability between replicate clones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PeakID</th>\n",
       "      <th>chr</th>\n",
       "      <th>end</th>\n",
       "      <th>in_ensg</th>\n",
       "      <th>in_gene</th>\n",
       "      <th>itype</th>\n",
       "      <th>mid</th>\n",
       "      <th>start</th>\n",
       "      <th>quality_p</th>\n",
       "      <th>clonal_p</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_77062</th>\n",
       "      <td>Interval_77062</td>\n",
       "      <td>9</td>\n",
       "      <td>115080094</td>\n",
       "      <td>ENSG00000119314</td>\n",
       "      <td>PTBP3</td>\n",
       "      <td>intron</td>\n",
       "      <td>115079731.5</td>\n",
       "      <td>115079369</td>\n",
       "      <td>0.104812</td>\n",
       "      <td>0.117727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_18293</th>\n",
       "      <td>Interval_18293</td>\n",
       "      <td>12</td>\n",
       "      <td>117544007</td>\n",
       "      <td>ENSG00000258285</td>\n",
       "      <td>RP11-103B5.2</td>\n",
       "      <td>intron</td>\n",
       "      <td>117543889.5</td>\n",
       "      <td>117543772</td>\n",
       "      <td>0.521875</td>\n",
       "      <td>0.163279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_51938</th>\n",
       "      <td>Interval_51938</td>\n",
       "      <td>3</td>\n",
       "      <td>156852899</td>\n",
       "      <td>None</td>\n",
       "      <td>nan</td>\n",
       "      <td>Intergen</td>\n",
       "      <td>156852750.0</td>\n",
       "      <td>156852601</td>\n",
       "      <td>0.028644</td>\n",
       "      <td>0.001930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_33832</th>\n",
       "      <td>Interval_33832</td>\n",
       "      <td>19</td>\n",
       "      <td>1672563</td>\n",
       "      <td>None</td>\n",
       "      <td>nan</td>\n",
       "      <td>Intergen</td>\n",
       "      <td>1672247.0</td>\n",
       "      <td>1671931</td>\n",
       "      <td>0.016356</td>\n",
       "      <td>0.320371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_13433</th>\n",
       "      <td>Interval_13433</td>\n",
       "      <td>11</td>\n",
       "      <td>82997741</td>\n",
       "      <td>ENSG00000137500</td>\n",
       "      <td>CCDC90B</td>\n",
       "      <td>promoter-TSS</td>\n",
       "      <td>82997180.5</td>\n",
       "      <td>82996620</td>\n",
       "      <td>0.010913</td>\n",
       "      <td>0.055141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_8387</th>\n",
       "      <td>Interval_8387</td>\n",
       "      <td>10</td>\n",
       "      <td>33624658</td>\n",
       "      <td>ENSG00000099250</td>\n",
       "      <td>NRP1</td>\n",
       "      <td>promoter-TSS</td>\n",
       "      <td>33623619.5</td>\n",
       "      <td>33622581</td>\n",
       "      <td>0.067399</td>\n",
       "      <td>0.001397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_17712</th>\n",
       "      <td>Interval_17712</td>\n",
       "      <td>12</td>\n",
       "      <td>96890243</td>\n",
       "      <td>ENSG00000188596</td>\n",
       "      <td>C12orf55</td>\n",
       "      <td>intron</td>\n",
       "      <td>96889859.5</td>\n",
       "      <td>96889476</td>\n",
       "      <td>0.029034</td>\n",
       "      <td>0.249430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_31505</th>\n",
       "      <td>Interval_31505</td>\n",
       "      <td>17</td>\n",
       "      <td>73744122</td>\n",
       "      <td>ENSG00000132470</td>\n",
       "      <td>ITGB4</td>\n",
       "      <td>intron</td>\n",
       "      <td>73744009.0</td>\n",
       "      <td>73743896</td>\n",
       "      <td>0.076459</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_18056</th>\n",
       "      <td>Interval_18056</td>\n",
       "      <td>12</td>\n",
       "      <td>110243675</td>\n",
       "      <td>ENSG00000111199</td>\n",
       "      <td>TRPV4</td>\n",
       "      <td>intron</td>\n",
       "      <td>110243346.5</td>\n",
       "      <td>110243018</td>\n",
       "      <td>0.081048</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_10427</th>\n",
       "      <td>Interval_10427</td>\n",
       "      <td>10</td>\n",
       "      <td>114392835</td>\n",
       "      <td>ENSG00000151532</td>\n",
       "      <td>VTI1A</td>\n",
       "      <td>intron</td>\n",
       "      <td>114392701.5</td>\n",
       "      <td>114392568</td>\n",
       "      <td>0.273358</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80599 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        PeakID chr        end          in_ensg       in_gene  \\\n",
       "PeakID                                                                         \n",
       "Interval_77062  Interval_77062   9  115080094  ENSG00000119314         PTBP3   \n",
       "Interval_18293  Interval_18293  12  117544007  ENSG00000258285  RP11-103B5.2   \n",
       "Interval_51938  Interval_51938   3  156852899             None           nan   \n",
       "Interval_33832  Interval_33832  19    1672563             None           nan   \n",
       "Interval_13433  Interval_13433  11   82997741  ENSG00000137500       CCDC90B   \n",
       "...                        ...  ..        ...              ...           ...   \n",
       "Interval_8387    Interval_8387  10   33624658  ENSG00000099250          NRP1   \n",
       "Interval_17712  Interval_17712  12   96890243  ENSG00000188596      C12orf55   \n",
       "Interval_31505  Interval_31505  17   73744122  ENSG00000132470         ITGB4   \n",
       "Interval_18056  Interval_18056  12  110243675  ENSG00000111199         TRPV4   \n",
       "Interval_10427  Interval_10427  10  114392835  ENSG00000151532         VTI1A   \n",
       "\n",
       "                       itype          mid      start  quality_p  clonal_p  \n",
       "PeakID                                                                     \n",
       "Interval_77062        intron  115079731.5  115079369   0.104812  0.117727  \n",
       "Interval_18293        intron  117543889.5  117543772   0.521875  0.163279  \n",
       "Interval_51938      Intergen  156852750.0  156852601   0.028644  0.001930  \n",
       "Interval_33832      Intergen    1672247.0    1671931   0.016356  0.320371  \n",
       "Interval_13433  promoter-TSS   82997180.5   82996620   0.010913  0.055141  \n",
       "...                      ...          ...        ...        ...       ...  \n",
       "Interval_8387   promoter-TSS   33623619.5   33622581   0.067399  0.001397  \n",
       "Interval_17712        intron   96889859.5   96889476   0.029034  0.249430  \n",
       "Interval_31505        intron   73744009.0   73743896   0.076459  1.000000  \n",
       "Interval_18056        intron  110243346.5  110243018   0.081048  1.000000  \n",
       "Interval_10427        intron  114392701.5  114392568   0.273358  1.000000  \n",
       "\n",
       "[80599 rows x 10 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATACR_mins = ATACR_data.to_df().min()\n",
    "ATACR_diffs = ATACR_data.to_df().max() - ATACR_mins\n",
    "\n",
    "peak_mindiffs = pd.DataFrame(ATACR_mins, columns=['min_value'])\n",
    "peak_mindiffs['Diff'] = ATACR_diffs\n",
    "peak_mindiffs['Cut'] = pd.cut(peak_mindiffs.min_value, bins=bins)\n",
    "\n",
    "p_df = {}\n",
    "\n",
    "ATAC_data.var['clonal_p'] = 1.0\n",
    "\n",
    "for cut in cuts:\n",
    "    p_df[cut] = peak_mindiffs[peak_mindiffs['Cut'] == cut].copy()\n",
    "    diff_df[cut] \n",
    "    p_df[cut]['ss'] = np.searchsorted(diff_df[cut], p_df[cut].Diff)\n",
    "    num_samps = len(diff_df[cut])\n",
    "    p_df[cut]['p_val'] = np.clip(60 * (1 - (p_df[cut]['ss'] / (1+num_samps))),0,1) \n",
    "    # 12 * 10 / 2 pairwise comparisons.\n",
    "    cp_series = p_df[cut].p_val\n",
    "    ATAC_data.var.loc[cp_series.index, 'clonal_p'] = cp_series\n",
    "    \n",
    "ATAC_data.var\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-89-d4a5f65f06fe>:38: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "CQ_df = ATAC_data.var[['clonal_p','quality_p']]\n",
    "qbins = np.arange(0,1.001,0.1)\n",
    "qcuts = pd.cut(CQ_df.quality_p, qbins)\n",
    "iindex = pd.IntervalIndex(qcuts.unique()) #.from_tuples([(x,x+1 - 1e-10) for x in np.arange(0,1,0.1)])\n",
    "plot_df = pd.DataFrame(index = iindex)\n",
    "plot_df['nsig'] = CQ_df.groupby(qcuts).apply(lambda x : (x < 0.01).sum()).clonal_p\n",
    "plot_df['count'] = CQ_df.groupby(qcuts).count().clonal_p\n",
    "plot_df['ratio'] = plot_df['nsig'] / plot_df['count']\n",
    "plot_df['interval'] = plot_df.index#.apply(lambda x : x.left())\n",
    "plot_df['mid'] = plot_df['interval'].apply(lambda x : x.mid)\n",
    "plot_df = plot_df.sort_values(by='mid')\n",
    "\n",
    "sns.set_style('whitegrid')\n",
    "figsize=(60*mm, 45*mm)\n",
    "fig, ax = plt.subplots(figsize=figsize, dpi=300)\n",
    "setfontsize()\n",
    "\n",
    "sns.barplot(x='mid', y='count', data=plot_df, color='steelblue', alpha=0.25, ax=ax, label='All peaks')\n",
    "sns.barplot(x='mid', y='nsig', data=plot_df, color='orange', label='Variable peaks ($p<0.01$)',ax=ax)\n",
    "#for row in plot_df:\n",
    "#    ax.text(row['interval'], row['nsig'], 'Hello')\n",
    "for j in range(len(plot_df)):\n",
    "    nsigj = plot_df.iloc[j]['nsig']\n",
    "    ratioj = plot_df.iloc[j]['ratio']\n",
    "    ax.text(j,nsigj,'{:1.0%}'.format(ratioj), ha='center')\n",
    "\n",
    "ax.set_xticks([])\n",
    "ax.set_xticklabels([]) #plot_df['interval'])\n",
    "ax.legend()\n",
    "plt.xticks(rotation=45)\n",
    "\n",
    "ax.set_xlabel('Quality p-value (replicate closeness)')\n",
    "ax.set_ylabel('Number of peaks')\n",
    "plt.suptitle('Over 15% of high-quality peaks are variable')\n",
    "\n",
    "set_size(fig, figsize)\n",
    "plt.savefig(\"Panels/Clonal_peaks.pdf\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Interval_13433', 'Interval_27763', 'Interval_4280', 'Interval_6501',\n",
       "       'Interval_68580', 'Interval_72287', 'Interval_7590', 'Interval_38287',\n",
       "       'Interval_49328', 'Interval_1833',\n",
       "       ...\n",
       "       'Interval_9823', 'Interval_63568', 'Interval_54425', 'Interval_33873',\n",
       "       'Interval_2270', 'Interval_69022', 'Interval_56910', 'Interval_57434',\n",
       "       'Interval_28417', 'Interval_8387'],\n",
       "      dtype='object', name='PeakID', length=16370)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tall_peaks = (ATACR_data.to_df().max() > np.log1p(50))\n",
    "tall_peaks = tall_peaks[tall_peaks].index\n",
    "tall_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-38-126e96806b74>:39: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "CQ_df = ATAC_data.var[['clonal_p','quality_p']]\n",
    "CQ_df = CQ_df.loc[tall_peaks]\n",
    "qbins = np.arange(0,1.001,0.1)\n",
    "qcuts = pd.cut(CQ_df.quality_p, qbins)\n",
    "iindex = pd.IntervalIndex(qcuts.unique()) #.from_tuples([(x,x+1 - 1e-10) for x in np.arange(0,1,0.1)])\n",
    "plot_df = pd.DataFrame(index = iindex)\n",
    "plot_df['nsig'] = CQ_df.groupby(qcuts).apply(lambda x : (x < 0.01).sum()).clonal_p\n",
    "plot_df['count'] = CQ_df.groupby(qcuts).count().clonal_p\n",
    "plot_df['ratio'] = plot_df['nsig'] / plot_df['count']\n",
    "plot_df['interval'] = plot_df.index#.apply(lambda x : x.left())\n",
    "plot_df['mid'] = plot_df['interval'].apply(lambda x : x.mid)\n",
    "plot_df = plot_df.sort_values(by='mid')\n",
    "\n",
    "sns.set_style('whitegrid')\n",
    "figsize=(60*mm, 45*mm)\n",
    "fig, ax = plt.subplots(figsize=figsize, dpi=300)\n",
    "setfontsize()\n",
    "\n",
    "sns.barplot(x='mid', y='count', data=plot_df, color='steelblue', alpha=0.25, ax=ax, label='All tall peaks (ht > 50)')\n",
    "sns.barplot(x='mid', y='nsig', data=plot_df, color='orange', label='Variable peaks ($p<0.01$)',ax=ax)\n",
    "#for row in plot_df:\n",
    "#    ax.text(row['interval'], row['nsig'], 'Hello')\n",
    "for j in range(len(plot_df)):\n",
    "    nsigj = plot_df.iloc[j]['nsig']\n",
    "    ratioj = plot_df.iloc[j]['ratio']\n",
    "    ax.text(j,nsigj,'{:1.0%}'.format(ratioj), ha='center')\n",
    "\n",
    "ax.set_xticks([])\n",
    "ax.set_xticklabels([]) #plot_df['interval'])\n",
    "ax.legend()\n",
    "plt.xticks(rotation=45)\n",
    "\n",
    "ax.set_xlabel('Quality p-value (replicate closeness)')\n",
    "ax.set_ylabel('Number of peaks')\n",
    "plt.suptitle('Over 20% of tall peaks are variable')\n",
    "\n",
    "set_size(fig, figsize)\n",
    "plt.savefig(\"Panels/ClonalTall_peaks.pdf\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We consider the \"good\" peaks to be those with quality p-value > 0.1.  In other words, these are peaks whose intra-replicate distances are all below the 90th percentile.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_good = ATAC_data[:,ATAC_data.var.quality_p > 0.1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_look = ATAC_good.var.clonal_p.sort_values()[:1000].index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is a typical clonal high-quality peak."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PeakID\n",
       "Interval_32862    0.000504\n",
       "Interval_38111    0.000504\n",
       "Interval_28339    0.000504\n",
       "Interval_63852    0.000504\n",
       "Interval_21089    0.000504\n",
       "Interval_39532    0.000504\n",
       "Interval_74596    0.000504\n",
       "Interval_60161    0.000504\n",
       "Interval_38170    0.000504\n",
       "Interval_53024    0.000504\n",
       "Name: clonal_p, dtype: float64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATAC_data.var.loc[peak_look].clonal_p.sort_values()[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-16-afdeac0540c6>:18: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "desetfontsize()\n",
    "ATAC_plot('Interval_35602')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-16-afdeac0540c6>:18: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "ATAC_plot('Interval_6767')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Panel:  Peak heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-94-14b3260dc3e0>:45: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "var_peaks = ATAC_good.var.clonal_p.sort_values()[:1000].index\n",
    "\n",
    "plot_df = ATAC_good.to_df()[var_peaks].copy()\n",
    "\n",
    "plot_df.index = ATAC_good.obs.ATAC_clone\n",
    "plot_df.sort_index(inplace=True)\n",
    "\n",
    "plot_df = (plot_df - plot_df.mean()) / plot_df.std()\n",
    "\n",
    "#display(plot_df.idxmax().sort_values())\n",
    "\n",
    "var_peaks = plot_df.idxmax().sort_values().index\n",
    "plot_df = plot_df[var_peaks]\n",
    "\n",
    "figsize=(60*mm, 90*mm)\n",
    "fig,ax = plt.subplots(1,1,figsize=figsize, dpi=300)\n",
    "setfontsize()\n",
    "\n",
    "sns.heatmap(plot_df, vmin=-2.5, vmax=2.5, cmap='coolwarm', \n",
    "            xticklabels=False, yticklabels=False,\n",
    "            cbar=False, ax=ax)\n",
    "\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "for y,clone in [(11,1), (5,15), (15,22), (17,23), (26,8), (28,9)]:\n",
    "    ax.text(-0.1,y,str(clone) + ' {', va='center', ha='right', fontsize=7)\n",
    "    \n",
    "for j,clone in enumerate(plot_df.index):\n",
    "    par = clone.split('_')[0]\n",
    "    if par not in ['1','15','22','23','8','9','5a','5b']:\n",
    "        ax.text(-0.1,j+0.5,par + ' ', va='center', ha='right', fontsize=5)\n",
    "    \n",
    "ax.text(-0.1,22.5,'5a  {', va='center', ha='right', fontsize=6)\n",
    "ax.text(-0.1,23.5,'5b  {', va='center', ha='right', fontsize = 6)\n",
    "\n",
    "from matplotlib import rc\n",
    "\n",
    "rc('text', usetex=True)\n",
    "\n",
    "#ax.text(0.5,0,r'\\underline{Clone}', va='bottom', ha='right', fontsize=6)\n",
    "plt.suptitle('{} clonal peaks, {} samples'.format(len(var_peaks), len(plot_df)), y=0.95)\n",
    "set_size(fig, figsize)\n",
    "plt.savefig(\"Panels/PeakHeatmap.pdf\", bbox_inches='tight')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Peak-gene correlations\n",
    "\n",
    "Now we create a giant correlation matrix -- one for each chromosome -- between gene expression (clonal average) and peak height (clonal average), for all genes and peaks on that chromosome.  We only use the \"good clones\" (16 out of 24 clones) found in preprocessing, when computing these correlations.  We also restrict to \"good\" peaks, where replicates have close peak heights."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['WASH7P', 'AL627309.1', 'RP11-34P13.13', 'MTND2P28', 'hsa-mir-6723',\n",
       "       'RP5-857K21.7', 'MTATP8P1', 'MTATP6P1', 'RP5-857K21.11',\n",
       "       'RP11-206L10.9',\n",
       "       ...\n",
       "       'VBP1-1', 'RAB39B-1', 'TMLHE-1', 'VAMP7', 'IL9R', 'EIF4A1P2', 'SRIP3',\n",
       "       'AC010970.2', 'RNA5-8SP6', 'PSMA6P1'],\n",
       "      dtype='object', name='gene_name', length=11881)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = adata.var_names\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Interval_77062', 'Interval_18293', 'Interval_48454', 'Interval_27763',\n",
       "       'Interval_53540', 'Interval_46787', 'Interval_23335', 'Interval_40339',\n",
       "       'Interval_76361', 'Interval_15915',\n",
       "       ...\n",
       "       'Interval_27261', 'Interval_76540', 'Interval_78442', 'Interval_57434',\n",
       "       'Interval_58433', 'Interval_62113', 'Interval_45676', 'Interval_43639',\n",
       "       'Interval_35777', 'Interval_10427'],\n",
       "      dtype='object', name='PeakID', length=63909)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks = ATAC_good.var_names\n",
    "peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X', 'Y']\n"
     ]
    }
   ],
   "source": [
    "chromes = [str(j) for j in range(1,23)] + ['X','Y']\n",
    "print(chromes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the good clones, we create dataframes of the clonal averages for gene expression and peak height."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_clones = adata.obs[adata.obs['Good clone'] == True].Clone_ID.unique()\n",
    "test_clones = [clone for clone in test_clones if clone in ATAC_data.obs.Clone_ID.unique()]\n",
    "\n",
    "pd_peak = ATAC_data[ATAC_data.obs['Clone_ID'].isin(test_clones)]\n",
    "pd_gene = adata[adata.obs['Clone_ID'].isin(test_clones)]\n",
    "\n",
    "gene_clonal = pd_gene.to_df().groupby(pd_gene.obs.Clone_ID).mean()\n",
    "peak_clonal = pd_peak.to_df().groupby(pd_peak.obs.Clone_ID).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>WASH7P</th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>RP11-34P13.13</th>\n",
       "      <th>MTND2P28</th>\n",
       "      <th>hsa-mir-6723</th>\n",
       "      <th>RP5-857K21.7</th>\n",
       "      <th>MTATP8P1</th>\n",
       "      <th>MTATP6P1</th>\n",
       "      <th>RP5-857K21.11</th>\n",
       "      <th>RP11-206L10.9</th>\n",
       "      <th>...</th>\n",
       "      <th>VBP1-1</th>\n",
       "      <th>RAB39B-1</th>\n",
       "      <th>TMLHE-1</th>\n",
       "      <th>VAMP7</th>\n",
       "      <th>IL9R</th>\n",
       "      <th>EIF4A1P2</th>\n",
       "      <th>SRIP3</th>\n",
       "      <th>AC010970.2</th>\n",
       "      <th>RNA5-8SP6</th>\n",
       "      <th>PSMA6P1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Clone_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P1_A03</th>\n",
       "      <td>0.618283</td>\n",
       "      <td>0.945561</td>\n",
       "      <td>1.025549</td>\n",
       "      <td>3.235500</td>\n",
       "      <td>2.083895</td>\n",
       "      <td>0.750680</td>\n",
       "      <td>0.459195</td>\n",
       "      <td>5.755206</td>\n",
       "      <td>0.250811</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.447827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.491540</td>\n",
       "      <td>6.902879</td>\n",
       "      <td>1.868600</td>\n",
       "      <td>0.802588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A05</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.712187</td>\n",
       "      <td>1.456248</td>\n",
       "      <td>0.407113</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.386560</td>\n",
       "      <td>0.229657</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.643641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.078488</td>\n",
       "      <td>2.631706</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.135689</td>\n",
       "      <td>1.724590</td>\n",
       "      <td>6.760322</td>\n",
       "      <td>1.604085</td>\n",
       "      <td>0.302625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A08</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.053064</td>\n",
       "      <td>1.616868</td>\n",
       "      <td>0.697563</td>\n",
       "      <td>0.414321</td>\n",
       "      <td>5.804655</td>\n",
       "      <td>0.227296</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.286690</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.964008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.283545</td>\n",
       "      <td>6.123337</td>\n",
       "      <td>1.978948</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A12</th>\n",
       "      <td>0.190624</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.038536</td>\n",
       "      <td>1.861662</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.472903</td>\n",
       "      <td>5.478726</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.888312</td>\n",
       "      <td>...</td>\n",
       "      <td>4.638396</td>\n",
       "      <td>0.864891</td>\n",
       "      <td>3.823669</td>\n",
       "      <td>2.864282</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.230631</td>\n",
       "      <td>1.922047</td>\n",
       "      <td>6.479896</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.199741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_B10</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.115203</td>\n",
       "      <td>2.256257</td>\n",
       "      <td>0.699134</td>\n",
       "      <td>0.425098</td>\n",
       "      <td>5.694868</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.665425</td>\n",
       "      <td>0.772898</td>\n",
       "      <td>3.731382</td>\n",
       "      <td>2.401685</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.325636</td>\n",
       "      <td>1.543150</td>\n",
       "      <td>6.618936</td>\n",
       "      <td>3.252234</td>\n",
       "      <td>0.182327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_C11</th>\n",
       "      <td>0.195251</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.748705</td>\n",
       "      <td>1.980835</td>\n",
       "      <td>0.540095</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.840600</td>\n",
       "      <td>0.252954</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.885841</td>\n",
       "      <td>0.892146</td>\n",
       "      <td>1.207822</td>\n",
       "      <td>2.805009</td>\n",
       "      <td>2.455279</td>\n",
       "      <td>0.105118</td>\n",
       "      <td>1.013093</td>\n",
       "      <td>6.190698</td>\n",
       "      <td>1.319050</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_D09</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.579358</td>\n",
       "      <td>1.508479</td>\n",
       "      <td>0.527772</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>5.743250</td>\n",
       "      <td>0.213638</td>\n",
       "      <td>0.949451</td>\n",
       "      <td>...</td>\n",
       "      <td>4.358072</td>\n",
       "      <td>0.686928</td>\n",
       "      <td>1.850092</td>\n",
       "      <td>2.886688</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.248796</td>\n",
       "      <td>1.610327</td>\n",
       "      <td>6.392721</td>\n",
       "      <td>2.055962</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_F10</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.925440</td>\n",
       "      <td>1.696129</td>\n",
       "      <td>0.626677</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.868581</td>\n",
       "      <td>0.273484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.180732</td>\n",
       "      <td>0.902162</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.396651</td>\n",
       "      <td>1.221312</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.020712</td>\n",
       "      <td>6.250185</td>\n",
       "      <td>1.113877</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_H01</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.109386</td>\n",
       "      <td>1.951495</td>\n",
       "      <td>0.371031</td>\n",
       "      <td>0.505814</td>\n",
       "      <td>5.827295</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.212268</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.494714</td>\n",
       "      <td>1.758571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.138110</td>\n",
       "      <td>1.123212</td>\n",
       "      <td>6.057616</td>\n",
       "      <td>1.935471</td>\n",
       "      <td>0.426746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_H10</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.253430</td>\n",
       "      <td>1.487985</td>\n",
       "      <td>0.211150</td>\n",
       "      <td>0.885398</td>\n",
       "      <td>5.617558</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.177901</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.762414</td>\n",
       "      <td>1.625854</td>\n",
       "      <td>0.302063</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.940247</td>\n",
       "      <td>6.774329</td>\n",
       "      <td>1.562626</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_A04</th>\n",
       "      <td>0.577054</td>\n",
       "      <td>0.203521</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.239484</td>\n",
       "      <td>1.336314</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.568577</td>\n",
       "      <td>0.234326</td>\n",
       "      <td>0.957820</td>\n",
       "      <td>...</td>\n",
       "      <td>2.828667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.955809</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.326913</td>\n",
       "      <td>6.448241</td>\n",
       "      <td>1.566463</td>\n",
       "      <td>0.393528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_B11</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.310599</td>\n",
       "      <td>2.166867</td>\n",
       "      <td>0.973400</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.709068</td>\n",
       "      <td>0.615366</td>\n",
       "      <td>1.189926</td>\n",
       "      <td>...</td>\n",
       "      <td>4.456919</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.221027</td>\n",
       "      <td>2.097311</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.121822</td>\n",
       "      <td>2.415108</td>\n",
       "      <td>7.193824</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.181933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_G01</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.007948</td>\n",
       "      <td>0.545674</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.469179</td>\n",
       "      <td>4.773428</td>\n",
       "      <td>0.257901</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.927383</td>\n",
       "      <td>0.352053</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.155642</td>\n",
       "      <td>2.634030</td>\n",
       "      <td>7.148538</td>\n",
       "      <td>2.947898</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_G09</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.711262</td>\n",
       "      <td>1.457363</td>\n",
       "      <td>0.747793</td>\n",
       "      <td>0.461915</td>\n",
       "      <td>5.398914</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.849241</td>\n",
       "      <td>2.137543</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.615978</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.978685</td>\n",
       "      <td>6.366894</td>\n",
       "      <td>0.950816</td>\n",
       "      <td>0.414203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_H04</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.317622</td>\n",
       "      <td>1.713998</td>\n",
       "      <td>0.343014</td>\n",
       "      <td>0.477543</td>\n",
       "      <td>5.297565</td>\n",
       "      <td>0.299365</td>\n",
       "      <td>0.352106</td>\n",
       "      <td>...</td>\n",
       "      <td>2.516015</td>\n",
       "      <td>1.169482</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.435665</td>\n",
       "      <td>6.590279</td>\n",
       "      <td>1.350423</td>\n",
       "      <td>0.322009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_H05</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.048599</td>\n",
       "      <td>0.480044</td>\n",
       "      <td>0.387329</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.366103</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.405826</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.193881</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.825652</td>\n",
       "      <td>7.100769</td>\n",
       "      <td>2.881787</td>\n",
       "      <td>0.234537</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows × 11881 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name    WASH7P  AL627309.1  RP11-34P13.13  MTND2P28  hsa-mir-6723  \\\n",
       "Clone_ID                                                                 \n",
       "P1_A03     0.618283    0.945561       1.025549  3.235500      2.083895   \n",
       "P1_A05     0.000000    0.000000       0.000000  2.712187      1.456248   \n",
       "P1_A08     0.000000    0.000000       0.000000  3.053064      1.616868   \n",
       "P1_A12     0.190624    0.000000       0.000000  3.038536      1.861662   \n",
       "P1_B10     0.000000    0.000000       0.000000  3.115203      2.256257   \n",
       "P1_C11     0.195251    0.000000       0.000000  2.748705      1.980835   \n",
       "P1_D09     0.000000    0.000000       0.000000  2.579358      1.508479   \n",
       "P1_F10     0.000000    0.000000       0.000000  2.925440      1.696129   \n",
       "P1_H01     0.000000    0.000000       0.000000  3.109386      1.951495   \n",
       "P1_H10     0.000000    0.000000       0.000000  2.253430      1.487985   \n",
       "P2_A04     0.577054    0.203521       0.000000  3.239484      1.336314   \n",
       "P2_B11     0.000000    0.000000       0.000000  3.310599      2.166867   \n",
       "P2_G01     0.000000    0.000000       0.000000  3.007948      0.545674   \n",
       "P2_G09     0.000000    0.000000       0.000000  2.711262      1.457363   \n",
       "P2_H04     0.000000    0.000000       0.000000  2.317622      1.713998   \n",
       "P2_H05     0.000000    0.000000       0.000000  2.048599      0.480044   \n",
       "\n",
       "gene_name  RP5-857K21.7  MTATP8P1  MTATP6P1  RP5-857K21.11  RP11-206L10.9  \\\n",
       "Clone_ID                                                                    \n",
       "P1_A03         0.750680  0.459195  5.755206       0.250811       0.000000   \n",
       "P1_A05         0.407113  0.000000  5.386560       0.229657       0.000000   \n",
       "P1_A08         0.697563  0.414321  5.804655       0.227296       0.000000   \n",
       "P1_A12         0.000000  0.472903  5.478726       0.000000       0.888312   \n",
       "P1_B10         0.699134  0.425098  5.694868       0.000000       0.000000   \n",
       "P1_C11         0.540095  0.000000  5.840600       0.252954       0.000000   \n",
       "P1_D09         0.527772  0.454545  5.743250       0.213638       0.949451   \n",
       "P1_F10         0.626677  0.000000  5.868581       0.273484       0.000000   \n",
       "P1_H01         0.371031  0.505814  5.827295       0.000000       0.000000   \n",
       "P1_H10         0.211150  0.885398  5.617558       0.000000       0.000000   \n",
       "P2_A04         0.000000  0.000000  5.568577       0.234326       0.957820   \n",
       "P2_B11         0.973400  0.000000  5.709068       0.615366       1.189926   \n",
       "P2_G01         0.000000  0.469179  4.773428       0.257901       0.000000   \n",
       "P2_G09         0.747793  0.461915  5.398914       0.000000       0.000000   \n",
       "P2_H04         0.343014  0.477543  5.297565       0.299365       0.352106   \n",
       "P2_H05         0.387329  0.000000  4.366103       0.000000       0.000000   \n",
       "\n",
       "gene_name  ...    VBP1-1  RAB39B-1   TMLHE-1     VAMP7      IL9R  EIF4A1P2  \\\n",
       "Clone_ID   ...                                                               \n",
       "P1_A03     ...  3.447827  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "P1_A05     ...  4.643641  0.000000  1.078488  2.631706  0.000000  0.135689   \n",
       "P1_A08     ...  4.286690  0.000000  2.964008  0.000000  0.000000  0.000000   \n",
       "P1_A12     ...  4.638396  0.864891  3.823669  2.864282  0.000000  0.230631   \n",
       "P1_B10     ...  2.665425  0.772898  3.731382  2.401685  0.000000  0.325636   \n",
       "P1_C11     ...  3.885841  0.892146  1.207822  2.805009  2.455279  0.105118   \n",
       "P1_D09     ...  4.358072  0.686928  1.850092  2.886688  0.000000  0.248796   \n",
       "P1_F10     ...  4.180732  0.902162  0.000000  1.396651  1.221312  0.000000   \n",
       "P1_H01     ...  4.212268  0.000000  1.494714  1.758571  0.000000  0.138110   \n",
       "P1_H10     ...  4.177901  0.000000  0.762414  1.625854  0.302063  0.000000   \n",
       "P2_A04     ...  2.828667  0.000000  2.955809  0.000000  0.000000  0.000000   \n",
       "P2_B11     ...  4.456919  0.000000  4.221027  2.097311  0.000000  0.121822   \n",
       "P2_G01     ...  0.000000  0.000000  1.927383  0.352053  0.000000  0.155642   \n",
       "P2_G09     ...  4.849241  2.137543  0.000000  2.615978  0.000000  0.000000   \n",
       "P2_H04     ...  2.516015  1.169482  0.000000  0.000000  0.000000  0.000000   \n",
       "P2_H05     ...  0.000000  1.405826  0.000000  1.193881  0.000000  0.000000   \n",
       "\n",
       "gene_name     SRIP3  AC010970.2  RNA5-8SP6   PSMA6P1  \n",
       "Clone_ID                                              \n",
       "P1_A03     1.491540    6.902879   1.868600  0.802588  \n",
       "P1_A05     1.724590    6.760322   1.604085  0.302625  \n",
       "P1_A08     1.283545    6.123337   1.978948  0.000000  \n",
       "P1_A12     1.922047    6.479896   0.000000  0.199741  \n",
       "P1_B10     1.543150    6.618936   3.252234  0.182327  \n",
       "P1_C11     1.013093    6.190698   1.319050  0.000000  \n",
       "P1_D09     1.610327    6.392721   2.055962  0.000000  \n",
       "P1_F10     2.020712    6.250185   1.113877  0.000000  \n",
       "P1_H01     1.123212    6.057616   1.935471  0.426746  \n",
       "P1_H10     2.940247    6.774329   1.562626  0.000000  \n",
       "P2_A04     2.326913    6.448241   1.566463  0.393528  \n",
       "P2_B11     2.415108    7.193824   0.000000  0.181933  \n",
       "P2_G01     2.634030    7.148538   2.947898  0.000000  \n",
       "P2_G09     1.978685    6.366894   0.950816  0.414203  \n",
       "P2_H04     1.435665    6.590279   1.350423  0.322009  \n",
       "P2_H05     1.825652    7.100769   2.881787  0.234537  \n",
       "\n",
       "[16 rows x 11881 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_clonal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_77062</th>\n",
       "      <th>Interval_18293</th>\n",
       "      <th>Interval_51938</th>\n",
       "      <th>Interval_33832</th>\n",
       "      <th>Interval_13433</th>\n",
       "      <th>Interval_48454</th>\n",
       "      <th>Interval_27763</th>\n",
       "      <th>Interval_53540</th>\n",
       "      <th>Interval_46169</th>\n",
       "      <th>Interval_46787</th>\n",
       "      <th>...</th>\n",
       "      <th>Interval_58433</th>\n",
       "      <th>Interval_62113</th>\n",
       "      <th>Interval_45676</th>\n",
       "      <th>Interval_43639</th>\n",
       "      <th>Interval_35777</th>\n",
       "      <th>Interval_8387</th>\n",
       "      <th>Interval_17712</th>\n",
       "      <th>Interval_31505</th>\n",
       "      <th>Interval_18056</th>\n",
       "      <th>Interval_10427</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Clone_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P1_A03</th>\n",
       "      <td>2.397638</td>\n",
       "      <td>0.396998</td>\n",
       "      <td>2.529633</td>\n",
       "      <td>2.150198</td>\n",
       "      <td>5.617923</td>\n",
       "      <td>1.967855</td>\n",
       "      <td>3.278499</td>\n",
       "      <td>3.233557</td>\n",
       "      <td>0.939784</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.185111</td>\n",
       "      <td>2.035630</td>\n",
       "      <td>3.081982</td>\n",
       "      <td>0.561223</td>\n",
       "      <td>0.659377</td>\n",
       "      <td>4.251013</td>\n",
       "      <td>1.816327</td>\n",
       "      <td>1.216354</td>\n",
       "      <td>1.990757</td>\n",
       "      <td>0.324315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A05</th>\n",
       "      <td>2.717578</td>\n",
       "      <td>2.192299</td>\n",
       "      <td>1.972389</td>\n",
       "      <td>2.088377</td>\n",
       "      <td>5.437860</td>\n",
       "      <td>0.633384</td>\n",
       "      <td>3.517708</td>\n",
       "      <td>2.451669</td>\n",
       "      <td>0.633384</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.827997</td>\n",
       "      <td>2.088377</td>\n",
       "      <td>2.525063</td>\n",
       "      <td>1.018107</td>\n",
       "      <td>1.690073</td>\n",
       "      <td>3.377151</td>\n",
       "      <td>2.372459</td>\n",
       "      <td>1.295254</td>\n",
       "      <td>1.972389</td>\n",
       "      <td>0.633384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A08</th>\n",
       "      <td>2.928915</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.287843</td>\n",
       "      <td>1.691356</td>\n",
       "      <td>5.560565</td>\n",
       "      <td>2.287843</td>\n",
       "      <td>3.709698</td>\n",
       "      <td>3.762476</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.040741</td>\n",
       "      <td>1.167335</td>\n",
       "      <td>2.802991</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.465355</td>\n",
       "      <td>2.490481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.691356</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A12</th>\n",
       "      <td>2.225148</td>\n",
       "      <td>1.634604</td>\n",
       "      <td>2.426444</td>\n",
       "      <td>2.426444</td>\n",
       "      <td>5.683740</td>\n",
       "      <td>1.119627</td>\n",
       "      <td>3.585825</td>\n",
       "      <td>3.326052</td>\n",
       "      <td>1.634604</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.972849</td>\n",
       "      <td>2.426444</td>\n",
       "      <td>2.593931</td>\n",
       "      <td>1.119627</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.235591</td>\n",
       "      <td>1.972849</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.634604</td>\n",
       "      <td>1.634604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_B10</th>\n",
       "      <td>2.987230</td>\n",
       "      <td>2.205004</td>\n",
       "      <td>1.853198</td>\n",
       "      <td>2.670728</td>\n",
       "      <td>5.436592</td>\n",
       "      <td>2.205004</td>\n",
       "      <td>3.504991</td>\n",
       "      <td>3.114438</td>\n",
       "      <td>1.853198</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.227275</td>\n",
       "      <td>1.853198</td>\n",
       "      <td>3.582720</td>\n",
       "      <td>1.305652</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.223213</td>\n",
       "      <td>1.853198</td>\n",
       "      <td>2.205004</td>\n",
       "      <td>1.853198</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_C11</th>\n",
       "      <td>2.751107</td>\n",
       "      <td>0.571870</td>\n",
       "      <td>2.591159</td>\n",
       "      <td>1.970270</td>\n",
       "      <td>5.644123</td>\n",
       "      <td>1.856460</td>\n",
       "      <td>3.638525</td>\n",
       "      <td>3.083403</td>\n",
       "      <td>1.580603</td>\n",
       "      <td>1.407641</td>\n",
       "      <td>...</td>\n",
       "      <td>1.407641</td>\n",
       "      <td>1.407641</td>\n",
       "      <td>3.436934</td>\n",
       "      <td>1.407641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.461447</td>\n",
       "      <td>1.580603</td>\n",
       "      <td>1.580603</td>\n",
       "      <td>1.407641</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_D09</th>\n",
       "      <td>2.464640</td>\n",
       "      <td>1.027191</td>\n",
       "      <td>3.449496</td>\n",
       "      <td>2.299183</td>\n",
       "      <td>5.493403</td>\n",
       "      <td>3.114337</td>\n",
       "      <td>3.700081</td>\n",
       "      <td>3.607240</td>\n",
       "      <td>1.523099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.100815</td>\n",
       "      <td>1.523099</td>\n",
       "      <td>2.464640</td>\n",
       "      <td>1.853109</td>\n",
       "      <td>1.027191</td>\n",
       "      <td>3.031354</td>\n",
       "      <td>2.841349</td>\n",
       "      <td>1.853109</td>\n",
       "      <td>2.464640</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_F10</th>\n",
       "      <td>2.149178</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.824024</td>\n",
       "      <td>2.065458</td>\n",
       "      <td>5.619921</td>\n",
       "      <td>1.974084</td>\n",
       "      <td>3.443976</td>\n",
       "      <td>3.507894</td>\n",
       "      <td>0.524072</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.065458</td>\n",
       "      <td>1.491657</td>\n",
       "      <td>2.864107</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.635758</td>\n",
       "      <td>4.014984</td>\n",
       "      <td>1.873513</td>\n",
       "      <td>1.323239</td>\n",
       "      <td>1.761685</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_H01</th>\n",
       "      <td>2.556990</td>\n",
       "      <td>0.624559</td>\n",
       "      <td>1.892109</td>\n",
       "      <td>2.524847</td>\n",
       "      <td>5.518234</td>\n",
       "      <td>1.727383</td>\n",
       "      <td>3.856328</td>\n",
       "      <td>3.475958</td>\n",
       "      <td>1.315365</td>\n",
       "      <td>0.233445</td>\n",
       "      <td>...</td>\n",
       "      <td>2.290025</td>\n",
       "      <td>1.509294</td>\n",
       "      <td>1.788409</td>\n",
       "      <td>0.759774</td>\n",
       "      <td>1.367123</td>\n",
       "      <td>4.081071</td>\n",
       "      <td>2.330552</td>\n",
       "      <td>1.482836</td>\n",
       "      <td>1.971055</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_H10</th>\n",
       "      <td>2.260768</td>\n",
       "      <td>1.267416</td>\n",
       "      <td>2.869512</td>\n",
       "      <td>2.697917</td>\n",
       "      <td>5.432869</td>\n",
       "      <td>2.418915</td>\n",
       "      <td>3.799819</td>\n",
       "      <td>1.578473</td>\n",
       "      <td>2.039231</td>\n",
       "      <td>0.172310</td>\n",
       "      <td>...</td>\n",
       "      <td>3.005765</td>\n",
       "      <td>1.842303</td>\n",
       "      <td>2.550780</td>\n",
       "      <td>0.369957</td>\n",
       "      <td>2.041536</td>\n",
       "      <td>4.472290</td>\n",
       "      <td>1.878427</td>\n",
       "      <td>1.694529</td>\n",
       "      <td>1.807208</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_A04</th>\n",
       "      <td>2.370404</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.967492</td>\n",
       "      <td>2.469283</td>\n",
       "      <td>5.697934</td>\n",
       "      <td>2.166221</td>\n",
       "      <td>3.824545</td>\n",
       "      <td>0.732249</td>\n",
       "      <td>1.467717</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.334129</td>\n",
       "      <td>1.469240</td>\n",
       "      <td>2.954350</td>\n",
       "      <td>0.557643</td>\n",
       "      <td>0.287043</td>\n",
       "      <td>3.197900</td>\n",
       "      <td>1.540386</td>\n",
       "      <td>1.422612</td>\n",
       "      <td>1.746686</td>\n",
       "      <td>0.287043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_B11</th>\n",
       "      <td>2.246812</td>\n",
       "      <td>0.662547</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.714423</td>\n",
       "      <td>5.558466</td>\n",
       "      <td>2.246812</td>\n",
       "      <td>4.033247</td>\n",
       "      <td>3.964330</td>\n",
       "      <td>1.340035</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.892865</td>\n",
       "      <td>1.560018</td>\n",
       "      <td>2.650151</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.340035</td>\n",
       "      <td>3.723480</td>\n",
       "      <td>2.650151</td>\n",
       "      <td>1.057600</td>\n",
       "      <td>2.142159</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_G01</th>\n",
       "      <td>1.948509</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.775430</td>\n",
       "      <td>2.623850</td>\n",
       "      <td>5.629741</td>\n",
       "      <td>1.707229</td>\n",
       "      <td>3.436920</td>\n",
       "      <td>2.986031</td>\n",
       "      <td>1.180754</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.445115</td>\n",
       "      <td>1.707229</td>\n",
       "      <td>1.948509</td>\n",
       "      <td>1.560538</td>\n",
       "      <td>1.560538</td>\n",
       "      <td>3.613867</td>\n",
       "      <td>2.050333</td>\n",
       "      <td>1.180754</td>\n",
       "      <td>1.948509</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_G09</th>\n",
       "      <td>2.707259</td>\n",
       "      <td>1.632013</td>\n",
       "      <td>2.128885</td>\n",
       "      <td>1.243721</td>\n",
       "      <td>5.444118</td>\n",
       "      <td>1.243721</td>\n",
       "      <td>3.869145</td>\n",
       "      <td>2.991982</td>\n",
       "      <td>1.243721</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.760704</td>\n",
       "      <td>1.910997</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.600391</td>\n",
       "      <td>4.111967</td>\n",
       "      <td>1.632013</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.456596</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_H04</th>\n",
       "      <td>2.718706</td>\n",
       "      <td>1.758107</td>\n",
       "      <td>1.969373</td>\n",
       "      <td>2.339709</td>\n",
       "      <td>5.705107</td>\n",
       "      <td>2.115431</td>\n",
       "      <td>3.431467</td>\n",
       "      <td>1.571595</td>\n",
       "      <td>1.329488</td>\n",
       "      <td>0.320538</td>\n",
       "      <td>...</td>\n",
       "      <td>2.635186</td>\n",
       "      <td>1.819228</td>\n",
       "      <td>1.717655</td>\n",
       "      <td>0.680518</td>\n",
       "      <td>1.845484</td>\n",
       "      <td>3.823444</td>\n",
       "      <td>1.816530</td>\n",
       "      <td>1.227406</td>\n",
       "      <td>1.618448</td>\n",
       "      <td>0.320538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_H05</th>\n",
       "      <td>2.027774</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000840</td>\n",
       "      <td>1.945744</td>\n",
       "      <td>5.533689</td>\n",
       "      <td>1.454262</td>\n",
       "      <td>3.602494</td>\n",
       "      <td>0.553023</td>\n",
       "      <td>1.395205</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.547333</td>\n",
       "      <td>1.342413</td>\n",
       "      <td>2.070768</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.553023</td>\n",
       "      <td>3.902930</td>\n",
       "      <td>1.962229</td>\n",
       "      <td>1.742703</td>\n",
       "      <td>1.536204</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows × 80599 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID    Interval_77062  Interval_18293  Interval_51938  Interval_33832  \\\n",
       "Clone_ID                                                                   \n",
       "P1_A03          2.397638        0.396998        2.529633        2.150198   \n",
       "P1_A05          2.717578        2.192299        1.972389        2.088377   \n",
       "P1_A08          2.928915        0.000000        2.287843        1.691356   \n",
       "P1_A12          2.225148        1.634604        2.426444        2.426444   \n",
       "P1_B10          2.987230        2.205004        1.853198        2.670728   \n",
       "P1_C11          2.751107        0.571870        2.591159        1.970270   \n",
       "P1_D09          2.464640        1.027191        3.449496        2.299183   \n",
       "P1_F10          2.149178        0.000000        2.824024        2.065458   \n",
       "P1_H01          2.556990        0.624559        1.892109        2.524847   \n",
       "P1_H10          2.260768        1.267416        2.869512        2.697917   \n",
       "P2_A04          2.370404        0.000000        2.967492        2.469283   \n",
       "P2_B11          2.246812        0.662547        0.000000        2.714423   \n",
       "P2_G01          1.948509        0.000000        2.775430        2.623850   \n",
       "P2_G09          2.707259        1.632013        2.128885        1.243721   \n",
       "P2_H04          2.718706        1.758107        1.969373        2.339709   \n",
       "P2_H05          2.027774        0.000000        2.000840        1.945744   \n",
       "\n",
       "PeakID    Interval_13433  Interval_48454  Interval_27763  Interval_53540  \\\n",
       "Clone_ID                                                                   \n",
       "P1_A03          5.617923        1.967855        3.278499        3.233557   \n",
       "P1_A05          5.437860        0.633384        3.517708        2.451669   \n",
       "P1_A08          5.560565        2.287843        3.709698        3.762476   \n",
       "P1_A12          5.683740        1.119627        3.585825        3.326052   \n",
       "P1_B10          5.436592        2.205004        3.504991        3.114438   \n",
       "P1_C11          5.644123        1.856460        3.638525        3.083403   \n",
       "P1_D09          5.493403        3.114337        3.700081        3.607240   \n",
       "P1_F10          5.619921        1.974084        3.443976        3.507894   \n",
       "P1_H01          5.518234        1.727383        3.856328        3.475958   \n",
       "P1_H10          5.432869        2.418915        3.799819        1.578473   \n",
       "P2_A04          5.697934        2.166221        3.824545        0.732249   \n",
       "P2_B11          5.558466        2.246812        4.033247        3.964330   \n",
       "P2_G01          5.629741        1.707229        3.436920        2.986031   \n",
       "P2_G09          5.444118        1.243721        3.869145        2.991982   \n",
       "P2_H04          5.705107        2.115431        3.431467        1.571595   \n",
       "P2_H05          5.533689        1.454262        3.602494        0.553023   \n",
       "\n",
       "PeakID    Interval_46169  Interval_46787  ...  Interval_58433  Interval_62113  \\\n",
       "Clone_ID                                  ...                                   \n",
       "P1_A03          0.939784        0.000000  ...        2.185111        2.035630   \n",
       "P1_A05          0.633384        0.000000  ...        2.827997        2.088377   \n",
       "P1_A08          0.000000        0.000000  ...        3.040741        1.167335   \n",
       "P1_A12          1.634604        0.000000  ...        1.972849        2.426444   \n",
       "P1_B10          1.853198        0.000000  ...        3.227275        1.853198   \n",
       "P1_C11          1.580603        1.407641  ...        1.407641        1.407641   \n",
       "P1_D09          1.523099        0.000000  ...        2.100815        1.523099   \n",
       "P1_F10          0.524072        0.000000  ...        2.065458        1.491657   \n",
       "P1_H01          1.315365        0.233445  ...        2.290025        1.509294   \n",
       "P1_H10          2.039231        0.172310  ...        3.005765        1.842303   \n",
       "P2_A04          1.467717        0.000000  ...        2.334129        1.469240   \n",
       "P2_B11          1.340035        0.000000  ...        1.892865        1.560018   \n",
       "P2_G01          1.180754        0.000000  ...        2.445115        1.707229   \n",
       "P2_G09          1.243721        0.000000  ...        2.760704        1.910997   \n",
       "P2_H04          1.329488        0.320538  ...        2.635186        1.819228   \n",
       "P2_H05          1.395205        0.000000  ...        2.547333        1.342413   \n",
       "\n",
       "PeakID    Interval_45676  Interval_43639  Interval_35777  Interval_8387  \\\n",
       "Clone_ID                                                                  \n",
       "P1_A03          3.081982        0.561223        0.659377       4.251013   \n",
       "P1_A05          2.525063        1.018107        1.690073       3.377151   \n",
       "P1_A08          2.802991        0.000000        0.000000       3.465355   \n",
       "P1_A12          2.593931        1.119627        0.000000       4.235591   \n",
       "P1_B10          3.582720        1.305652        0.000000       4.223213   \n",
       "P1_C11          3.436934        1.407641        0.000000       3.461447   \n",
       "P1_D09          2.464640        1.853109        1.027191       3.031354   \n",
       "P1_F10          2.864107        0.000000        1.635758       4.014984   \n",
       "P1_H01          1.788409        0.759774        1.367123       4.081071   \n",
       "P1_H10          2.550780        0.369957        2.041536       4.472290   \n",
       "P2_A04          2.954350        0.557643        0.287043       3.197900   \n",
       "P2_B11          2.650151        0.000000        1.340035       3.723480   \n",
       "P2_G01          1.948509        1.560538        1.560538       3.613867   \n",
       "P2_G09          0.000000        0.000000        0.600391       4.111967   \n",
       "P2_H04          1.717655        0.680518        1.845484       3.823444   \n",
       "P2_H05          2.070768        0.000000        0.553023       3.902930   \n",
       "\n",
       "PeakID    Interval_17712  Interval_31505  Interval_18056  Interval_10427  \n",
       "Clone_ID                                                                  \n",
       "P1_A03          1.816327        1.216354        1.990757        0.324315  \n",
       "P1_A05          2.372459        1.295254        1.972389        0.633384  \n",
       "P1_A08          2.490481        0.000000        1.691356        0.000000  \n",
       "P1_A12          1.972849        0.000000        1.634604        1.634604  \n",
       "P1_B10          1.853198        2.205004        1.853198        0.000000  \n",
       "P1_C11          1.580603        1.580603        1.407641        0.000000  \n",
       "P1_D09          2.841349        1.853109        2.464640        0.000000  \n",
       "P1_F10          1.873513        1.323239        1.761685        0.000000  \n",
       "P1_H01          2.330552        1.482836        1.971055        0.000000  \n",
       "P1_H10          1.878427        1.694529        1.807208        0.000000  \n",
       "P2_A04          1.540386        1.422612        1.746686        0.287043  \n",
       "P2_B11          2.650151        1.057600        2.142159        0.000000  \n",
       "P2_G01          2.050333        1.180754        1.948509        0.000000  \n",
       "P2_G09          1.632013        0.000000        1.456596        0.000000  \n",
       "P2_H04          1.816530        1.227406        1.618448        0.320538  \n",
       "P2_H05          1.962229        1.742703        1.536204        0.000000  \n",
       "\n",
       "[16 rows x 80599 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peak_clonal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following function creates the giant correlation matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pg_corr(chrome, method='pearson'):\n",
    "    pd_peak = peak_clonal.loc[:,ATAC_data.var.chr == chrome]\n",
    "    pd_gene = gene_clonal.loc[:,adata.var.chr == chrome]\n",
    "    \n",
    "    print('Making corr. matrix for chromosome {} with {} peaks and {} genes.'.format(\n",
    "        chrome, pd_peak.shape[1], pd_gene.shape[1]))\n",
    "\n",
    "    A = pd_peak\n",
    "    E = pd_gene\n",
    "    \n",
    "    if method == 'spearman':\n",
    "        A = A.rank()\n",
    "        E = E.rank()\n",
    "    Az = (A - A.mean())\n",
    "    Ez = (E - E.mean())\n",
    "    C = Az.T.dot(Ez).div(len(A)).div(Ez.std(ddof=0)).div(Az.std(ddof=0), axis=0)\n",
    "    return C.sort_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It will take just a few seconds to make the correlation matrices for every chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making corr. matrix for chromosome 1 with 7289 peaks and 1248 genes.\n",
      "Making corr. matrix for chromosome 2 with 6684 peaks and 802 genes.\n",
      "Making corr. matrix for chromosome 3 with 5425 peaks and 681 genes.\n",
      "Making corr. matrix for chromosome 4 with 4310 peaks and 429 genes.\n",
      "Making corr. matrix for chromosome 5 with 4407 peaks and 550 genes.\n",
      "Making corr. matrix for chromosome 6 with 5201 peaks and 648 genes.\n",
      "Making corr. matrix for chromosome 7 with 4297 peaks and 569 genes.\n",
      "Making corr. matrix for chromosome 8 with 3449 peaks and 380 genes.\n",
      "Making corr. matrix for chromosome 9 with 3257 peaks and 465 genes.\n",
      "Making corr. matrix for chromosome 10 with 3640 peaks and 460 genes.\n",
      "Making corr. matrix for chromosome 11 with 3888 peaks and 627 genes.\n",
      "Making corr. matrix for chromosome 12 with 3987 peaks and 666 genes.\n",
      "Making corr. matrix for chromosome 13 with 2159 peaks and 210 genes.\n",
      "Making corr. matrix for chromosome 14 with 2512 peaks and 388 genes.\n",
      "Making corr. matrix for chromosome 15 with 2421 peaks and 389 genes.\n",
      "Making corr. matrix for chromosome 16 with 2653 peaks and 535 genes.\n",
      "Making corr. matrix for chromosome 17 with 3544 peaks and 716 genes.\n",
      "Making corr. matrix for chromosome 18 with 1595 peaks and 166 genes.\n",
      "Making corr. matrix for chromosome 19 with 2944 peaks and 807 genes.\n",
      "Making corr. matrix for chromosome 20 with 1915 peaks and 307 genes.\n",
      "Making corr. matrix for chromosome 21 with 964 peaks and 123 genes.\n",
      "Making corr. matrix for chromosome 22 with 1550 peaks and 275 genes.\n",
      "Making corr. matrix for chromosome X with 2505 peaks and 402 genes.\n",
      "Making corr. matrix for chromosome Y with 3 peaks and 5 genes.\n"
     ]
    }
   ],
   "source": [
    "pg_corrs = {}\n",
    "for chrome in chromes:\n",
    "    pg_corrs[chrome] = pg_corr(chrome).fillna(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Looking at correlations within windows.\n",
    "\n",
    "We look at peaks within a window around each gene.  For a close-up look, we consider peaks within 50kb upstream of the TSS, together with the region between the gene start and end sites.  (We add a little 5kb wiggle-room too.)  For a broader look, we look at peaks within 500kb of the TSS.  The following function finds the peaks within such windows around a given gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def peaks_near(gene, window=50000, tol = 5000.0, method='TSSUp'):\n",
    "    gene_chrome = adata.var.chr[gene]\n",
    "    gene_TSS = adata.var.TSS[gene]\n",
    "    gene_start = adata.var.start[gene]\n",
    "    gene_end = adata.var.end[gene]\n",
    "    gene_dir = adata.var.dir[gene]\n",
    "\n",
    "    p_df = ATAC_good.var[ATAC_good.var.chr == gene_chrome]\n",
    "    dist_peak_TSS = gene_dir * (gene_TSS - p_df.mid)\n",
    "    if method == 'TSSCenter':\n",
    "        p_range = abs(dist_peak_TSS) < window\n",
    "    elif method == 'TSSUp':\n",
    "        up_range = (dist_peak_TSS < window) & (dist_peak_TSS >= -tol)\n",
    "        in_range = (p_df.mid > gene_start - tol) & (p_df.mid < gene_end + tol)\n",
    "        p_range = up_range | in_range\n",
    "    else:\n",
    "        print('Method unknown.')\n",
    "        return None\n",
    "    p_df = p_df[p_range]\n",
    "    return p_df.index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following function adds up the coefficients of determination for all peaks within a window around a gene. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def total_corr_near(gene, window=500000):\n",
    "    pn = peaks_near(gene, window, method='TSSCenter')\n",
    "    gene_chrome = adata.var.chr[gene]\n",
    "    R_sq_list = pg_corrs[gene_chrome][gene][pn]**2\n",
    "    loc_list = ATAC_good.var.loc[pn].mid - adata.var.TSS[gene]\n",
    "    tcn_df = pd.DataFrame(index=pn)\n",
    "    tcn_df['R_sq'] = R_sq_list\n",
    "    tcn_df['relative_loc'] = loc_list\n",
    "    tcn_df['loc_bin'] = pd.cut(loc_list, bins=np.arange(-500000,500001,5000))\n",
    "    return tcn_df.groupby('loc_bin')['R_sq'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "loc_bin\n",
       "(-500000, -495000]    0.041686\n",
       "(-495000, -490000]    0.000000\n",
       "(-490000, -485000]    0.000000\n",
       "(-485000, -480000]    0.000000\n",
       "(-480000, -475000]    0.000000\n",
       "                        ...   \n",
       "(475000, 480000]      0.000000\n",
       "(480000, 485000]      0.000000\n",
       "(485000, 490000]      0.000000\n",
       "(490000, 495000]      0.000000\n",
       "(495000, 500000]      0.000000\n",
       "Name: R_sq, Length: 200, dtype: float32"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcn = total_corr_near('GNLY')\n",
    "tcn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We add up the peak-correlations within a 500000bp window around every gene.  This will take a minute or two."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 genes complete\n",
      "1000 genes complete\n",
      "2000 genes complete\n",
      "3000 genes complete\n",
      "4000 genes complete\n",
      "5000 genes complete\n",
      "6000 genes complete\n",
      "7000 genes complete\n",
      "8000 genes complete\n",
      "9000 genes complete\n",
      "10000 genes complete\n",
      "11000 genes complete\n"
     ]
    }
   ],
   "source": [
    "TSS_intervals = tcn.index\n",
    "genes_chrome = [gene for gene in genes if adata.var.chr[gene] in chromes]\n",
    "\n",
    "TSS_plot = pd.DataFrame(index = TSS_intervals, columns=genes_chrome)\n",
    "TSS_plot\n",
    "\n",
    "for j,gene in enumerate(genes_chrome):\n",
    "    if j%1000 == 0:\n",
    "        print('{} genes complete'.format(j))\n",
    "    TSS_plot[gene] = total_corr_near(gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Panel:  Total R^2 for peak~gene within 500kb of TSS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Total R^2</th>\n",
       "      <th>x</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>loc_bin</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(-500000, -495000]</th>\n",
       "      <td>132.924057</td>\n",
       "      <td>-497500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(-495000, -490000]</th>\n",
       "      <td>133.605408</td>\n",
       "      <td>-492500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(-490000, -485000]</th>\n",
       "      <td>137.357162</td>\n",
       "      <td>-487500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(-485000, -480000]</th>\n",
       "      <td>137.347290</td>\n",
       "      <td>-482500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(-480000, -475000]</th>\n",
       "      <td>134.227310</td>\n",
       "      <td>-477500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Total R^2       x\n",
       "loc_bin                               \n",
       "(-500000, -495000]  132.924057 -497500\n",
       "(-495000, -490000]  133.605408 -492500\n",
       "(-490000, -485000]  137.357162 -487500\n",
       "(-485000, -480000]  137.347290 -482500\n",
       "(-480000, -475000]  134.227310 -477500"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plot_df = pd.DataFrame(TSS_plot.sum(axis=1), columns=['Total R^2'])\n",
    "plot_df['x'] = np.arange(-500000,500000, 5000) + 2500\n",
    "plot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-58-4dda7e95f2db>:22: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "\n",
    "sns.set_style('white')\n",
    "figsize=(60*mm, 40*mm)\n",
    "fig, ax = plt.subplots(figsize=figsize, dpi=300)\n",
    "setfontsize()\n",
    "\n",
    "sns.lineplot(x='x', y='Total R^2', data=plot_df,  color='k', linewidth=0.5, ax=ax)\n",
    "\n",
    "ax.set_ylim(0,375)\n",
    "ax.axvline(0,color='red', linestyle='--', ymax=0.6, linewidth=0.5)\n",
    "ax.set_xticks([-500000,-50000,50000,500000])\n",
    "ax.set_xticklabels(['-500kb','-50kb','+50kb','+500kb'], rotation=45, y=0.08)\n",
    "ax.set_yticks([100,200,300])\n",
    "ax.set_yticklabels([100,200,300], x=0.03)\n",
    "ax.text(0,15,'TSS',color='red', ha='center', bbox={'facecolor':'white',  'pad':0})\n",
    "ax.axvline(50000,color='steelblue',linestyle='--',linewidth=0.5)\n",
    "ax.axvline(-50000,color='steelblue',linestyle='--',linewidth=0.5)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('Total $R^2$')\n",
    "set_size(fig, figsize)\n",
    "plt.savefig(\"Panels/R2_near_TSS.pdf\", bbox_inches='tight')\n",
    "plt.suptitle('Gene $\\sim$ peak $R^2$ near all TSS', y=1.02)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Peak gene correlations, close to each gene\n",
    "\n",
    "We perform principal component regression to study the effect of peak heights (all peaks within a window) on gene expression.  Our window includes the whole start-end length of the gene, plus 50kb upstream of the TSS.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoters = ATAC_data.var[ATAC_data.var.itype == 'promoter-TSS'].groupby('in_ensg')['PeakID'].unique()\n",
    "num_promoters = ATAC_data.var[ATAC_data.var.itype == 'promoter-TSS'].groupby('in_ensg')['PeakID'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def peak_corrmatrix(peaklist):\n",
    "    pc_df = peak_clonal[peaklist]\n",
    "    return pc_df.corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PCR_analysis(gene, peaklist=None, window=50000, theta = 0.05, verbose=False):\n",
    "    if not peaklist:\n",
    "        peaklist = peaks_near(gene, window=window)\n",
    "    else:\n",
    "        peaklist = peaklist\n",
    "        \n",
    "    out_df = pd.DataFrame(np.eye(len(peaklist)), index = peaklist, columns = peaklist)\n",
    "    \n",
    "    \n",
    "    chrome = adata.var.chr[gene]\n",
    "    corr_with_gene = pg_corrs[chrome][gene][peaklist]\n",
    "    #display(corr_with_gene**2)\n",
    "    max_peaks = max(0,len(peak_clonal) - 1) # Cannot use too many peaks!\n",
    "    cwg = corr_with_gene.sort_values(ascending=False)[:max_peaks] \n",
    "    sel_peaks = cwg[cwg**2 > theta].index\n",
    "    \n",
    "    #print(sel_peaks)\n",
    "    out_df = out_df.loc[sel_peaks]\n",
    "    \n",
    "    # display(out_df)\n",
    "    \n",
    "    peak_corrs = peak_corrmatrix(sel_peaks)\n",
    "    \n",
    "    #display(peak_corrs)\n",
    "    #display(corr_with_gene**2)\n",
    "    C = peak_corrs\n",
    "    #C = C.multiply(corr_with_gene)\n",
    "    #C = C.multiply(corr_with_gene, axis=0)\n",
    "    eigens = np.linalg.eigh(C) #[0]\n",
    "    #display(C)\n",
    "    evals = eigens[0]\n",
    "    evecs = eigens[1]\n",
    "    #print(evecs)\n",
    "    realpeaks = peak_clonal[sel_peaks]\n",
    "    normpeaks = (realpeaks - realpeaks.mean()) / realpeaks.std()\n",
    "    #display(normpeaks)\n",
    "    eigenpeaks = pd.DataFrame(np.dot(normpeaks, evecs), index=normpeaks.index)\n",
    "    eig_cols = ['Eigenpeak {}'.format(str(j)) for j in range(len(sel_peaks))]\n",
    "    out_df[eig_cols] = evecs\n",
    "    eigenpeaks.columns = eig_cols\n",
    "    if verbose: \n",
    "        display(eigenpeaks)\n",
    "        display(eigenpeaks.corr())\n",
    "    out_df.loc['{}_R'.format(gene), peaklist] = corr_with_gene\n",
    "    out_df.loc['{}_Rsq'.format(gene), peaklist] = corr_with_gene**2\n",
    "    \n",
    "    gene_ensg = adata.var.loc[gene].ENSG\n",
    "    if gene_ensg in promoters.index:\n",
    "        prom_list = promoters[gene_ensg]\n",
    "    else:\n",
    "        prom_list = []\n",
    "        \n",
    "    out_df.loc[gene+'_promoter', peaklist] = peaklist.isin(prom_list)\n",
    "    #print(peaklist.isin(prom_list))\n",
    "    #display(out_df)\n",
    "    \n",
    "    out_df.loc['{}_R'.format(gene), eig_cols] = (eigenpeaks).corrwith(gene_clonal[gene])\n",
    "    out_df.loc['{}_Rsq'.format(gene), eig_cols] = out_df.loc['{}_R'.format(gene), eig_cols]**2\n",
    "    \n",
    "    #insig_eigs = (out_df.loc['{}_Rsq'.format(gene), eig_cols] < 2*theta)\n",
    "    #insig_eigs = insig_eigs[insig_eigs].index\n",
    "    #print(insig_eigs)\n",
    "    #out_df.drop(columns=insig_eigs, inplace=True)\n",
    "    #display(out_df)\n",
    "    return out_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We carry out a brief analysis of all (non-mito) genes here.  It will take a minute or two."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 genes completed\n",
      "100 genes complete\n",
      "500 genes completed\n",
      "1000 genes completed\n",
      "1500 genes completed\n",
      "2000 genes completed\n",
      "2500 genes completed\n",
      "3000 genes completed\n",
      "3500 genes completed\n",
      "4000 genes completed\n",
      "4500 genes completed\n",
      "5000 genes completed\n",
      "5500 genes completed\n",
      "6000 genes completed\n",
      "6500 genes completed\n",
      "7000 genes completed\n",
      "7500 genes completed\n",
      "8000 genes completed\n",
      "8500 genes completed\n",
      "9000 genes completed\n",
      "9500 genes completed\n",
      "10000 genes completed\n",
      "10500 genes completed\n",
      "11000 genes completed\n",
      "11500 genes completed\n"
     ]
    }
   ],
   "source": [
    "PCR_ans = {}\n",
    "for j,gene in enumerate(adata.var_names.unique()):\n",
    "    if j == 100:\n",
    "        print('100 genes complete')\n",
    "    if j%500 == 0:\n",
    "        print('{} genes completed'.format(j))\n",
    "    if adata.var.chr[gene] != 'MT': # No Mito genes in ATAC data.\n",
    "        PCR_ans[gene] = PCR_analysis(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_38954</th>\n",
       "      <th>Interval_38953</th>\n",
       "      <th>Interval_38958</th>\n",
       "      <th>Interval_38960</th>\n",
       "      <th>Interval_38952</th>\n",
       "      <th>Interval_38956</th>\n",
       "      <th>Interval_38957</th>\n",
       "      <th>Interval_38955</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <th>Eigenpeak 4</th>\n",
       "      <th>Eigenpeak 5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_38960</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.649201</td>\n",
       "      <td>0.319882</td>\n",
       "      <td>0.436807</td>\n",
       "      <td>0.074543</td>\n",
       "      <td>-0.051089</td>\n",
       "      <td>-0.526542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_38958</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.037467</td>\n",
       "      <td>0.247217</td>\n",
       "      <td>-0.705323</td>\n",
       "      <td>0.480680</td>\n",
       "      <td>-0.238539</td>\n",
       "      <td>-0.389931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_38952</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.225503</td>\n",
       "      <td>-0.311749</td>\n",
       "      <td>-0.413907</td>\n",
       "      <td>-0.742980</td>\n",
       "      <td>-0.051701</td>\n",
       "      <td>-0.354893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_38957</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.330066</td>\n",
       "      <td>-0.434762</td>\n",
       "      <td>0.335812</td>\n",
       "      <td>0.092549</td>\n",
       "      <td>-0.695269</td>\n",
       "      <td>-0.311937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_38956</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.037681</td>\n",
       "      <td>-0.634704</td>\n",
       "      <td>0.018647</td>\n",
       "      <td>0.382499</td>\n",
       "      <td>0.585323</td>\n",
       "      <td>-0.326306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_38954</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.644918</td>\n",
       "      <td>-0.384063</td>\n",
       "      <td>-0.165180</td>\n",
       "      <td>0.237697</td>\n",
       "      <td>-0.334387</td>\n",
       "      <td>0.490895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNLY_R</th>\n",
       "      <td>-0.478716</td>\n",
       "      <td>0.099327</td>\n",
       "      <td>0.517197</td>\n",
       "      <td>0.793927</td>\n",
       "      <td>0.50086</td>\n",
       "      <td>0.276469</td>\n",
       "      <td>0.459019</td>\n",
       "      <td>0.088697</td>\n",
       "      <td>-0.348087</td>\n",
       "      <td>0.050861</td>\n",
       "      <td>0.017245</td>\n",
       "      <td>-0.031298</td>\n",
       "      <td>-0.174513</td>\n",
       "      <td>-0.771071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNLY_Rsq</th>\n",
       "      <td>0.229169</td>\n",
       "      <td>0.009866</td>\n",
       "      <td>0.267493</td>\n",
       "      <td>0.63032</td>\n",
       "      <td>0.250861</td>\n",
       "      <td>0.076435</td>\n",
       "      <td>0.210699</td>\n",
       "      <td>0.007867</td>\n",
       "      <td>0.121164</td>\n",
       "      <td>0.002587</td>\n",
       "      <td>0.000297</td>\n",
       "      <td>0.000980</td>\n",
       "      <td>0.030455</td>\n",
       "      <td>0.594551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNLY_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID         Interval_38954 Interval_38953 Interval_38958 Interval_38960  \\\n",
       "PeakID                                                                       \n",
       "Interval_38960            0.0            0.0            0.0            1.0   \n",
       "Interval_38958            0.0            0.0            1.0            0.0   \n",
       "Interval_38952            0.0            0.0            0.0            0.0   \n",
       "Interval_38957            0.0            0.0            0.0            0.0   \n",
       "Interval_38956            0.0            0.0            0.0            0.0   \n",
       "Interval_38954            1.0            0.0            0.0            0.0   \n",
       "GNLY_R              -0.478716       0.099327       0.517197       0.793927   \n",
       "GNLY_Rsq             0.229169       0.009866       0.267493        0.63032   \n",
       "GNLY_promoter           False          False          False           True   \n",
       "\n",
       "PeakID         Interval_38952 Interval_38956 Interval_38957 Interval_38955  \\\n",
       "PeakID                                                                       \n",
       "Interval_38960            0.0            0.0            0.0            0.0   \n",
       "Interval_38958            0.0            0.0            0.0            0.0   \n",
       "Interval_38952            1.0            0.0            0.0            0.0   \n",
       "Interval_38957            0.0            0.0            1.0            0.0   \n",
       "Interval_38956            0.0            1.0            0.0            0.0   \n",
       "Interval_38954            0.0            0.0            0.0            0.0   \n",
       "GNLY_R                0.50086       0.276469       0.459019       0.088697   \n",
       "GNLY_Rsq             0.250861       0.076435       0.210699       0.007867   \n",
       "GNLY_promoter           False          False          False          False   \n",
       "\n",
       "PeakID          Eigenpeak 0  Eigenpeak 1  Eigenpeak 2  Eigenpeak 3  \\\n",
       "PeakID                                                               \n",
       "Interval_38960    -0.649201     0.319882     0.436807     0.074543   \n",
       "Interval_38958     0.037467     0.247217    -0.705323     0.480680   \n",
       "Interval_38952    -0.225503    -0.311749    -0.413907    -0.742980   \n",
       "Interval_38957     0.330066    -0.434762     0.335812     0.092549   \n",
       "Interval_38956    -0.037681    -0.634704     0.018647     0.382499   \n",
       "Interval_38954    -0.644918    -0.384063    -0.165180     0.237697   \n",
       "GNLY_R            -0.348087     0.050861     0.017245    -0.031298   \n",
       "GNLY_Rsq           0.121164     0.002587     0.000297     0.000980   \n",
       "GNLY_promoter           NaN          NaN          NaN          NaN   \n",
       "\n",
       "PeakID          Eigenpeak 4  Eigenpeak 5  \n",
       "PeakID                                    \n",
       "Interval_38960    -0.051089    -0.526542  \n",
       "Interval_38958    -0.238539    -0.389931  \n",
       "Interval_38952    -0.051701    -0.354893  \n",
       "Interval_38957    -0.695269    -0.311937  \n",
       "Interval_38956     0.585323    -0.326306  \n",
       "Interval_38954    -0.334387     0.490895  \n",
       "GNLY_R            -0.174513    -0.771071  \n",
       "GNLY_Rsq           0.030455     0.594551  \n",
       "GNLY_promoter           NaN          NaN  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_ans['GNLY']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can hunt for *repressor* peaks by looking for very negative R values as below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_7914</th>\n",
       "      <th>Interval_7912</th>\n",
       "      <th>Interval_7915</th>\n",
       "      <th>Interval_7916</th>\n",
       "      <th>Interval_7910</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_7914</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.707107</td>\n",
       "      <td>-0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_7915</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.707107</td>\n",
       "      <td>0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTPLA_R</th>\n",
       "      <td>0.414435</td>\n",
       "      <td>0.167362</td>\n",
       "      <td>-0.824415</td>\n",
       "      <td>-0.107373</td>\n",
       "      <td>-0.04491</td>\n",
       "      <td>0.415283</td>\n",
       "      <td>-0.712245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTPLA_Rsq</th>\n",
       "      <td>0.171756</td>\n",
       "      <td>0.02801</td>\n",
       "      <td>0.67966</td>\n",
       "      <td>0.011529</td>\n",
       "      <td>0.002017</td>\n",
       "      <td>0.172460</td>\n",
       "      <td>0.507292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTPLA_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID         Interval_7914 Interval_7912 Interval_7915 Interval_7916  \\\n",
       "PeakID                                                                   \n",
       "Interval_7914            1.0           0.0           0.0           0.0   \n",
       "Interval_7915            0.0           0.0           1.0           0.0   \n",
       "PTPLA_R             0.414435      0.167362     -0.824415     -0.107373   \n",
       "PTPLA_Rsq           0.171756       0.02801       0.67966      0.011529   \n",
       "PTPLA_promoter         False         False         False         False   \n",
       "\n",
       "PeakID         Interval_7910  Eigenpeak 0  Eigenpeak 1  \n",
       "PeakID                                                  \n",
       "Interval_7914            0.0    -0.707107    -0.707107  \n",
       "Interval_7915            0.0    -0.707107     0.707107  \n",
       "PTPLA_R             -0.04491     0.415283    -0.712245  \n",
       "PTPLA_Rsq           0.002017     0.172460     0.507292  \n",
       "PTPLA_promoter         False          NaN          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_12707</th>\n",
       "      <th>Interval_12708</th>\n",
       "      <th>Interval_12706</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_12706</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.707107</td>\n",
       "      <td>0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_12707</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.707107</td>\n",
       "      <td>0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLA2_R</th>\n",
       "      <td>-0.832008</td>\n",
       "      <td>-0.010158</td>\n",
       "      <td>-0.30152</td>\n",
       "      <td>-0.417184</td>\n",
       "      <td>-0.734286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLA2_Rsq</th>\n",
       "      <td>0.692237</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.090914</td>\n",
       "      <td>0.174043</td>\n",
       "      <td>0.539176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLA2_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID         Interval_12707 Interval_12708 Interval_12706  Eigenpeak 0  \\\n",
       "PeakID                                                                     \n",
       "Interval_12706            0.0            0.0            1.0    -0.707107   \n",
       "Interval_12707            1.0            0.0            0.0     0.707107   \n",
       "POLA2_R             -0.832008      -0.010158       -0.30152    -0.417184   \n",
       "POLA2_Rsq            0.692237       0.000103       0.090914     0.174043   \n",
       "POLA2_promoter          False           True          False          NaN   \n",
       "\n",
       "PeakID          Eigenpeak 1  \n",
       "PeakID                       \n",
       "Interval_12706     0.707107  \n",
       "Interval_12707     0.707107  \n",
       "POLA2_R           -0.734286  \n",
       "POLA2_Rsq          0.539176  \n",
       "POLA2_promoter          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_28631</th>\n",
       "      <th>Interval_28630</th>\n",
       "      <th>Interval_28633</th>\n",
       "      <th>Interval_28632</th>\n",
       "      <th>Interval_28636</th>\n",
       "      <th>Interval_28634</th>\n",
       "      <th>Interval_28635</th>\n",
       "      <th>Interval_28629</th>\n",
       "      <th>Interval_28628</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <th>Eigenpeak 4</th>\n",
       "      <th>Eigenpeak 5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_28629</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.171204</td>\n",
       "      <td>0.458659</td>\n",
       "      <td>-0.265992</td>\n",
       "      <td>0.678584</td>\n",
       "      <td>0.397227</td>\n",
       "      <td>-0.267028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_28630</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.229430</td>\n",
       "      <td>0.288613</td>\n",
       "      <td>0.531537</td>\n",
       "      <td>-0.273252</td>\n",
       "      <td>0.694289</td>\n",
       "      <td>0.157573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_28634</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.335186</td>\n",
       "      <td>0.434567</td>\n",
       "      <td>0.398577</td>\n",
       "      <td>0.387441</td>\n",
       "      <td>-0.521862</td>\n",
       "      <td>0.342765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_28633</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.182545</td>\n",
       "      <td>-0.659626</td>\n",
       "      <td>0.057835</td>\n",
       "      <td>0.519632</td>\n",
       "      <td>0.277506</td>\n",
       "      <td>0.425676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_28635</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.384583</td>\n",
       "      <td>0.251296</td>\n",
       "      <td>-0.693749</td>\n",
       "      <td>-0.207122</td>\n",
       "      <td>0.103673</td>\n",
       "      <td>0.503994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_28628</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.790237</td>\n",
       "      <td>0.138675</td>\n",
       "      <td>0.056743</td>\n",
       "      <td>-0.042776</td>\n",
       "      <td>0.008720</td>\n",
       "      <td>0.592595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WDR81_R</th>\n",
       "      <td>0.211021</td>\n",
       "      <td>-0.281517</td>\n",
       "      <td>-0.39769</td>\n",
       "      <td>0.084281</td>\n",
       "      <td>-0.213297</td>\n",
       "      <td>-0.360108</td>\n",
       "      <td>-0.692189</td>\n",
       "      <td>0.355854</td>\n",
       "      <td>-0.829867</td>\n",
       "      <td>0.192515</td>\n",
       "      <td>-0.185262</td>\n",
       "      <td>0.031259</td>\n",
       "      <td>0.146408</td>\n",
       "      <td>-0.046731</td>\n",
       "      <td>-0.792307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WDR81_Rsq</th>\n",
       "      <td>0.04453</td>\n",
       "      <td>0.079252</td>\n",
       "      <td>0.158157</td>\n",
       "      <td>0.007103</td>\n",
       "      <td>0.045496</td>\n",
       "      <td>0.129677</td>\n",
       "      <td>0.479125</td>\n",
       "      <td>0.126632</td>\n",
       "      <td>0.688678</td>\n",
       "      <td>0.037062</td>\n",
       "      <td>0.034322</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>0.021435</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.627750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WDR81_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID         Interval_28631 Interval_28630 Interval_28633 Interval_28632  \\\n",
       "PeakID                                                                       \n",
       "Interval_28629            0.0            0.0            0.0            0.0   \n",
       "Interval_28630            0.0            1.0            0.0            0.0   \n",
       "Interval_28634            0.0            0.0            0.0            0.0   \n",
       "Interval_28633            0.0            0.0            1.0            0.0   \n",
       "Interval_28635            0.0            0.0            0.0            0.0   \n",
       "Interval_28628            0.0            0.0            0.0            0.0   \n",
       "WDR81_R              0.211021      -0.281517       -0.39769       0.084281   \n",
       "WDR81_Rsq             0.04453       0.079252       0.158157       0.007103   \n",
       "WDR81_promoter          False          False          False          False   \n",
       "\n",
       "PeakID         Interval_28636 Interval_28634 Interval_28635 Interval_28629  \\\n",
       "PeakID                                                                       \n",
       "Interval_28629            0.0            0.0            0.0            1.0   \n",
       "Interval_28630            0.0            0.0            0.0            0.0   \n",
       "Interval_28634            0.0            1.0            0.0            0.0   \n",
       "Interval_28633            0.0            0.0            0.0            0.0   \n",
       "Interval_28635            0.0            0.0            1.0            0.0   \n",
       "Interval_28628            0.0            0.0            0.0            0.0   \n",
       "WDR81_R             -0.213297      -0.360108      -0.692189       0.355854   \n",
       "WDR81_Rsq            0.045496       0.129677       0.479125       0.126632   \n",
       "WDR81_promoter          False          False           True          False   \n",
       "\n",
       "PeakID         Interval_28628  Eigenpeak 0  Eigenpeak 1  Eigenpeak 2  \\\n",
       "PeakID                                                                 \n",
       "Interval_28629            0.0    -0.171204     0.458659    -0.265992   \n",
       "Interval_28630            0.0     0.229430     0.288613     0.531537   \n",
       "Interval_28634            0.0     0.335186     0.434567     0.398577   \n",
       "Interval_28633            0.0     0.182545    -0.659626     0.057835   \n",
       "Interval_28635            0.0     0.384583     0.251296    -0.693749   \n",
       "Interval_28628            1.0    -0.790237     0.138675     0.056743   \n",
       "WDR81_R             -0.829867     0.192515    -0.185262     0.031259   \n",
       "WDR81_Rsq            0.688678     0.037062     0.034322     0.000977   \n",
       "WDR81_promoter          False          NaN          NaN          NaN   \n",
       "\n",
       "PeakID          Eigenpeak 3  Eigenpeak 4  Eigenpeak 5  \n",
       "PeakID                                                 \n",
       "Interval_28629     0.678584     0.397227    -0.267028  \n",
       "Interval_28630    -0.273252     0.694289     0.157573  \n",
       "Interval_28634     0.387441    -0.521862     0.342765  \n",
       "Interval_28633     0.519632     0.277506     0.425676  \n",
       "Interval_28635    -0.207122     0.103673     0.503994  \n",
       "Interval_28628    -0.042776     0.008720     0.592595  \n",
       "WDR81_R            0.146408    -0.046731    -0.792307  \n",
       "WDR81_Rsq          0.021435     0.002184     0.627750  \n",
       "WDR81_promoter          NaN          NaN          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_31680</th>\n",
       "      <th>Interval_31679</th>\n",
       "      <th>Interval_31675</th>\n",
       "      <th>Interval_31677</th>\n",
       "      <th>Interval_31678</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_31677</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNRC6C_R</th>\n",
       "      <td>0.164693</td>\n",
       "      <td>-0.091602</td>\n",
       "      <td>-0.197118</td>\n",
       "      <td>-0.82714</td>\n",
       "      <td>-0.136104</td>\n",
       "      <td>-0.82714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNRC6C_Rsq</th>\n",
       "      <td>0.027124</td>\n",
       "      <td>0.008391</td>\n",
       "      <td>0.038855</td>\n",
       "      <td>0.68416</td>\n",
       "      <td>0.018524</td>\n",
       "      <td>0.68416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNRC6C_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID          Interval_31680 Interval_31679 Interval_31675 Interval_31677  \\\n",
       "PeakID                                                                        \n",
       "Interval_31677             0.0            0.0            0.0            1.0   \n",
       "TNRC6C_R              0.164693      -0.091602      -0.197118       -0.82714   \n",
       "TNRC6C_Rsq            0.027124       0.008391       0.038855        0.68416   \n",
       "TNRC6C_promoter          False          False          False          False   \n",
       "\n",
       "PeakID          Interval_31678  Eigenpeak 0  \n",
       "PeakID                                       \n",
       "Interval_31677             0.0      1.00000  \n",
       "TNRC6C_R             -0.136104     -0.82714  \n",
       "TNRC6C_Rsq            0.018524      0.68416  \n",
       "TNRC6C_promoter           True          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_34572</th>\n",
       "      <th>Interval_34574</th>\n",
       "      <th>Interval_34569</th>\n",
       "      <th>Interval_34567</th>\n",
       "      <th>Interval_34566</th>\n",
       "      <th>Interval_34568</th>\n",
       "      <th>Interval_34577</th>\n",
       "      <th>Interval_34565</th>\n",
       "      <th>Interval_34570</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <th>Eigenpeak 4</th>\n",
       "      <th>Eigenpeak 5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_34567</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.036138</td>\n",
       "      <td>-0.237097</td>\n",
       "      <td>0.343555</td>\n",
       "      <td>-0.793847</td>\n",
       "      <td>-0.325860</td>\n",
       "      <td>-0.296769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_34569</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.442847</td>\n",
       "      <td>-0.086588</td>\n",
       "      <td>-0.538722</td>\n",
       "      <td>0.147835</td>\n",
       "      <td>-0.611262</td>\n",
       "      <td>-0.332673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_34572</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.193282</td>\n",
       "      <td>0.641365</td>\n",
       "      <td>0.423904</td>\n",
       "      <td>0.288386</td>\n",
       "      <td>-0.288039</td>\n",
       "      <td>-0.453283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_34565</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.654904</td>\n",
       "      <td>0.228392</td>\n",
       "      <td>0.041701</td>\n",
       "      <td>-0.170380</td>\n",
       "      <td>0.577882</td>\n",
       "      <td>-0.392711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_34574</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.576956</td>\n",
       "      <td>0.062438</td>\n",
       "      <td>-0.561124</td>\n",
       "      <td>-0.172969</td>\n",
       "      <td>0.289491</td>\n",
       "      <td>-0.484396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_34566</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.058694</td>\n",
       "      <td>-0.684746</td>\n",
       "      <td>0.308957</td>\n",
       "      <td>0.453694</td>\n",
       "      <td>0.139481</td>\n",
       "      <td>-0.454895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNASEH2A_R</th>\n",
       "      <td>-0.550169</td>\n",
       "      <td>-0.623425</td>\n",
       "      <td>-0.292098</td>\n",
       "      <td>-0.284012</td>\n",
       "      <td>-0.804227</td>\n",
       "      <td>-0.057409</td>\n",
       "      <td>0.155131</td>\n",
       "      <td>-0.56181</td>\n",
       "      <td>0.190978</td>\n",
       "      <td>0.069978</td>\n",
       "      <td>0.209717</td>\n",
       "      <td>-0.144635</td>\n",
       "      <td>-0.149954</td>\n",
       "      <td>-0.161315</td>\n",
       "      <td>0.777317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNASEH2A_Rsq</th>\n",
       "      <td>0.302686</td>\n",
       "      <td>0.388658</td>\n",
       "      <td>0.085321</td>\n",
       "      <td>0.080663</td>\n",
       "      <td>0.646781</td>\n",
       "      <td>0.003296</td>\n",
       "      <td>0.024066</td>\n",
       "      <td>0.315631</td>\n",
       "      <td>0.036473</td>\n",
       "      <td>0.004897</td>\n",
       "      <td>0.043981</td>\n",
       "      <td>0.020919</td>\n",
       "      <td>0.022486</td>\n",
       "      <td>0.026022</td>\n",
       "      <td>0.604222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNASEH2A_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID            Interval_34572 Interval_34574 Interval_34569 Interval_34567  \\\n",
       "PeakID                                                                          \n",
       "Interval_34567               0.0            0.0            0.0            1.0   \n",
       "Interval_34569               0.0            0.0            1.0            0.0   \n",
       "Interval_34572               1.0            0.0            0.0            0.0   \n",
       "Interval_34565               0.0            0.0            0.0            0.0   \n",
       "Interval_34574               0.0            1.0            0.0            0.0   \n",
       "Interval_34566               0.0            0.0            0.0            0.0   \n",
       "RNASEH2A_R             -0.550169      -0.623425      -0.292098      -0.284012   \n",
       "RNASEH2A_Rsq            0.302686       0.388658       0.085321       0.080663   \n",
       "RNASEH2A_promoter          False          False          False          False   \n",
       "\n",
       "PeakID            Interval_34566 Interval_34568 Interval_34577 Interval_34565  \\\n",
       "PeakID                                                                          \n",
       "Interval_34567               0.0            0.0            0.0            0.0   \n",
       "Interval_34569               0.0            0.0            0.0            0.0   \n",
       "Interval_34572               0.0            0.0            0.0            0.0   \n",
       "Interval_34565               0.0            0.0            0.0            1.0   \n",
       "Interval_34574               0.0            0.0            0.0            0.0   \n",
       "Interval_34566               1.0            0.0            0.0            0.0   \n",
       "RNASEH2A_R             -0.804227      -0.057409       0.155131       -0.56181   \n",
       "RNASEH2A_Rsq            0.646781       0.003296       0.024066       0.315631   \n",
       "RNASEH2A_promoter          False          False           True          False   \n",
       "\n",
       "PeakID            Interval_34570  Eigenpeak 0  Eigenpeak 1  Eigenpeak 2  \\\n",
       "PeakID                                                                    \n",
       "Interval_34567               0.0    -0.036138    -0.237097     0.343555   \n",
       "Interval_34569               0.0     0.442847    -0.086588    -0.538722   \n",
       "Interval_34572               0.0    -0.193282     0.641365     0.423904   \n",
       "Interval_34565               0.0     0.654904     0.228392     0.041701   \n",
       "Interval_34574               0.0    -0.576956     0.062438    -0.561124   \n",
       "Interval_34566               0.0    -0.058694    -0.684746     0.308957   \n",
       "RNASEH2A_R              0.190978     0.069978     0.209717    -0.144635   \n",
       "RNASEH2A_Rsq            0.036473     0.004897     0.043981     0.020919   \n",
       "RNASEH2A_promoter          False          NaN          NaN          NaN   \n",
       "\n",
       "PeakID             Eigenpeak 3  Eigenpeak 4  Eigenpeak 5  \n",
       "PeakID                                                    \n",
       "Interval_34567       -0.793847    -0.325860    -0.296769  \n",
       "Interval_34569        0.147835    -0.611262    -0.332673  \n",
       "Interval_34572        0.288386    -0.288039    -0.453283  \n",
       "Interval_34565       -0.170380     0.577882    -0.392711  \n",
       "Interval_34574       -0.172969     0.289491    -0.484396  \n",
       "Interval_34566        0.453694     0.139481    -0.454895  \n",
       "RNASEH2A_R           -0.149954    -0.161315     0.777317  \n",
       "RNASEH2A_Rsq          0.022486     0.026022     0.604222  \n",
       "RNASEH2A_promoter          NaN          NaN          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_34644</th>\n",
       "      <th>Interval_34647</th>\n",
       "      <th>Interval_34641</th>\n",
       "      <th>Interval_34648</th>\n",
       "      <th>Interval_34643</th>\n",
       "      <th>Interval_34640</th>\n",
       "      <th>Interval_34642</th>\n",
       "      <th>Interval_34646</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_34641</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.707107</td>\n",
       "      <td>0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_34642</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.707107</td>\n",
       "      <td>0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM4_R</th>\n",
       "      <td>-0.159354</td>\n",
       "      <td>-0.096856</td>\n",
       "      <td>-0.555371</td>\n",
       "      <td>0.007067</td>\n",
       "      <td>-0.175304</td>\n",
       "      <td>0.217909</td>\n",
       "      <td>-0.807004</td>\n",
       "      <td>0.063407</td>\n",
       "      <td>-0.248750</td>\n",
       "      <td>-0.789642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM4_Rsq</th>\n",
       "      <td>0.025394</td>\n",
       "      <td>0.009381</td>\n",
       "      <td>0.308436</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>0.030732</td>\n",
       "      <td>0.047484</td>\n",
       "      <td>0.651255</td>\n",
       "      <td>0.00402</td>\n",
       "      <td>0.061877</td>\n",
       "      <td>0.623534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM4_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID          Interval_34644 Interval_34647 Interval_34641 Interval_34648  \\\n",
       "PeakID                                                                        \n",
       "Interval_34641             0.0            0.0            1.0            0.0   \n",
       "Interval_34642             0.0            0.0            0.0            0.0   \n",
       "ZSWIM4_R             -0.159354      -0.096856      -0.555371       0.007067   \n",
       "ZSWIM4_Rsq            0.025394       0.009381       0.308436        0.00005   \n",
       "ZSWIM4_promoter          False          False          False          False   \n",
       "\n",
       "PeakID          Interval_34643 Interval_34640 Interval_34642 Interval_34646  \\\n",
       "PeakID                                                                        \n",
       "Interval_34641             0.0            0.0            0.0            0.0   \n",
       "Interval_34642             0.0            0.0            1.0            0.0   \n",
       "ZSWIM4_R             -0.175304       0.217909      -0.807004       0.063407   \n",
       "ZSWIM4_Rsq            0.030732       0.047484       0.651255        0.00402   \n",
       "ZSWIM4_promoter          False          False          False          False   \n",
       "\n",
       "PeakID           Eigenpeak 0  Eigenpeak 1  \n",
       "PeakID                                     \n",
       "Interval_34641     -0.707107     0.707107  \n",
       "Interval_34642      0.707107     0.707107  \n",
       "ZSWIM4_R           -0.248750    -0.789642  \n",
       "ZSWIM4_Rsq          0.061877     0.623534  \n",
       "ZSWIM4_promoter          NaN          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_35840</th>\n",
       "      <th>Interval_35820</th>\n",
       "      <th>Interval_35823</th>\n",
       "      <th>Interval_35822</th>\n",
       "      <th>Interval_35835</th>\n",
       "      <th>Interval_35827</th>\n",
       "      <th>Interval_35818</th>\n",
       "      <th>Interval_35817</th>\n",
       "      <th>Interval_35829</th>\n",
       "      <th>Interval_35821</th>\n",
       "      <th>...</th>\n",
       "      <th>Interval_35836</th>\n",
       "      <th>Interval_35830</th>\n",
       "      <th>Interval_35831</th>\n",
       "      <th>Interval_35834</th>\n",
       "      <th>Interval_35824</th>\n",
       "      <th>Interval_35828</th>\n",
       "      <th>Interval_35837</th>\n",
       "      <th>Interval_35839</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_35826</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.707107</td>\n",
       "      <td>0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_35832</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.707107</td>\n",
       "      <td>0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MARK4_R</th>\n",
       "      <td>0.062365</td>\n",
       "      <td>0.009838</td>\n",
       "      <td>-0.848172</td>\n",
       "      <td>-0.055361</td>\n",
       "      <td>-0.464867</td>\n",
       "      <td>-0.157907</td>\n",
       "      <td>-0.243045</td>\n",
       "      <td>0.001945</td>\n",
       "      <td>-0.326618</td>\n",
       "      <td>-0.174052</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.034096</td>\n",
       "      <td>-0.263781</td>\n",
       "      <td>-0.270783</td>\n",
       "      <td>-0.204832</td>\n",
       "      <td>-0.512891</td>\n",
       "      <td>-0.202013</td>\n",
       "      <td>-0.044096</td>\n",
       "      <td>-0.020382</td>\n",
       "      <td>-0.361588</td>\n",
       "      <td>0.008554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MARK4_Rsq</th>\n",
       "      <td>0.003889</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>0.719396</td>\n",
       "      <td>0.003065</td>\n",
       "      <td>0.216101</td>\n",
       "      <td>0.024935</td>\n",
       "      <td>0.059071</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.106679</td>\n",
       "      <td>0.030294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001163</td>\n",
       "      <td>0.069581</td>\n",
       "      <td>0.073323</td>\n",
       "      <td>0.041956</td>\n",
       "      <td>0.263057</td>\n",
       "      <td>0.040809</td>\n",
       "      <td>0.001944</td>\n",
       "      <td>0.000415</td>\n",
       "      <td>0.130746</td>\n",
       "      <td>0.000073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MARK4_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID         Interval_35840 Interval_35820 Interval_35823 Interval_35822  \\\n",
       "PeakID                                                                       \n",
       "Interval_35826            0.0            0.0            0.0            0.0   \n",
       "Interval_35832            0.0            0.0            0.0            0.0   \n",
       "MARK4_R              0.062365       0.009838      -0.848172      -0.055361   \n",
       "MARK4_Rsq            0.003889       0.000097       0.719396       0.003065   \n",
       "MARK4_promoter          False          False          False          False   \n",
       "\n",
       "PeakID         Interval_35835 Interval_35827 Interval_35818 Interval_35817  \\\n",
       "PeakID                                                                       \n",
       "Interval_35826            0.0            0.0            0.0            0.0   \n",
       "Interval_35832            0.0            0.0            0.0            0.0   \n",
       "MARK4_R             -0.464867      -0.157907      -0.243045       0.001945   \n",
       "MARK4_Rsq            0.216101       0.024935       0.059071       0.000004   \n",
       "MARK4_promoter          False          False          False          False   \n",
       "\n",
       "PeakID         Interval_35829 Interval_35821  ... Interval_35836  \\\n",
       "PeakID                                        ...                  \n",
       "Interval_35826            0.0            0.0  ...            0.0   \n",
       "Interval_35832            0.0            0.0  ...            0.0   \n",
       "MARK4_R             -0.326618      -0.174052  ...      -0.034096   \n",
       "MARK4_Rsq            0.106679       0.030294  ...       0.001163   \n",
       "MARK4_promoter          False          False  ...           True   \n",
       "\n",
       "PeakID         Interval_35830 Interval_35831 Interval_35834 Interval_35824  \\\n",
       "PeakID                                                                       \n",
       "Interval_35826            0.0            0.0            0.0            0.0   \n",
       "Interval_35832            0.0            0.0            0.0            0.0   \n",
       "MARK4_R             -0.263781      -0.270783      -0.204832      -0.512891   \n",
       "MARK4_Rsq            0.069581       0.073323       0.041956       0.263057   \n",
       "MARK4_promoter          False          False          False          False   \n",
       "\n",
       "PeakID         Interval_35828 Interval_35837 Interval_35839 Eigenpeak 0  \\\n",
       "PeakID                                                                    \n",
       "Interval_35826            0.0            0.0            0.0   -0.707107   \n",
       "Interval_35832            0.0            0.0            0.0    0.707107   \n",
       "MARK4_R             -0.202013      -0.044096      -0.020382   -0.361588   \n",
       "MARK4_Rsq            0.040809       0.001944       0.000415    0.130746   \n",
       "MARK4_promoter          False          False          False         NaN   \n",
       "\n",
       "PeakID         Eigenpeak 1  \n",
       "PeakID                      \n",
       "Interval_35826    0.707107  \n",
       "Interval_35832    0.707107  \n",
       "MARK4_R           0.008554  \n",
       "MARK4_Rsq         0.000073  \n",
       "MARK4_promoter         NaN  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_35870</th>\n",
       "      <th>Interval_35878</th>\n",
       "      <th>Interval_35872</th>\n",
       "      <th>Interval_35880</th>\n",
       "      <th>Interval_35871</th>\n",
       "      <th>Interval_35879</th>\n",
       "      <th>Interval_35873</th>\n",
       "      <th>Interval_35869</th>\n",
       "      <th>Interval_35874</th>\n",
       "      <th>Interval_35875</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <th>Eigenpeak 4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_35871</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.528200</td>\n",
       "      <td>0.167217</td>\n",
       "      <td>0.039205</td>\n",
       "      <td>0.752104</td>\n",
       "      <td>-0.354747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_35878</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.195303</td>\n",
       "      <td>-0.216130</td>\n",
       "      <td>0.903591</td>\n",
       "      <td>-0.009972</td>\n",
       "      <td>-0.313955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_35872</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.260821</td>\n",
       "      <td>-0.703140</td>\n",
       "      <td>-0.401249</td>\n",
       "      <td>0.118780</td>\n",
       "      <td>-0.512305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_35874</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.473638</td>\n",
       "      <td>0.656427</td>\n",
       "      <td>-0.142473</td>\n",
       "      <td>-0.072382</td>\n",
       "      <td>-0.565006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_35869</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.624900</td>\n",
       "      <td>-0.004832</td>\n",
       "      <td>-0.026195</td>\n",
       "      <td>-0.644121</td>\n",
       "      <td>-0.440338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EML2_R</th>\n",
       "      <td>0.195169</td>\n",
       "      <td>-0.225418</td>\n",
       "      <td>-0.339943</td>\n",
       "      <td>-0.117802</td>\n",
       "      <td>0.234278</td>\n",
       "      <td>0.090075</td>\n",
       "      <td>-0.017614</td>\n",
       "      <td>-0.825912</td>\n",
       "      <td>-0.379751</td>\n",
       "      <td>-0.100648</td>\n",
       "      <td>-0.181194</td>\n",
       "      <td>0.126359</td>\n",
       "      <td>0.018520</td>\n",
       "      <td>0.657724</td>\n",
       "      <td>0.482080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EML2_Rsq</th>\n",
       "      <td>0.038091</td>\n",
       "      <td>0.050813</td>\n",
       "      <td>0.115561</td>\n",
       "      <td>0.013877</td>\n",
       "      <td>0.054886</td>\n",
       "      <td>0.008114</td>\n",
       "      <td>0.00031</td>\n",
       "      <td>0.682131</td>\n",
       "      <td>0.144211</td>\n",
       "      <td>0.01013</td>\n",
       "      <td>0.032831</td>\n",
       "      <td>0.015967</td>\n",
       "      <td>0.000343</td>\n",
       "      <td>0.432601</td>\n",
       "      <td>0.232401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EML2_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID         Interval_35870 Interval_35878 Interval_35872 Interval_35880  \\\n",
       "PeakID                                                                       \n",
       "Interval_35871            0.0            0.0            0.0            0.0   \n",
       "Interval_35878            0.0            1.0            0.0            0.0   \n",
       "Interval_35872            0.0            0.0            1.0            0.0   \n",
       "Interval_35874            0.0            0.0            0.0            0.0   \n",
       "Interval_35869            0.0            0.0            0.0            0.0   \n",
       "EML2_R               0.195169      -0.225418      -0.339943      -0.117802   \n",
       "EML2_Rsq             0.038091       0.050813       0.115561       0.013877   \n",
       "EML2_promoter           False          False           True          False   \n",
       "\n",
       "PeakID         Interval_35871 Interval_35879 Interval_35873 Interval_35869  \\\n",
       "PeakID                                                                       \n",
       "Interval_35871            1.0            0.0            0.0            0.0   \n",
       "Interval_35878            0.0            0.0            0.0            0.0   \n",
       "Interval_35872            0.0            0.0            0.0            0.0   \n",
       "Interval_35874            0.0            0.0            0.0            0.0   \n",
       "Interval_35869            0.0            0.0            0.0            1.0   \n",
       "EML2_R               0.234278       0.090075      -0.017614      -0.825912   \n",
       "EML2_Rsq             0.054886       0.008114        0.00031       0.682131   \n",
       "EML2_promoter            True          False           True          False   \n",
       "\n",
       "PeakID         Interval_35874 Interval_35875  Eigenpeak 0  Eigenpeak 1  \\\n",
       "PeakID                                                                   \n",
       "Interval_35871            0.0            0.0     0.528200     0.167217   \n",
       "Interval_35878            0.0            0.0    -0.195303    -0.216130   \n",
       "Interval_35872            0.0            0.0    -0.260821    -0.703140   \n",
       "Interval_35874            1.0            0.0    -0.473638     0.656427   \n",
       "Interval_35869            0.0            0.0     0.624900    -0.004832   \n",
       "EML2_R              -0.379751      -0.100648    -0.181194     0.126359   \n",
       "EML2_Rsq             0.144211        0.01013     0.032831     0.015967   \n",
       "EML2_promoter           False           True          NaN          NaN   \n",
       "\n",
       "PeakID          Eigenpeak 2  Eigenpeak 3  Eigenpeak 4  \n",
       "PeakID                                                 \n",
       "Interval_35871     0.039205     0.752104    -0.354747  \n",
       "Interval_35878     0.903591    -0.009972    -0.313955  \n",
       "Interval_35872    -0.401249     0.118780    -0.512305  \n",
       "Interval_35874    -0.142473    -0.072382    -0.565006  \n",
       "Interval_35869    -0.026195    -0.644121    -0.440338  \n",
       "EML2_R             0.018520     0.657724     0.482080  \n",
       "EML2_Rsq           0.000343     0.432601     0.232401  \n",
       "EML2_promoter           NaN          NaN          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_46380</th>\n",
       "      <th>Interval_46377</th>\n",
       "      <th>Interval_46379</th>\n",
       "      <th>Interval_46378</th>\n",
       "      <th>Interval_46381</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_46381</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.020265</td>\n",
       "      <td>-0.930805</td>\n",
       "      <td>-0.364953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_46380</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.710271</td>\n",
       "      <td>0.243497</td>\n",
       "      <td>-0.660473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_46379</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.703636</td>\n",
       "      <td>0.272600</td>\n",
       "      <td>-0.656190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THAP7-AS1_R</th>\n",
       "      <td>-0.70981</td>\n",
       "      <td>0.04764</td>\n",
       "      <td>-0.833162</td>\n",
       "      <td>-0.132839</td>\n",
       "      <td>-0.290844</td>\n",
       "      <td>-0.134563</td>\n",
       "      <td>-0.136507</td>\n",
       "      <td>0.866363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THAP7-AS1_Rsq</th>\n",
       "      <td>0.50383</td>\n",
       "      <td>0.00227</td>\n",
       "      <td>0.694158</td>\n",
       "      <td>0.017646</td>\n",
       "      <td>0.08459</td>\n",
       "      <td>0.018107</td>\n",
       "      <td>0.018634</td>\n",
       "      <td>0.750585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THAP7-AS1_promoter</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID             Interval_46380 Interval_46377 Interval_46379  \\\n",
       "PeakID                                                            \n",
       "Interval_46381                0.0            0.0            0.0   \n",
       "Interval_46380                1.0            0.0            0.0   \n",
       "Interval_46379                0.0            0.0            1.0   \n",
       "THAP7-AS1_R              -0.70981        0.04764      -0.833162   \n",
       "THAP7-AS1_Rsq             0.50383        0.00227       0.694158   \n",
       "THAP7-AS1_promoter           True          False          False   \n",
       "\n",
       "PeakID             Interval_46378 Interval_46381  Eigenpeak 0  Eigenpeak 1  \\\n",
       "PeakID                                                                       \n",
       "Interval_46381                0.0            1.0     0.020265    -0.930805   \n",
       "Interval_46380                0.0            0.0    -0.710271     0.243497   \n",
       "Interval_46379                0.0            0.0     0.703636     0.272600   \n",
       "THAP7-AS1_R             -0.132839      -0.290844    -0.134563    -0.136507   \n",
       "THAP7-AS1_Rsq            0.017646        0.08459     0.018107     0.018634   \n",
       "THAP7-AS1_promoter          False          False          NaN          NaN   \n",
       "\n",
       "PeakID              Eigenpeak 2  \n",
       "PeakID                           \n",
       "Interval_46381        -0.364953  \n",
       "Interval_46380        -0.660473  \n",
       "Interval_46379        -0.656190  \n",
       "THAP7-AS1_R            0.866363  \n",
       "THAP7-AS1_Rsq          0.750585  \n",
       "THAP7-AS1_promoter          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_48815</th>\n",
       "      <th>Interval_48811</th>\n",
       "      <th>Interval_48810</th>\n",
       "      <th>Interval_48816</th>\n",
       "      <th>Interval_48814</th>\n",
       "      <th>Interval_48813</th>\n",
       "      <th>Interval_48809</th>\n",
       "      <th>Interval_48812</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_48812</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.735659</td>\n",
       "      <td>0.206526</td>\n",
       "      <td>-0.645100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_48810</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.659016</td>\n",
       "      <td>0.438331</td>\n",
       "      <td>-0.611198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_48814</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.156539</td>\n",
       "      <td>0.874764</td>\n",
       "      <td>0.458566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CX3CR1_R</th>\n",
       "      <td>-0.146545</td>\n",
       "      <td>-0.073746</td>\n",
       "      <td>0.272845</td>\n",
       "      <td>-0.13615</td>\n",
       "      <td>-0.84793</td>\n",
       "      <td>0.072962</td>\n",
       "      <td>-0.220777</td>\n",
       "      <td>0.502327</td>\n",
       "      <td>-0.082997</td>\n",
       "      <td>-0.571394</td>\n",
       "      <td>-0.673613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CX3CR1_Rsq</th>\n",
       "      <td>0.021475</td>\n",
       "      <td>0.005438</td>\n",
       "      <td>0.074444</td>\n",
       "      <td>0.018537</td>\n",
       "      <td>0.718986</td>\n",
       "      <td>0.005323</td>\n",
       "      <td>0.048742</td>\n",
       "      <td>0.252333</td>\n",
       "      <td>0.006889</td>\n",
       "      <td>0.326491</td>\n",
       "      <td>0.453755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CX3CR1_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID          Interval_48815 Interval_48811 Interval_48810 Interval_48816  \\\n",
       "PeakID                                                                        \n",
       "Interval_48812             0.0            0.0            0.0            0.0   \n",
       "Interval_48810             0.0            0.0            1.0            0.0   \n",
       "Interval_48814             0.0            0.0            0.0            0.0   \n",
       "CX3CR1_R             -0.146545      -0.073746       0.272845       -0.13615   \n",
       "CX3CR1_Rsq            0.021475       0.005438       0.074444       0.018537   \n",
       "CX3CR1_promoter          False          False          False          False   \n",
       "\n",
       "PeakID          Interval_48814 Interval_48813 Interval_48809 Interval_48812  \\\n",
       "PeakID                                                                        \n",
       "Interval_48812             0.0            0.0            0.0            1.0   \n",
       "Interval_48810             0.0            0.0            0.0            0.0   \n",
       "Interval_48814             1.0            0.0            0.0            0.0   \n",
       "CX3CR1_R              -0.84793       0.072962      -0.220777       0.502327   \n",
       "CX3CR1_Rsq            0.718986       0.005323       0.048742       0.252333   \n",
       "CX3CR1_promoter          False          False          False           True   \n",
       "\n",
       "PeakID           Eigenpeak 0  Eigenpeak 1  Eigenpeak 2  \n",
       "PeakID                                                  \n",
       "Interval_48812     -0.735659     0.206526    -0.645100  \n",
       "Interval_48810      0.659016     0.438331    -0.611198  \n",
       "Interval_48814     -0.156539     0.874764     0.458566  \n",
       "CX3CR1_R           -0.082997    -0.571394    -0.673613  \n",
       "CX3CR1_Rsq          0.006889     0.326491     0.453755  \n",
       "CX3CR1_promoter          NaN          NaN          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_49196</th>\n",
       "      <th>Interval_49194</th>\n",
       "      <th>Interval_49192</th>\n",
       "      <th>Interval_49197</th>\n",
       "      <th>Interval_49193</th>\n",
       "      <th>Interval_49191</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_49196</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.707107</td>\n",
       "      <td>0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_49197</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.707107</td>\n",
       "      <td>0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHX30_R</th>\n",
       "      <td>0.232945</td>\n",
       "      <td>-0.062689</td>\n",
       "      <td>0.147127</td>\n",
       "      <td>-0.804288</td>\n",
       "      <td>-0.060495</td>\n",
       "      <td>-0.163753</td>\n",
       "      <td>-0.804256</td>\n",
       "      <td>-0.373760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHX30_Rsq</th>\n",
       "      <td>0.054264</td>\n",
       "      <td>0.00393</td>\n",
       "      <td>0.021646</td>\n",
       "      <td>0.646879</td>\n",
       "      <td>0.00366</td>\n",
       "      <td>0.026815</td>\n",
       "      <td>0.646828</td>\n",
       "      <td>0.139697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHX30_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID         Interval_49196 Interval_49194 Interval_49192 Interval_49197  \\\n",
       "PeakID                                                                       \n",
       "Interval_49196            1.0            0.0            0.0            0.0   \n",
       "Interval_49197            0.0            0.0            0.0            1.0   \n",
       "DHX30_R              0.232945      -0.062689       0.147127      -0.804288   \n",
       "DHX30_Rsq            0.054264        0.00393       0.021646       0.646879   \n",
       "DHX30_promoter          False          False          False          False   \n",
       "\n",
       "PeakID         Interval_49193 Interval_49191  Eigenpeak 0  Eigenpeak 1  \n",
       "PeakID                                                                  \n",
       "Interval_49196            0.0            0.0    -0.707107     0.707107  \n",
       "Interval_49197            0.0            0.0     0.707107     0.707107  \n",
       "DHX30_R             -0.060495      -0.163753    -0.804256    -0.373760  \n",
       "DHX30_Rsq             0.00366       0.026815     0.646828     0.139697  \n",
       "DHX30_promoter           True          False          NaN          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_50636</th>\n",
       "      <th>Interval_50646</th>\n",
       "      <th>Interval_50642</th>\n",
       "      <th>Interval_50641</th>\n",
       "      <th>Interval_50644</th>\n",
       "      <th>Interval_50639</th>\n",
       "      <th>Interval_50638</th>\n",
       "      <th>Interval_50649</th>\n",
       "      <th>Interval_50643</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <th>Eigenpeak 4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_50646</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.064360</td>\n",
       "      <td>0.766872</td>\n",
       "      <td>-0.244797</td>\n",
       "      <td>0.339253</td>\n",
       "      <td>-0.482439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_50644</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.609542</td>\n",
       "      <td>-0.165451</td>\n",
       "      <td>-0.307274</td>\n",
       "      <td>-0.569877</td>\n",
       "      <td>-0.426505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_50636</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.039170</td>\n",
       "      <td>-0.077353</td>\n",
       "      <td>-0.890044</td>\n",
       "      <td>0.143093</td>\n",
       "      <td>0.424062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_50639</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.229679</td>\n",
       "      <td>0.567817</td>\n",
       "      <td>0.023851</td>\n",
       "      <td>-0.699016</td>\n",
       "      <td>0.368292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_50643</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.755003</td>\n",
       "      <td>0.236926</td>\n",
       "      <td>0.230022</td>\n",
       "      <td>0.225940</td>\n",
       "      <td>0.519498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHLDB2_R</th>\n",
       "      <td>-0.467084</td>\n",
       "      <td>0.623617</td>\n",
       "      <td>-0.035791</td>\n",
       "      <td>0.200649</td>\n",
       "      <td>0.560877</td>\n",
       "      <td>-0.484676</td>\n",
       "      <td>-0.047488</td>\n",
       "      <td>0.085494</td>\n",
       "      <td>-0.822058</td>\n",
       "      <td>-0.214463</td>\n",
       "      <td>-0.083320</td>\n",
       "      <td>-0.141645</td>\n",
       "      <td>-0.021583</td>\n",
       "      <td>-0.806268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHLDB2_Rsq</th>\n",
       "      <td>0.218167</td>\n",
       "      <td>0.388898</td>\n",
       "      <td>0.001281</td>\n",
       "      <td>0.04026</td>\n",
       "      <td>0.314583</td>\n",
       "      <td>0.234911</td>\n",
       "      <td>0.002255</td>\n",
       "      <td>0.007309</td>\n",
       "      <td>0.67578</td>\n",
       "      <td>0.045994</td>\n",
       "      <td>0.006942</td>\n",
       "      <td>0.020063</td>\n",
       "      <td>0.000466</td>\n",
       "      <td>0.650068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHLDB2_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID          Interval_50636 Interval_50646 Interval_50642 Interval_50641  \\\n",
       "PeakID                                                                        \n",
       "Interval_50646             0.0            1.0            0.0            0.0   \n",
       "Interval_50644             0.0            0.0            0.0            0.0   \n",
       "Interval_50636             1.0            0.0            0.0            0.0   \n",
       "Interval_50639             0.0            0.0            0.0            0.0   \n",
       "Interval_50643             0.0            0.0            0.0            0.0   \n",
       "PHLDB2_R             -0.467084       0.623617      -0.035791       0.200649   \n",
       "PHLDB2_Rsq            0.218167       0.388898       0.001281        0.04026   \n",
       "PHLDB2_promoter          False          False          False          False   \n",
       "\n",
       "PeakID          Interval_50644 Interval_50639 Interval_50638 Interval_50649  \\\n",
       "PeakID                                                                        \n",
       "Interval_50646             0.0            0.0            0.0            0.0   \n",
       "Interval_50644             1.0            0.0            0.0            0.0   \n",
       "Interval_50636             0.0            0.0            0.0            0.0   \n",
       "Interval_50639             0.0            1.0            0.0            0.0   \n",
       "Interval_50643             0.0            0.0            0.0            0.0   \n",
       "PHLDB2_R              0.560877      -0.484676      -0.047488       0.085494   \n",
       "PHLDB2_Rsq            0.314583       0.234911       0.002255       0.007309   \n",
       "PHLDB2_promoter          False          False          False          False   \n",
       "\n",
       "PeakID          Interval_50643  Eigenpeak 0  Eigenpeak 1  Eigenpeak 2  \\\n",
       "PeakID                                                                  \n",
       "Interval_50646             0.0     0.064360     0.766872    -0.244797   \n",
       "Interval_50644             0.0     0.609542    -0.165451    -0.307274   \n",
       "Interval_50636             0.0    -0.039170    -0.077353    -0.890044   \n",
       "Interval_50639             0.0    -0.229679     0.567817     0.023851   \n",
       "Interval_50643             1.0     0.755003     0.236926     0.230022   \n",
       "PHLDB2_R             -0.822058    -0.214463    -0.083320    -0.141645   \n",
       "PHLDB2_Rsq             0.67578     0.045994     0.006942     0.020063   \n",
       "PHLDB2_promoter          False          NaN          NaN          NaN   \n",
       "\n",
       "PeakID           Eigenpeak 3  Eigenpeak 4  \n",
       "PeakID                                     \n",
       "Interval_50646      0.339253    -0.482439  \n",
       "Interval_50644     -0.569877    -0.426505  \n",
       "Interval_50636      0.143093     0.424062  \n",
       "Interval_50639     -0.699016     0.368292  \n",
       "Interval_50643      0.225940     0.519498  \n",
       "PHLDB2_R           -0.021583    -0.806268  \n",
       "PHLDB2_Rsq          0.000466     0.650068  \n",
       "PHLDB2_promoter          NaN          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_62659</th>\n",
       "      <th>Interval_62655</th>\n",
       "      <th>Interval_62672</th>\n",
       "      <th>Interval_62656</th>\n",
       "      <th>Interval_62674</th>\n",
       "      <th>Interval_62675</th>\n",
       "      <th>Interval_62661</th>\n",
       "      <th>Interval_62671</th>\n",
       "      <th>Interval_62657</th>\n",
       "      <th>Interval_62663</th>\n",
       "      <th>...</th>\n",
       "      <th>Interval_62680</th>\n",
       "      <th>Interval_62676</th>\n",
       "      <th>Interval_62678</th>\n",
       "      <th>Interval_62679</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <th>Eigenpeak 4</th>\n",
       "      <th>Eigenpeak 5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_62671</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.382206</td>\n",
       "      <td>0.366269</td>\n",
       "      <td>0.108898</td>\n",
       "      <td>-0.193444</td>\n",
       "      <td>-0.772613</td>\n",
       "      <td>-0.271212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_62668</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.507081</td>\n",
       "      <td>0.638341</td>\n",
       "      <td>-0.034938</td>\n",
       "      <td>0.130182</td>\n",
       "      <td>0.199180</td>\n",
       "      <td>-0.526829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_62663</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.067430</td>\n",
       "      <td>-0.204399</td>\n",
       "      <td>-0.571452</td>\n",
       "      <td>-0.730330</td>\n",
       "      <td>0.076152</td>\n",
       "      <td>-0.296539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_62658</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.387912</td>\n",
       "      <td>-0.617260</td>\n",
       "      <td>0.035258</td>\n",
       "      <td>0.321552</td>\n",
       "      <td>-0.402266</td>\n",
       "      <td>-0.449508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_62677</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.113173</td>\n",
       "      <td>-0.172486</td>\n",
       "      <td>0.768639</td>\n",
       "      <td>-0.389649</td>\n",
       "      <td>0.303173</td>\n",
       "      <td>-0.350563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_62669</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.654948</td>\n",
       "      <td>-0.076348</td>\n",
       "      <td>-0.261370</td>\n",
       "      <td>0.396267</td>\n",
       "      <td>0.322282</td>\n",
       "      <td>-0.485806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM65B_R</th>\n",
       "      <td>-0.109015</td>\n",
       "      <td>-0.153073</td>\n",
       "      <td>0.142392</td>\n",
       "      <td>-0.082905</td>\n",
       "      <td>-0.424297</td>\n",
       "      <td>-0.140013</td>\n",
       "      <td>0.075671</td>\n",
       "      <td>0.452342</td>\n",
       "      <td>-0.047006</td>\n",
       "      <td>0.318944</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.84397</td>\n",
       "      <td>-0.372134</td>\n",
       "      <td>-0.277167</td>\n",
       "      <td>-0.445709</td>\n",
       "      <td>0.004413</td>\n",
       "      <td>0.271128</td>\n",
       "      <td>-0.024230</td>\n",
       "      <td>-0.181451</td>\n",
       "      <td>-0.181565</td>\n",
       "      <td>-0.460778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM65B_Rsq</th>\n",
       "      <td>0.011884</td>\n",
       "      <td>0.023431</td>\n",
       "      <td>0.020276</td>\n",
       "      <td>0.006873</td>\n",
       "      <td>0.180028</td>\n",
       "      <td>0.019604</td>\n",
       "      <td>0.005726</td>\n",
       "      <td>0.204613</td>\n",
       "      <td>0.00221</td>\n",
       "      <td>0.101725</td>\n",
       "      <td>...</td>\n",
       "      <td>0.712285</td>\n",
       "      <td>0.138483</td>\n",
       "      <td>0.076822</td>\n",
       "      <td>0.198657</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.073510</td>\n",
       "      <td>0.000587</td>\n",
       "      <td>0.032925</td>\n",
       "      <td>0.032966</td>\n",
       "      <td>0.212316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM65B_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID          Interval_62659 Interval_62655 Interval_62672 Interval_62656  \\\n",
       "PeakID                                                                        \n",
       "Interval_62671             0.0            0.0            0.0            0.0   \n",
       "Interval_62668             0.0            0.0            0.0            0.0   \n",
       "Interval_62663             0.0            0.0            0.0            0.0   \n",
       "Interval_62658             0.0            0.0            0.0            0.0   \n",
       "Interval_62677             0.0            0.0            0.0            0.0   \n",
       "Interval_62669             0.0            0.0            0.0            0.0   \n",
       "FAM65B_R             -0.109015      -0.153073       0.142392      -0.082905   \n",
       "FAM65B_Rsq            0.011884       0.023431       0.020276       0.006873   \n",
       "FAM65B_promoter          False          False          False          False   \n",
       "\n",
       "PeakID          Interval_62674 Interval_62675 Interval_62661 Interval_62671  \\\n",
       "PeakID                                                                        \n",
       "Interval_62671             0.0            0.0            0.0            1.0   \n",
       "Interval_62668             0.0            0.0            0.0            0.0   \n",
       "Interval_62663             0.0            0.0            0.0            0.0   \n",
       "Interval_62658             0.0            0.0            0.0            0.0   \n",
       "Interval_62677             0.0            0.0            0.0            0.0   \n",
       "Interval_62669             0.0            0.0            0.0            0.0   \n",
       "FAM65B_R             -0.424297      -0.140013       0.075671       0.452342   \n",
       "FAM65B_Rsq            0.180028       0.019604       0.005726       0.204613   \n",
       "FAM65B_promoter          False          False           True          False   \n",
       "\n",
       "PeakID          Interval_62657 Interval_62663  ... Interval_62680  \\\n",
       "PeakID                                         ...                  \n",
       "Interval_62671             0.0            0.0  ...            0.0   \n",
       "Interval_62668             0.0            0.0  ...            0.0   \n",
       "Interval_62663             0.0            1.0  ...            0.0   \n",
       "Interval_62658             0.0            0.0  ...            0.0   \n",
       "Interval_62677             0.0            0.0  ...            0.0   \n",
       "Interval_62669             0.0            0.0  ...            0.0   \n",
       "FAM65B_R             -0.047006       0.318944  ...       -0.84397   \n",
       "FAM65B_Rsq             0.00221       0.101725  ...       0.712285   \n",
       "FAM65B_promoter          False          False  ...          False   \n",
       "\n",
       "PeakID          Interval_62676 Interval_62678 Interval_62679 Eigenpeak 0  \\\n",
       "PeakID                                                                     \n",
       "Interval_62671             0.0            0.0            0.0    0.382206   \n",
       "Interval_62668             0.0            0.0            0.0   -0.507081   \n",
       "Interval_62663             0.0            0.0            0.0   -0.067430   \n",
       "Interval_62658             0.0            0.0            0.0   -0.387912   \n",
       "Interval_62677             0.0            0.0            0.0    0.113173   \n",
       "Interval_62669             0.0            0.0            0.0    0.654948   \n",
       "FAM65B_R             -0.372134      -0.277167      -0.445709    0.004413   \n",
       "FAM65B_Rsq            0.138483       0.076822       0.198657    0.000019   \n",
       "FAM65B_promoter          False          False          False         NaN   \n",
       "\n",
       "PeakID          Eigenpeak 1 Eigenpeak 2 Eigenpeak 3 Eigenpeak 4 Eigenpeak 5  \n",
       "PeakID                                                                       \n",
       "Interval_62671     0.366269    0.108898   -0.193444   -0.772613   -0.271212  \n",
       "Interval_62668     0.638341   -0.034938    0.130182    0.199180   -0.526829  \n",
       "Interval_62663    -0.204399   -0.571452   -0.730330    0.076152   -0.296539  \n",
       "Interval_62658    -0.617260    0.035258    0.321552   -0.402266   -0.449508  \n",
       "Interval_62677    -0.172486    0.768639   -0.389649    0.303173   -0.350563  \n",
       "Interval_62669    -0.076348   -0.261370    0.396267    0.322282   -0.485806  \n",
       "FAM65B_R           0.271128   -0.024230   -0.181451   -0.181565   -0.460778  \n",
       "FAM65B_Rsq         0.073510    0.000587    0.032925    0.032966    0.212316  \n",
       "FAM65B_promoter         NaN         NaN         NaN         NaN         NaN  \n",
       "\n",
       "[9 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_63905</th>\n",
       "      <th>Interval_63900</th>\n",
       "      <th>Interval_63908</th>\n",
       "      <th>Interval_63901</th>\n",
       "      <th>Interval_63904</th>\n",
       "      <th>Interval_63906</th>\n",
       "      <th>Interval_63899</th>\n",
       "      <th>Interval_63907</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <th>Eigenpeak 4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_63901</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.054611</td>\n",
       "      <td>-0.697250</td>\n",
       "      <td>-0.277057</td>\n",
       "      <td>-0.476678</td>\n",
       "      <td>-0.454838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_63904</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.366790</td>\n",
       "      <td>0.379933</td>\n",
       "      <td>0.506878</td>\n",
       "      <td>-0.664412</td>\n",
       "      <td>-0.150826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_63907</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.064720</td>\n",
       "      <td>-0.448504</td>\n",
       "      <td>0.766189</td>\n",
       "      <td>0.409253</td>\n",
       "      <td>-0.200304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_63905</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.637302</td>\n",
       "      <td>-0.238712</td>\n",
       "      <td>0.272951</td>\n",
       "      <td>-0.404217</td>\n",
       "      <td>0.546780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_63906</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.672413</td>\n",
       "      <td>-0.333698</td>\n",
       "      <td>-0.069039</td>\n",
       "      <td>-0.021362</td>\n",
       "      <td>0.656722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAQR8_R</th>\n",
       "      <td>-0.386708</td>\n",
       "      <td>-0.171704</td>\n",
       "      <td>0.172362</td>\n",
       "      <td>0.366373</td>\n",
       "      <td>0.306026</td>\n",
       "      <td>-0.904939</td>\n",
       "      <td>-0.157974</td>\n",
       "      <td>0.294247</td>\n",
       "      <td>-0.336263</td>\n",
       "      <td>0.166089</td>\n",
       "      <td>0.232729</td>\n",
       "      <td>-0.072473</td>\n",
       "      <td>-0.813968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAQR8_Rsq</th>\n",
       "      <td>0.149543</td>\n",
       "      <td>0.029482</td>\n",
       "      <td>0.029709</td>\n",
       "      <td>0.134229</td>\n",
       "      <td>0.093652</td>\n",
       "      <td>0.818914</td>\n",
       "      <td>0.024956</td>\n",
       "      <td>0.086581</td>\n",
       "      <td>0.113073</td>\n",
       "      <td>0.027585</td>\n",
       "      <td>0.054163</td>\n",
       "      <td>0.005252</td>\n",
       "      <td>0.662544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAQR8_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID         Interval_63905 Interval_63900 Interval_63908 Interval_63901  \\\n",
       "PeakID                                                                       \n",
       "Interval_63901            0.0            0.0            0.0            1.0   \n",
       "Interval_63904            0.0            0.0            0.0            0.0   \n",
       "Interval_63907            0.0            0.0            0.0            0.0   \n",
       "Interval_63905            1.0            0.0            0.0            0.0   \n",
       "Interval_63906            0.0            0.0            0.0            0.0   \n",
       "PAQR8_R             -0.386708      -0.171704       0.172362       0.366373   \n",
       "PAQR8_Rsq            0.149543       0.029482       0.029709       0.134229   \n",
       "PAQR8_promoter          False          False          False          False   \n",
       "\n",
       "PeakID         Interval_63904 Interval_63906 Interval_63899 Interval_63907  \\\n",
       "PeakID                                                                       \n",
       "Interval_63901            0.0            0.0            0.0            0.0   \n",
       "Interval_63904            1.0            0.0            0.0            0.0   \n",
       "Interval_63907            0.0            0.0            0.0            1.0   \n",
       "Interval_63905            0.0            0.0            0.0            0.0   \n",
       "Interval_63906            0.0            1.0            0.0            0.0   \n",
       "PAQR8_R              0.306026      -0.904939      -0.157974       0.294247   \n",
       "PAQR8_Rsq            0.093652       0.818914       0.024956       0.086581   \n",
       "PAQR8_promoter          False          False          False          False   \n",
       "\n",
       "PeakID          Eigenpeak 0  Eigenpeak 1  Eigenpeak 2  Eigenpeak 3  \\\n",
       "PeakID                                                               \n",
       "Interval_63901     0.054611    -0.697250    -0.277057    -0.476678   \n",
       "Interval_63904     0.366790     0.379933     0.506878    -0.664412   \n",
       "Interval_63907     0.064720    -0.448504     0.766189     0.409253   \n",
       "Interval_63905    -0.637302    -0.238712     0.272951    -0.404217   \n",
       "Interval_63906     0.672413    -0.333698    -0.069039    -0.021362   \n",
       "PAQR8_R           -0.336263     0.166089     0.232729    -0.072473   \n",
       "PAQR8_Rsq          0.113073     0.027585     0.054163     0.005252   \n",
       "PAQR8_promoter          NaN          NaN          NaN          NaN   \n",
       "\n",
       "PeakID          Eigenpeak 4  \n",
       "PeakID                       \n",
       "Interval_63901    -0.454838  \n",
       "Interval_63904    -0.150826  \n",
       "Interval_63907    -0.200304  \n",
       "Interval_63905     0.546780  \n",
       "Interval_63906     0.656722  \n",
       "PAQR8_R           -0.813968  \n",
       "PAQR8_Rsq          0.662544  \n",
       "PAQR8_promoter          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_69616</th>\n",
       "      <th>Interval_69611</th>\n",
       "      <th>Interval_69615</th>\n",
       "      <th>Interval_69613</th>\n",
       "      <th>Interval_69617</th>\n",
       "      <th>Interval_69614</th>\n",
       "      <th>Interval_69609</th>\n",
       "      <th>Interval_69618</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_69614</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.547623</td>\n",
       "      <td>0.627741</td>\n",
       "      <td>-0.553218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_69613</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.428161</td>\n",
       "      <td>0.778284</td>\n",
       "      <td>0.459295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_69611</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.718879</td>\n",
       "      <td>-0.014654</td>\n",
       "      <td>0.694981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMS2P1_R</th>\n",
       "      <td>-0.119458</td>\n",
       "      <td>-0.80737</td>\n",
       "      <td>0.161394</td>\n",
       "      <td>-0.266771</td>\n",
       "      <td>0.099508</td>\n",
       "      <td>0.36348</td>\n",
       "      <td>-0.031862</td>\n",
       "      <td>0.012143</td>\n",
       "      <td>-0.420483</td>\n",
       "      <td>0.033076</td>\n",
       "      <td>-0.691239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMS2P1_Rsq</th>\n",
       "      <td>0.01427</td>\n",
       "      <td>0.651846</td>\n",
       "      <td>0.026048</td>\n",
       "      <td>0.071167</td>\n",
       "      <td>0.009902</td>\n",
       "      <td>0.132118</td>\n",
       "      <td>0.001015</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>0.176806</td>\n",
       "      <td>0.001094</td>\n",
       "      <td>0.477812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMS2P1_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID          Interval_69616 Interval_69611 Interval_69615 Interval_69613  \\\n",
       "PeakID                                                                        \n",
       "Interval_69614             0.0            0.0            0.0            0.0   \n",
       "Interval_69613             0.0            0.0            0.0            1.0   \n",
       "Interval_69611             0.0            1.0            0.0            0.0   \n",
       "PMS2P1_R             -0.119458       -0.80737       0.161394      -0.266771   \n",
       "PMS2P1_Rsq             0.01427       0.651846       0.026048       0.071167   \n",
       "PMS2P1_promoter          False          False          False          False   \n",
       "\n",
       "PeakID          Interval_69617 Interval_69614 Interval_69609 Interval_69618  \\\n",
       "PeakID                                                                        \n",
       "Interval_69614             0.0            1.0            0.0            0.0   \n",
       "Interval_69613             0.0            0.0            0.0            0.0   \n",
       "Interval_69611             0.0            0.0            0.0            0.0   \n",
       "PMS2P1_R              0.099508        0.36348      -0.031862       0.012143   \n",
       "PMS2P1_Rsq            0.009902       0.132118       0.001015       0.000147   \n",
       "PMS2P1_promoter          False          False          False          False   \n",
       "\n",
       "PeakID           Eigenpeak 0  Eigenpeak 1  Eigenpeak 2  \n",
       "PeakID                                                  \n",
       "Interval_69614      0.547623     0.627741    -0.553218  \n",
       "Interval_69613     -0.428161     0.778284     0.459295  \n",
       "Interval_69611      0.718879    -0.014654     0.694981  \n",
       "PMS2P1_R           -0.420483     0.033076    -0.691239  \n",
       "PMS2P1_Rsq          0.176806     0.001094     0.477812  \n",
       "PMS2P1_promoter          NaN          NaN          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for gene in PCR_ans.keys():\n",
    "    PCR_df = PCR_ans[gene]\n",
    "    int_cols = [col for col in PCR_df.columns if col[:5] == 'Inter']\n",
    "    R_vals = PCR_df.loc[gene+'_R', int_cols]\n",
    "    if R_vals.min() < -0.8:\n",
    "        display(PCR_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Panel:  Principal Component Regression -- summary.\n",
    "\n",
    "We create a dataframe summarizing our PCR analysis, and make a plot of the peaks found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000 genes processed.\n",
      "2000 genes processed.\n",
      "3000 genes processed.\n",
      "4000 genes processed.\n",
      "5000 genes processed.\n",
      "6000 genes processed.\n",
      "7000 genes processed.\n",
      "8000 genes processed.\n",
      "9000 genes processed.\n",
      "10000 genes processed.\n",
      "11000 genes processed.\n"
     ]
    }
   ],
   "source": [
    "PCR_summary = pd.DataFrame(index = PCR_ans.keys(), \n",
    "                           columns=['promoter_R_sq', 'num_eig'])\n",
    "\n",
    "i = 0\n",
    "for gene in PCR_ans.keys():\n",
    "    i += 1\n",
    "    if i%1000 == 0:\n",
    "        print('{} genes processed.'.format(i))\n",
    "    PCR_df = PCR_ans[gene]\n",
    "    \n",
    "    intcols = [col for col in PCR_df.columns if col[:5] == 'Inter']\n",
    "    PCR_summary.loc[gene,'num_peaks'] = len(intcols)\n",
    "    \n",
    "    eigencols = [col for col in PCR_df.columns if col[:3] == 'Eig']\n",
    "    promcols = (PCR_df.loc[gene+'_promoter'] == True)\n",
    "    promcols = promcols[promcols].index\n",
    "    #print(promcols)\n",
    "    if len(promcols) > 0:\n",
    "        #print(PCR_df.loc[gene+'_Rsq',promcols].max())\n",
    "        PCR_summary.loc[gene, 'promoter_R_sq'] = PCR_df.loc[gene+'_Rsq',promcols].max()\n",
    "    else:\n",
    "        PCR_summary.loc[gene, 'promoter_R_sq'] = 0\n",
    "    R2_int = PCR_df.loc[gene+'_Rsq',intcols]\n",
    "    PCR_summary.loc[gene, 'num_sigpeaks'] = (R2_int > 0.2).sum()\n",
    "    if len(eigencols) > 0:\n",
    "        R2_eig = PCR_df.loc[gene+'_Rsq',eigencols]\n",
    "        num_eig = (R2_eig > 0.2).sum()\n",
    "        PCR_summary.loc[gene,'num_eig'] = num_eig\n",
    "        eig_R_sq = R2_eig.sort_values(ascending=False)\n",
    "        for j in range(len(eigencols)):\n",
    "            PCR_summary.loc[gene, 'eigen{}_R_sq'.format(j)] = eig_R_sq[j]\n",
    "    else:\n",
    "        PCR_summary.loc[gene,'num_eig'] = 0\n",
    "    #print(PCR_summary.loc[gene])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>promoter_R_sq</th>\n",
       "      <th>num_eig</th>\n",
       "      <th>num_peaks</th>\n",
       "      <th>num_sigpeaks</th>\n",
       "      <th>eigen0_R_sq</th>\n",
       "      <th>eigen1_R_sq</th>\n",
       "      <th>eigen2_R_sq</th>\n",
       "      <th>eigen3_R_sq</th>\n",
       "      <th>eigen4_R_sq</th>\n",
       "      <th>eigen5_R_sq</th>\n",
       "      <th>eigen6_R_sq</th>\n",
       "      <th>eigen7_R_sq</th>\n",
       "      <th>eigen8_R_sq</th>\n",
       "      <th>eigen9_R_sq</th>\n",
       "      <th>eigen10_R_sq</th>\n",
       "      <th>eigen11_R_sq</th>\n",
       "      <th>eigen12_R_sq</th>\n",
       "      <th>eigen13_R_sq</th>\n",
       "      <th>eigen14_R_sq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.13</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MTND2P28</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-mir-6723</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4A1P2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRIP3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC010970.2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA5-8SP6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PSMA6P1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11848 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              promoter_R_sq num_eig  num_peaks  num_sigpeaks  eigen0_R_sq  \\\n",
       "WASH7P                    0       0        2.0           0.0          NaN   \n",
       "AL627309.1                0       0        0.0           0.0          NaN   \n",
       "RP11-34P13.13             0       0        0.0           0.0          NaN   \n",
       "MTND2P28                  0       0        0.0           0.0          NaN   \n",
       "hsa-mir-6723              0       0        0.0           0.0          NaN   \n",
       "...                     ...     ...        ...           ...          ...   \n",
       "EIF4A1P2                  0       0        0.0           0.0          NaN   \n",
       "SRIP3                     0       0        0.0           0.0          NaN   \n",
       "AC010970.2                0       0        0.0           0.0          NaN   \n",
       "RNA5-8SP6                 0       0        0.0           0.0          NaN   \n",
       "PSMA6P1                   0       0        0.0           0.0          NaN   \n",
       "\n",
       "               eigen1_R_sq  eigen2_R_sq  eigen3_R_sq  eigen4_R_sq  \\\n",
       "WASH7P                 NaN          NaN          NaN          NaN   \n",
       "AL627309.1             NaN          NaN          NaN          NaN   \n",
       "RP11-34P13.13          NaN          NaN          NaN          NaN   \n",
       "MTND2P28               NaN          NaN          NaN          NaN   \n",
       "hsa-mir-6723           NaN          NaN          NaN          NaN   \n",
       "...                    ...          ...          ...          ...   \n",
       "EIF4A1P2               NaN          NaN          NaN          NaN   \n",
       "SRIP3                  NaN          NaN          NaN          NaN   \n",
       "AC010970.2             NaN          NaN          NaN          NaN   \n",
       "RNA5-8SP6              NaN          NaN          NaN          NaN   \n",
       "PSMA6P1                NaN          NaN          NaN          NaN   \n",
       "\n",
       "               eigen5_R_sq  eigen6_R_sq  eigen7_R_sq  eigen8_R_sq  \\\n",
       "WASH7P                 NaN          NaN          NaN          NaN   \n",
       "AL627309.1             NaN          NaN          NaN          NaN   \n",
       "RP11-34P13.13          NaN          NaN          NaN          NaN   \n",
       "MTND2P28               NaN          NaN          NaN          NaN   \n",
       "hsa-mir-6723           NaN          NaN          NaN          NaN   \n",
       "...                    ...          ...          ...          ...   \n",
       "EIF4A1P2               NaN          NaN          NaN          NaN   \n",
       "SRIP3                  NaN          NaN          NaN          NaN   \n",
       "AC010970.2             NaN          NaN          NaN          NaN   \n",
       "RNA5-8SP6              NaN          NaN          NaN          NaN   \n",
       "PSMA6P1                NaN          NaN          NaN          NaN   \n",
       "\n",
       "               eigen9_R_sq  eigen10_R_sq  eigen11_R_sq  eigen12_R_sq  \\\n",
       "WASH7P                 NaN           NaN           NaN           NaN   \n",
       "AL627309.1             NaN           NaN           NaN           NaN   \n",
       "RP11-34P13.13          NaN           NaN           NaN           NaN   \n",
       "MTND2P28               NaN           NaN           NaN           NaN   \n",
       "hsa-mir-6723           NaN           NaN           NaN           NaN   \n",
       "...                    ...           ...           ...           ...   \n",
       "EIF4A1P2               NaN           NaN           NaN           NaN   \n",
       "SRIP3                  NaN           NaN           NaN           NaN   \n",
       "AC010970.2             NaN           NaN           NaN           NaN   \n",
       "RNA5-8SP6              NaN           NaN           NaN           NaN   \n",
       "PSMA6P1                NaN           NaN           NaN           NaN   \n",
       "\n",
       "               eigen13_R_sq  eigen14_R_sq  \n",
       "WASH7P                  NaN           NaN  \n",
       "AL627309.1              NaN           NaN  \n",
       "RP11-34P13.13           NaN           NaN  \n",
       "MTND2P28                NaN           NaN  \n",
       "hsa-mir-6723            NaN           NaN  \n",
       "...                     ...           ...  \n",
       "EIF4A1P2                NaN           NaN  \n",
       "SRIP3                   NaN           NaN  \n",
       "AC010970.2              NaN           NaN  \n",
       "RNA5-8SP6               NaN           NaN  \n",
       "PSMA6P1                 NaN           NaN  \n",
       "\n",
       "[11848 rows x 19 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_genes = adata.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_genes = [gene for gene in plot_genes if gene in PCR_summary.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-69-d8f89d62e4ff>:25: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "sns.set_style('white')\n",
    "figsize=(65*mm, 40*mm)\n",
    "fig, ax = plt.subplots(figsize=figsize, dpi=300)\n",
    "setfontsize()\n",
    "\n",
    "bins=np.arange(0,75,1)\n",
    "sns.histplot(PCR_summary.loc[plot_genes,'num_peaks'], bins=bins, ax=ax, alpha=0.3, label='All peaks')\n",
    "sns.histplot(PCR_summary.loc[plot_genes,'num_sigpeaks'], bins=bins, ax=ax, color='orange', label='Peaks correlated ($R^2 > 0.2$) to gene')\n",
    "sns.histplot(PCR_summary.loc[plot_genes,'num_eig'], bins=bins, ax=ax, color='darkgreen', label='Eigenpeaks correlated ($R^2 > 0.2$) to gene')\n",
    "ax.legend()\n",
    "plt.yscale('log')\n",
    "plt.xscale('symlog',linthresh=20.0)\n",
    "xt = [0,1,2,3,4,5,10,20,30,40,50]\n",
    "ax.set_xticks([x + 0.5 for x in xt])\n",
    "ax.set_xticklabels([0,1,2,3,4,5,10,20,30,40,50], y=0.07)\n",
    "ax.set_yticks([1,10,100,1000,10000])\n",
    "ax.set_yticklabels(['1','10','100','1,000','10,000'], x=0.03)\n",
    "ax.set_xlim(0,80)\n",
    "ax.set_xlabel('Number of peaks within 50kb of gene')\n",
    "ax.set_ylabel('Count', labelpad=-5)\n",
    "plt.suptitle('Peaks within 50kb of {} genes'.format(len(PCR_summary.loc[plot_genes])))\n",
    "\n",
    "set_size(fig, figsize)\n",
    "plt.savefig(\"Panels/PCR_Peaks.pdf\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multiomics plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['P1_A03',\n",
       " 'P1_A05',\n",
       " 'P1_A08',\n",
       " 'P1_A12',\n",
       " 'P1_B10',\n",
       " 'P1_C11',\n",
       " 'P1_D09',\n",
       " 'P1_F10',\n",
       " 'P1_H01',\n",
       " 'P1_H10',\n",
       " 'P2_A04',\n",
       " 'P2_B11',\n",
       " 'P2_G01',\n",
       " 'P2_G09',\n",
       " 'P2_H04',\n",
       " 'P2_H05']"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pg_clones = list(peak_clonal.index)\n",
    "pg_clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>var_names</th>\n",
       "      <th>CD57 FITC</th>\n",
       "      <th>A2 YF DEXTRA APC</th>\n",
       "      <th>CD8 A7</th>\n",
       "      <th>CD62L BV670</th>\n",
       "      <th>PD1 BV711</th>\n",
       "      <th>CD27 BV786</th>\n",
       "      <th>CD94 PE</th>\n",
       "      <th>CD3 PC5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Clone_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P1_A03</th>\n",
       "      <td>902.903320</td>\n",
       "      <td>27506.250000</td>\n",
       "      <td>8444.175781</td>\n",
       "      <td>-208.290939</td>\n",
       "      <td>1376.424927</td>\n",
       "      <td>-97.778107</td>\n",
       "      <td>2359.556641</td>\n",
       "      <td>2292.757568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A05</th>\n",
       "      <td>803.705139</td>\n",
       "      <td>19328.957031</td>\n",
       "      <td>7142.438965</td>\n",
       "      <td>-32.177135</td>\n",
       "      <td>903.623413</td>\n",
       "      <td>-68.900253</td>\n",
       "      <td>393.199371</td>\n",
       "      <td>2615.705322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A08</th>\n",
       "      <td>576.234619</td>\n",
       "      <td>19552.335938</td>\n",
       "      <td>7188.782227</td>\n",
       "      <td>-37.302555</td>\n",
       "      <td>1303.275879</td>\n",
       "      <td>-41.017300</td>\n",
       "      <td>7642.252441</td>\n",
       "      <td>2269.702393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A12</th>\n",
       "      <td>413.486084</td>\n",
       "      <td>15468.637695</td>\n",
       "      <td>4991.385742</td>\n",
       "      <td>-77.271622</td>\n",
       "      <td>994.417175</td>\n",
       "      <td>-49.785648</td>\n",
       "      <td>279.779022</td>\n",
       "      <td>1625.113037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_B10</th>\n",
       "      <td>42.948910</td>\n",
       "      <td>14563.261719</td>\n",
       "      <td>8508.237305</td>\n",
       "      <td>-55.116821</td>\n",
       "      <td>2128.482666</td>\n",
       "      <td>22.142307</td>\n",
       "      <td>1226.074707</td>\n",
       "      <td>2014.063965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_C11</th>\n",
       "      <td>1643.779175</td>\n",
       "      <td>23151.318359</td>\n",
       "      <td>10985.375000</td>\n",
       "      <td>-18.524429</td>\n",
       "      <td>1235.683594</td>\n",
       "      <td>-62.457111</td>\n",
       "      <td>2831.641602</td>\n",
       "      <td>2888.821533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_D09</th>\n",
       "      <td>1441.065430</td>\n",
       "      <td>17640.460938</td>\n",
       "      <td>6325.950684</td>\n",
       "      <td>567.863037</td>\n",
       "      <td>2455.682129</td>\n",
       "      <td>-83.398872</td>\n",
       "      <td>60.759659</td>\n",
       "      <td>2319.213623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_F10</th>\n",
       "      <td>203.625443</td>\n",
       "      <td>16097.100586</td>\n",
       "      <td>7561.971680</td>\n",
       "      <td>-58.249008</td>\n",
       "      <td>1241.386597</td>\n",
       "      <td>-57.741810</td>\n",
       "      <td>1476.108765</td>\n",
       "      <td>1676.395996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_H01</th>\n",
       "      <td>1690.622925</td>\n",
       "      <td>14977.458984</td>\n",
       "      <td>6277.771973</td>\n",
       "      <td>4.625610</td>\n",
       "      <td>1099.460083</td>\n",
       "      <td>-65.979813</td>\n",
       "      <td>1879.674072</td>\n",
       "      <td>1961.431641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_H10</th>\n",
       "      <td>1068.334717</td>\n",
       "      <td>26251.025391</td>\n",
       "      <td>8835.803711</td>\n",
       "      <td>71.106735</td>\n",
       "      <td>555.246704</td>\n",
       "      <td>-97.844833</td>\n",
       "      <td>154.407333</td>\n",
       "      <td>2497.243896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_A04</th>\n",
       "      <td>1749.253052</td>\n",
       "      <td>22271.263672</td>\n",
       "      <td>7926.547363</td>\n",
       "      <td>-162.400391</td>\n",
       "      <td>741.104126</td>\n",
       "      <td>-83.873314</td>\n",
       "      <td>62.666954</td>\n",
       "      <td>2209.680908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_B11</th>\n",
       "      <td>905.793823</td>\n",
       "      <td>22156.519531</td>\n",
       "      <td>8163.387695</td>\n",
       "      <td>-191.425201</td>\n",
       "      <td>828.401001</td>\n",
       "      <td>-106.923340</td>\n",
       "      <td>2057.955078</td>\n",
       "      <td>2286.909668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_G01</th>\n",
       "      <td>2321.262451</td>\n",
       "      <td>29470.908203</td>\n",
       "      <td>9826.098633</td>\n",
       "      <td>-157.706314</td>\n",
       "      <td>1150.331055</td>\n",
       "      <td>-97.822929</td>\n",
       "      <td>288.096466</td>\n",
       "      <td>2334.123291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_G09</th>\n",
       "      <td>638.899292</td>\n",
       "      <td>22724.232422</td>\n",
       "      <td>8776.172852</td>\n",
       "      <td>-93.520004</td>\n",
       "      <td>1495.582153</td>\n",
       "      <td>-101.572876</td>\n",
       "      <td>7028.983398</td>\n",
       "      <td>2695.924805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_H04</th>\n",
       "      <td>247.371338</td>\n",
       "      <td>35838.644531</td>\n",
       "      <td>9158.025391</td>\n",
       "      <td>-105.397858</td>\n",
       "      <td>810.228943</td>\n",
       "      <td>25.330467</td>\n",
       "      <td>145.109039</td>\n",
       "      <td>2907.352783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_H05</th>\n",
       "      <td>3864.262207</td>\n",
       "      <td>17620.658203</td>\n",
       "      <td>9249.083984</td>\n",
       "      <td>34.258526</td>\n",
       "      <td>702.913269</td>\n",
       "      <td>-84.052254</td>\n",
       "      <td>63.130737</td>\n",
       "      <td>1906.631836</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "var_names    CD57 FITC  A2 YF DEXTRA APC        CD8 A7  CD62L BV670  \\\n",
       "Clone_ID                                                              \n",
       "P1_A03      902.903320      27506.250000   8444.175781  -208.290939   \n",
       "P1_A05      803.705139      19328.957031   7142.438965   -32.177135   \n",
       "P1_A08      576.234619      19552.335938   7188.782227   -37.302555   \n",
       "P1_A12      413.486084      15468.637695   4991.385742   -77.271622   \n",
       "P1_B10       42.948910      14563.261719   8508.237305   -55.116821   \n",
       "P1_C11     1643.779175      23151.318359  10985.375000   -18.524429   \n",
       "P1_D09     1441.065430      17640.460938   6325.950684   567.863037   \n",
       "P1_F10      203.625443      16097.100586   7561.971680   -58.249008   \n",
       "P1_H01     1690.622925      14977.458984   6277.771973     4.625610   \n",
       "P1_H10     1068.334717      26251.025391   8835.803711    71.106735   \n",
       "P2_A04     1749.253052      22271.263672   7926.547363  -162.400391   \n",
       "P2_B11      905.793823      22156.519531   8163.387695  -191.425201   \n",
       "P2_G01     2321.262451      29470.908203   9826.098633  -157.706314   \n",
       "P2_G09      638.899292      22724.232422   8776.172852   -93.520004   \n",
       "P2_H04      247.371338      35838.644531   9158.025391  -105.397858   \n",
       "P2_H05     3864.262207      17620.658203   9249.083984    34.258526   \n",
       "\n",
       "var_names    PD1 BV711  CD27 BV786      CD94 PE      CD3 PC5  \n",
       "Clone_ID                                                      \n",
       "P1_A03     1376.424927  -97.778107  2359.556641  2292.757568  \n",
       "P1_A05      903.623413  -68.900253   393.199371  2615.705322  \n",
       "P1_A08     1303.275879  -41.017300  7642.252441  2269.702393  \n",
       "P1_A12      994.417175  -49.785648   279.779022  1625.113037  \n",
       "P1_B10     2128.482666   22.142307  1226.074707  2014.063965  \n",
       "P1_C11     1235.683594  -62.457111  2831.641602  2888.821533  \n",
       "P1_D09     2455.682129  -83.398872    60.759659  2319.213623  \n",
       "P1_F10     1241.386597  -57.741810  1476.108765  1676.395996  \n",
       "P1_H01     1099.460083  -65.979813  1879.674072  1961.431641  \n",
       "P1_H10      555.246704  -97.844833   154.407333  2497.243896  \n",
       "P2_A04      741.104126  -83.873314    62.666954  2209.680908  \n",
       "P2_B11      828.401001 -106.923340  2057.955078  2286.909668  \n",
       "P2_G01     1150.331055  -97.822929   288.096466  2334.123291  \n",
       "P2_G09     1495.582153 -101.572876  7028.983398  2695.924805  \n",
       "P2_H04      810.228943   25.330467   145.109039  2907.352783  \n",
       "P2_H05      702.913269  -84.052254    63.130737  1906.631836  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_clonal = prot_data.to_df().groupby(prot_data.obs.Clone_ID).mean()\n",
    "prot_clonal = prot_clonal.loc[pg_clones]\n",
    "prot_clonal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.ticker import FormatStrFormatter\n",
    "\n",
    "def pgplot(gene, protein = None, cutoff=0.2, window=50000):\n",
    "    sns.set_style('white')\n",
    "    if protein:\n",
    "        num_rows = 3\n",
    "        hr = [2,1,1]\n",
    "    else:\n",
    "        num_rows = 2\n",
    "        hr=[2,1]\n",
    "        \n",
    "    figsize=(60*mm,15*mm*num_rows)\n",
    "    fig, ax = plt.subplots(num_rows,1,figsize=figsize, dpi=300, sharex=True,\n",
    "                           gridspec_kw = {'hspace':0.05, 'height_ratios':hr})            \n",
    "    \n",
    "    plt.rcParams['text.usetex'] = False\n",
    "    setfontsize()\n",
    "\n",
    "    PCR_df = PCR_analysis(gene, window=window)\n",
    "    clone_order = list(gene_clonal[gene].sort_values().index)\n",
    "    \n",
    "    gplot_df = pd.DataFrame(columns=[gene,'Clone'])\n",
    "    gplot_df[gene] = adata.to_df()[gene]\n",
    "    \n",
    "    \n",
    "    gplot_df['Clone'] = pd.Categorical(adata.obs.Clone_ID, categories = clone_order, ordered = True)\n",
    "    gplot_df = gplot_df[gplot_df.Clone.isin(pg_clones)]\n",
    "    clone_order = [clone for clone in clone_order if clone in pg_clones]\n",
    "        \n",
    "    for clone in clone_order:\n",
    "        for row in range(num_rows):\n",
    "            ax[row].axvline(clone,linestyle=':',color='gray',linewidth=0.5, alpha=0.5, zorder=-5)\n",
    "    \n",
    "    # GENE EXPRESSION PLOT\n",
    "    \n",
    "    sns.scatterplot(x='Clone', y=gene, hue='Clone', data=gplot_df, ax=ax[0], legend=False)\n",
    "    gene_ticks = [gplot_df[gene].min(), gplot_df[gene].mean(), gplot_df[gene].max()]\n",
    "    ax[0].set_yticks(gene_ticks)\n",
    "    ax[0].set_yticklabels(gene_ticks, x=0.03)\n",
    "    ax[0].yaxis.set_major_formatter(FormatStrFormatter('%.1f'))\n",
    "    ax[0].set_ylabel(gene)\n",
    "    ax[0].yaxis.set_label_position(\"right\")\n",
    "    \n",
    "    # ATAC PLOT\n",
    "    peaklist = [peak for peak in PCR_df.index if peak[:8] == 'Interval']\n",
    "    peaklist = [peak for peak in peaklist if PCR_df.loc[gene+'_Rsq',peak] > cutoff]\n",
    "    pplot_df = ATAC_data.to_df()[peaklist].groupby(ATAC_data.obs.Clone_ID).mean()\n",
    "    pplot_df['Clone'] = pd.Categorical(pplot_df.index, categories = clone_order, ordered = True)\n",
    "    pplot_df = pplot_df[pplot_df.Clone.isin(pg_clones)]\n",
    "    \n",
    "    pplot_df[peaklist] = (pplot_df[peaklist] - pplot_df[peaklist].min()) #/ pplot_df[peaklist].max()\n",
    "    pplot_df[peaklist] = pplot_df[peaklist] / pplot_df[peaklist].max()\n",
    "    \n",
    "    no_peaks = 0\n",
    "    for peak in peaklist:\n",
    "        peak_Rsq = PCR_df.loc[gene+'_Rsq',peak]\n",
    "        if peak_Rsq > 0.05:\n",
    "            peak_label = '${:1.2f}$'.format(peak_Rsq)\n",
    "            no_peaks += 1\n",
    "        else:\n",
    "            peak_label = None\n",
    "        sns.lineplot(x='Clone', y=peak, data=pplot_df, ax=ax[-1], linewidth=0.7,\n",
    "                    label=peak_label, alpha=min(1,peak_Rsq*1.5), zorder=np.round(peak_Rsq*100).astype(int)) #, label=peak, legend=False)\n",
    "    \n",
    "    ax[-1].set_ylabel('Peak')\n",
    "    ax[-1].yaxis.set_label_position(\"right\")\n",
    "    ax[-1].set_yticks([])\n",
    "    ax[-1].set_xlabel('')\n",
    "    \n",
    "    # PROTEIN PLOT\n",
    "    if protein:\n",
    "        proplot_df = pd.DataFrame(prot_clonal[protein], columns=[protein])\n",
    "        proplot_df[protein] = proplot_df[protein] - proplot_df[protein].min()\n",
    "        proplot_df[protein] = proplot_df[protein] / proplot_df[protein].max()\n",
    "        proplot_df['Clone'] = proplot_df.index\n",
    "        sns.lineplot(x='Clone', y=protein, data=proplot_df, ax=ax[1], linewidth=0.7)\n",
    "        ax[1].text(1,0.5,protein, ha='left')\n",
    "        ax[1].set_ylabel('Prot')\n",
    "        ax[1].yaxis.set_label_position(\"right\")\n",
    "        ax[1].set_yticks([])\n",
    "        ax[1].set_xlabel('')\n",
    "        ax[1].spines['top'].set_color('lightgray')\n",
    "        ax[1].spines['bottom'].set_color('lightgray')\n",
    "    \n",
    "    # General formatting.\n",
    "    ax[0].spines['bottom'].set_color('lightgray')\n",
    "    ax[-1].spines['top'].set_color('lightgray')\n",
    "    \n",
    "    plt.xticks(rotation=45)\n",
    "    plt.xticks([])\n",
    "    \n",
    "    if protein:\n",
    "        title = '{}, {}, and {} peaks.'.format(gene, protein, len(peaklist))\n",
    "    else:\n",
    "        title = '{} and {} peaks.'.format(gene, len(peaklist))\n",
    "    plt.suptitle(title, y=1.02)\n",
    "    \n",
    "    # Plot legend\n",
    "    if no_peaks <= 6:\n",
    "        ncol = no_peaks\n",
    "    else:\n",
    "        ncol = (no_peaks + 1)//2\n",
    "    lgnd = plt.legend(bbox_to_anchor=(0.5, 0.05), title = 'Peak/Gene $R^2$',  \n",
    "                      loc=9, ncol=ncol, framealpha=0.5)\n",
    "    lgnd.get_frame().set_linewidth(0.5)\n",
    "    \n",
    "    # Save and show figure.\n",
    "    set_size(fig, figsize)\n",
    "    plt.savefig(\"Panels/MultiPlots/{}_multi.pdf\".format(gene), bbox_inches='tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('KLRD1', protein='CD94 PE')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('GNLY')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('CTLA4', window=50000, cutoff=0.05)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quick look at CTLA4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <th>Eigenpeak 4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Clone_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P1_A03</th>\n",
       "      <td>0.108131</td>\n",
       "      <td>-0.127294</td>\n",
       "      <td>0.463815</td>\n",
       "      <td>0.647343</td>\n",
       "      <td>-0.958572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A05</th>\n",
       "      <td>0.468872</td>\n",
       "      <td>0.278330</td>\n",
       "      <td>-0.824381</td>\n",
       "      <td>0.243862</td>\n",
       "      <td>0.377747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A08</th>\n",
       "      <td>-0.960432</td>\n",
       "      <td>0.277748</td>\n",
       "      <td>0.241828</td>\n",
       "      <td>1.252980</td>\n",
       "      <td>1.101614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_A12</th>\n",
       "      <td>-0.628404</td>\n",
       "      <td>1.068183</td>\n",
       "      <td>-0.416881</td>\n",
       "      <td>0.885308</td>\n",
       "      <td>1.645937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_B10</th>\n",
       "      <td>-0.113053</td>\n",
       "      <td>-0.060938</td>\n",
       "      <td>-0.022482</td>\n",
       "      <td>-1.077174</td>\n",
       "      <td>0.595303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_C11</th>\n",
       "      <td>-0.268532</td>\n",
       "      <td>-0.342675</td>\n",
       "      <td>-0.480757</td>\n",
       "      <td>-0.025326</td>\n",
       "      <td>-0.748896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_D09</th>\n",
       "      <td>0.213048</td>\n",
       "      <td>0.201016</td>\n",
       "      <td>1.564668</td>\n",
       "      <td>0.870184</td>\n",
       "      <td>-3.126437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_F10</th>\n",
       "      <td>1.240361</td>\n",
       "      <td>1.037594</td>\n",
       "      <td>-0.403025</td>\n",
       "      <td>-0.371436</td>\n",
       "      <td>0.765693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_H01</th>\n",
       "      <td>0.239806</td>\n",
       "      <td>-0.291135</td>\n",
       "      <td>-0.927339</td>\n",
       "      <td>-0.158456</td>\n",
       "      <td>-1.471329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P1_H10</th>\n",
       "      <td>0.195390</td>\n",
       "      <td>-0.808948</td>\n",
       "      <td>0.014993</td>\n",
       "      <td>-0.392953</td>\n",
       "      <td>2.281345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_A04</th>\n",
       "      <td>-0.388732</td>\n",
       "      <td>-0.376913</td>\n",
       "      <td>-1.302781</td>\n",
       "      <td>0.017427</td>\n",
       "      <td>-2.902167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_B11</th>\n",
       "      <td>-0.638678</td>\n",
       "      <td>0.408611</td>\n",
       "      <td>0.251309</td>\n",
       "      <td>-2.456379</td>\n",
       "      <td>-0.972239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_G01</th>\n",
       "      <td>0.196549</td>\n",
       "      <td>0.041028</td>\n",
       "      <td>0.074959</td>\n",
       "      <td>0.584707</td>\n",
       "      <td>-1.131353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_G09</th>\n",
       "      <td>0.395070</td>\n",
       "      <td>-0.679282</td>\n",
       "      <td>0.745723</td>\n",
       "      <td>0.262352</td>\n",
       "      <td>0.162009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_H04</th>\n",
       "      <td>-0.065769</td>\n",
       "      <td>0.011731</td>\n",
       "      <td>1.396599</td>\n",
       "      <td>-0.805244</td>\n",
       "      <td>1.526805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2_H05</th>\n",
       "      <td>0.006363</td>\n",
       "      <td>-0.637076</td>\n",
       "      <td>-0.376235</td>\n",
       "      <td>0.522805</td>\n",
       "      <td>2.854545</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Eigenpeak 0  Eigenpeak 1  Eigenpeak 2  Eigenpeak 3  Eigenpeak 4\n",
       "Clone_ID                                                                 \n",
       "P1_A03       0.108131    -0.127294     0.463815     0.647343    -0.958572\n",
       "P1_A05       0.468872     0.278330    -0.824381     0.243862     0.377747\n",
       "P1_A08      -0.960432     0.277748     0.241828     1.252980     1.101614\n",
       "P1_A12      -0.628404     1.068183    -0.416881     0.885308     1.645937\n",
       "P1_B10      -0.113053    -0.060938    -0.022482    -1.077174     0.595303\n",
       "P1_C11      -0.268532    -0.342675    -0.480757    -0.025326    -0.748896\n",
       "P1_D09       0.213048     0.201016     1.564668     0.870184    -3.126437\n",
       "P1_F10       1.240361     1.037594    -0.403025    -0.371436     0.765693\n",
       "P1_H01       0.239806    -0.291135    -0.927339    -0.158456    -1.471329\n",
       "P1_H10       0.195390    -0.808948     0.014993    -0.392953     2.281345\n",
       "P2_A04      -0.388732    -0.376913    -1.302781     0.017427    -2.902167\n",
       "P2_B11      -0.638678     0.408611     0.251309    -2.456379    -0.972239\n",
       "P2_G01       0.196549     0.041028     0.074959     0.584707    -1.131353\n",
       "P2_G09       0.395070    -0.679282     0.745723     0.262352     0.162009\n",
       "P2_H04      -0.065769     0.011731     1.396599    -0.805244     1.526805\n",
       "P2_H05       0.006363    -0.637076    -0.376235     0.522805     2.854545"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <th>Eigenpeak 4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.089875e-08</td>\n",
       "      <td>3.127984e-09</td>\n",
       "      <td>4.827125e-08</td>\n",
       "      <td>1.082314e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <td>3.089875e-08</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.393934e-09</td>\n",
       "      <td>1.638161e-08</td>\n",
       "      <td>1.362356e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <td>3.127984e-09</td>\n",
       "      <td>9.393934e-09</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-3.087906e-08</td>\n",
       "      <td>5.246818e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <td>4.827125e-08</td>\n",
       "      <td>1.638161e-08</td>\n",
       "      <td>-3.087906e-08</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-4.636430e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eigenpeak 4</th>\n",
       "      <td>1.082314e-07</td>\n",
       "      <td>1.362356e-08</td>\n",
       "      <td>5.246818e-08</td>\n",
       "      <td>-4.636430e-08</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Eigenpeak 0   Eigenpeak 1   Eigenpeak 2   Eigenpeak 3  \\\n",
       "Eigenpeak 0  1.000000e+00  3.089875e-08  3.127984e-09  4.827125e-08   \n",
       "Eigenpeak 1  3.089875e-08  1.000000e+00  9.393934e-09  1.638161e-08   \n",
       "Eigenpeak 2  3.127984e-09  9.393934e-09  1.000000e+00 -3.087906e-08   \n",
       "Eigenpeak 3  4.827125e-08  1.638161e-08 -3.087906e-08  1.000000e+00   \n",
       "Eigenpeak 4  1.082314e-07  1.362356e-08  5.246818e-08 -4.636430e-08   \n",
       "\n",
       "              Eigenpeak 4  \n",
       "Eigenpeak 0  1.082314e-07  \n",
       "Eigenpeak 1  1.362356e-08  \n",
       "Eigenpeak 2  5.246818e-08  \n",
       "Eigenpeak 3 -4.636430e-08  \n",
       "Eigenpeak 4  1.000000e+00  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PeakID</th>\n",
       "      <th>Interval_42078</th>\n",
       "      <th>Interval_42082</th>\n",
       "      <th>Interval_42083</th>\n",
       "      <th>Interval_42081</th>\n",
       "      <th>Interval_42076</th>\n",
       "      <th>Interval_42084</th>\n",
       "      <th>Interval_42079</th>\n",
       "      <th>Interval_42077</th>\n",
       "      <th>Eigenpeak 0</th>\n",
       "      <th>Eigenpeak 1</th>\n",
       "      <th>Eigenpeak 2</th>\n",
       "      <th>Eigenpeak 3</th>\n",
       "      <th>Eigenpeak 4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PeakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Interval_42078</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.330946</td>\n",
       "      <td>-0.316371</td>\n",
       "      <td>-0.593672</td>\n",
       "      <td>-0.520477</td>\n",
       "      <td>-0.408706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_42077</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.329080</td>\n",
       "      <td>0.782123</td>\n",
       "      <td>0.139561</td>\n",
       "      <td>-0.025039</td>\n",
       "      <td>-0.509790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_42084</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.320385</td>\n",
       "      <td>-0.534450</td>\n",
       "      <td>0.597491</td>\n",
       "      <td>0.208118</td>\n",
       "      <td>-0.459792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_42081</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.122882</td>\n",
       "      <td>-0.048133</td>\n",
       "      <td>-0.505103</td>\n",
       "      <td>0.786451</td>\n",
       "      <td>-0.330074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Interval_42082</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.815128</td>\n",
       "      <td>-0.015502</td>\n",
       "      <td>0.126297</td>\n",
       "      <td>-0.258184</td>\n",
       "      <td>-0.502708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTLA4_R</th>\n",
       "      <td>0.628518</td>\n",
       "      <td>0.330031</td>\n",
       "      <td>-0.093321</td>\n",
       "      <td>0.460634</td>\n",
       "      <td>-0.007708</td>\n",
       "      <td>0.484888</td>\n",
       "      <td>-0.010952</td>\n",
       "      <td>0.603023</td>\n",
       "      <td>-0.453329</td>\n",
       "      <td>-0.025000</td>\n",
       "      <td>-0.242348</td>\n",
       "      <td>0.039257</td>\n",
       "      <td>-0.639627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTLA4_Rsq</th>\n",
       "      <td>0.395035</td>\n",
       "      <td>0.10892</td>\n",
       "      <td>0.008709</td>\n",
       "      <td>0.212184</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.235117</td>\n",
       "      <td>0.00012</td>\n",
       "      <td>0.363637</td>\n",
       "      <td>0.205507</td>\n",
       "      <td>0.000625</td>\n",
       "      <td>0.058732</td>\n",
       "      <td>0.001541</td>\n",
       "      <td>0.409122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTLA4_promoter</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "PeakID         Interval_42078 Interval_42082 Interval_42083 Interval_42081  \\\n",
       "PeakID                                                                       \n",
       "Interval_42078            1.0            0.0            0.0            0.0   \n",
       "Interval_42077            0.0            0.0            0.0            0.0   \n",
       "Interval_42084            0.0            0.0            0.0            0.0   \n",
       "Interval_42081            0.0            0.0            0.0            1.0   \n",
       "Interval_42082            0.0            1.0            0.0            0.0   \n",
       "CTLA4_R              0.628518       0.330031      -0.093321       0.460634   \n",
       "CTLA4_Rsq            0.395035        0.10892       0.008709       0.212184   \n",
       "CTLA4_promoter          False           True           True          False   \n",
       "\n",
       "PeakID         Interval_42076 Interval_42084 Interval_42079 Interval_42077  \\\n",
       "PeakID                                                                       \n",
       "Interval_42078            0.0            0.0            0.0            0.0   \n",
       "Interval_42077            0.0            0.0            0.0            1.0   \n",
       "Interval_42084            0.0            1.0            0.0            0.0   \n",
       "Interval_42081            0.0            0.0            0.0            0.0   \n",
       "Interval_42082            0.0            0.0            0.0            0.0   \n",
       "CTLA4_R             -0.007708       0.484888      -0.010952       0.603023   \n",
       "CTLA4_Rsq            0.000059       0.235117        0.00012       0.363637   \n",
       "CTLA4_promoter          False          False          False          False   \n",
       "\n",
       "PeakID          Eigenpeak 0  Eigenpeak 1  Eigenpeak 2  Eigenpeak 3  \\\n",
       "PeakID                                                               \n",
       "Interval_42078    -0.330946    -0.316371    -0.593672    -0.520477   \n",
       "Interval_42077    -0.329080     0.782123     0.139561    -0.025039   \n",
       "Interval_42084    -0.320385    -0.534450     0.597491     0.208118   \n",
       "Interval_42081     0.122882    -0.048133    -0.505103     0.786451   \n",
       "Interval_42082     0.815128    -0.015502     0.126297    -0.258184   \n",
       "CTLA4_R           -0.453329    -0.025000    -0.242348     0.039257   \n",
       "CTLA4_Rsq          0.205507     0.000625     0.058732     0.001541   \n",
       "CTLA4_promoter          NaN          NaN          NaN          NaN   \n",
       "\n",
       "PeakID          Eigenpeak 4  \n",
       "PeakID                       \n",
       "Interval_42078    -0.408706  \n",
       "Interval_42077    -0.509790  \n",
       "Interval_42084    -0.459792  \n",
       "Interval_42081    -0.330074  \n",
       "Interval_42082    -0.502708  \n",
       "CTLA4_R           -0.639627  \n",
       "CTLA4_Rsq          0.409122  \n",
       "CTLA4_promoter          NaN  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_a = PCR_analysis('CTLA4', window=50000, verbose=True)\n",
    "PCR_a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'eigen_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-77-350f85726d40>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0meigen_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'eigen_df' is not defined"
     ]
    }
   ],
   "source": [
    "eigen_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCR_a.loc['CTLA4_Rsq'].sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaklist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCR_df.loc['CTLA4_Rsq'].sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCR_df.loc[peaklist, eigenpeaks]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pplot_df[peaklist]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style('white')\n",
    "\n",
    "num_rows = 3\n",
    "hr=[2,1,1]\n",
    "        \n",
    "fig, ax = plt.subplots(num_rows,1,figsize=(2.5,0.75*num_rows), dpi=300, sharex=True,\n",
    "                        gridspec_kw = {'hspace':0.05, 'height_ratios':hr})            \n",
    "    \n",
    "plt.rcParams['text.usetex'] = False\n",
    "setfontsize()\n",
    "\n",
    "gene = 'CTLA4'\n",
    "PCR_df = PCR_analysis(gene, window=50000)\n",
    "clone_order = list(gene_clonal[gene].sort_values().index)\n",
    "    \n",
    "gplot_df = pd.DataFrame(columns=[gene,'Clone'])\n",
    "gplot_df[gene] = adata.to_df()[gene]\n",
    "    \n",
    "    \n",
    "gplot_df['Clone'] = pd.Categorical(adata.obs.Clone_ID, categories = clone_order, ordered = True)\n",
    "gplot_df = gplot_df[gplot_df.Clone.isin(pg_clones)]\n",
    "clone_order = [clone for clone in clone_order if clone in pg_clones]\n",
    "\n",
    "for clone in clone_order:\n",
    "    for row in range(num_rows):\n",
    "        ax[row].axvline(clone,linestyle=':',color='gray',linewidth=0.5, alpha=0.5, zorder=-5)\n",
    "    \n",
    "# GENE EXPRESSION PLOT\n",
    "    \n",
    "sns.scatterplot(x='Clone', y=gene, hue='Clone', data=gplot_df, ax=ax[0], legend=False)\n",
    "gene_ticks = [gplot_df[gene].min(), gplot_df[gene].mean(), gplot_df[gene].max()]\n",
    "ax[0].set_yticks(gene_ticks)\n",
    "ax[0].set_yticklabels(gene_ticks, x=0.03)\n",
    "ax[0].yaxis.set_major_formatter(FormatStrFormatter('%.1f'))\n",
    "ax[0].set_ylabel(gene)\n",
    "ax[0].yaxis.set_label_position(\"right\") \n",
    "\n",
    "\n",
    "# ATAC PLOT\n",
    "peaklist = [peak for peak in PCR_df.index if peak[:8] == 'Interval']\n",
    "peaklist = [peak for peak in peaklist if PCR_df.loc[gene+'_Rsq',peak] > 0.05]\n",
    "pplot_df = ATAC_data.to_df()[peaklist].groupby(ATAC_data.obs.Clone_ID).mean()\n",
    "\n",
    "pplot_df['Clone'] = pd.Categorical(pplot_df.index, categories = clone_order, ordered = True)\n",
    "pplot_df = pplot_df[pplot_df.Clone.isin(pg_clones)]\n",
    "    \n",
    "pplot_df[peaklist] = (pplot_df[peaklist] - pplot_df[peaklist].min()) #/ pplot_df[peaklist].max()\n",
    "pplot_df[peaklist] = pplot_df[peaklist] / pplot_df[peaklist].max()\n",
    "    \n",
    "no_peaks = 0\n",
    "for peak in peaklist:\n",
    "    peak_Rsq = PCR_df.loc[gene+'_Rsq',peak]\n",
    "    if peak_Rsq > 0.05:\n",
    "        peak_label = '${:1.2f}$'.format(peak_Rsq)\n",
    "        no_peaks += 1\n",
    "    else:\n",
    "        peak_label = None\n",
    "    sns.lineplot(x='Clone', y=peak, data=pplot_df, ax=ax[1], linewidth=0.7,\n",
    "                label=peak_label, alpha=min(1,peak_Rsq*1.5), zorder=np.round(peak_Rsq*100).astype(int)) #, label=peak, legend=False)\n",
    "    \n",
    "ax[1].set_ylabel('Peak')\n",
    "ax[1].yaxis.set_label_position(\"right\")\n",
    "ax[1].set_yticks([])\n",
    "ax[1].set_xlabel('')\n",
    "lgnd = ax[1].legend(bbox_to_anchor=(0.5, 0.05), title = 'Peak/Gene $R^2$',  \n",
    "                    loc=9, ncol=ncol, framealpha=0.5)\n",
    "lgnd.get_frame().set_linewidth(0.5)\n",
    "\n",
    "eigenpeaks = PCR_df.loc['CTLA4_Rsq'].sort_values(ascending=False).index\n",
    "eigenpeaks = [epeak for epeak in eigenpeaks if epeak[:5] == 'Eigen']\n",
    "\n",
    "realpeaks = ATAC_data.to_df()[peaklist].groupby(ATAC_data.obs.Clone_ID).mean()\n",
    "realpeaks = realpeaks.loc[pg_clones]\n",
    "normpeaks = (realpeaks - realpeaks.mean()) / realpeaks.std()\n",
    "eigen_df = -normpeaks[peaklist].dot(PCR_df.loc[peaklist, eigenpeaks])\n",
    "eigen_df['Clone'] = pd.Categorical(eigen_df.index, categories = clone_order, ordered = True)\n",
    "eigen_df[eigenpeaks] = (eigen_df[eigenpeaks] - eigen_df[eigenpeaks].min()) #/ pplot_df[peaklist].max()\n",
    "eigen_df[eigenpeaks] = eigen_df[eigenpeaks] / eigen_df[eigenpeaks].max()\n",
    "no_peaks = 0\n",
    "for epeak in eigenpeaks:\n",
    "    peak_Rsq = PCR_df.loc[gene+'_Rsq',epeak]\n",
    "    if peak_Rsq > 0.00:\n",
    "        peak_label = '${:1.2f}$'.format(peak_Rsq)\n",
    "        no_peaks += 1\n",
    "    else:\n",
    "        peak_label=None\n",
    "    sns.lineplot(x='Clone', y=epeak, data=eigen_df, ax=ax[2], linewidth=0.7, label=peak_label,\n",
    "                alpha=min(1,peak_Rsq*1.5), zorder=np.round(peak_Rsq*100).astype(int))\n",
    "ax[2].set_ylabel('Eigenpeak')\n",
    "#ax[2].yaxis.set_label_position(\"right\")\n",
    "ax[2].set_yticks([])\n",
    "ax[2].set_xlabel('')\n",
    "\n",
    "# Plot legend\n",
    "if no_peaks <= 6:\n",
    "    ncol = no_peaks\n",
    "else:\n",
    "    ncol = (no_peaks + 1)//2\n",
    "lgnd = plt.legend(bbox_to_anchor=(0.5, 0.05), title = 'Eigenpeak/Gene $R^2$',  \n",
    "                    loc=9, ncol=ncol, framealpha=0.5)\n",
    "lgnd.get_frame().set_linewidth(0.5)\n",
    "\n",
    "ax[0].spines['bottom'].set_color('lightgray')\n",
    "ax[-1].spines['top'].set_color('lightgray')\n",
    "    \n",
    "plt.xticks(rotation=45)\n",
    "plt.xticks([])\n",
    "    \n",
    "# Save and show figure.\n",
    "set_size(fig, (2.5, 1.5))\n",
    "plt.savefig(\"Panels/MultiPlots/CTLA_Illustration.pdf\".format(gene), bbox_inches='tight')\n",
    "plt.show()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaklist = ['Interval_{}'.format(j) for j in ['42078','42082','42077','42084','42081']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pplot_df[peaklist].corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eigen_df[eigenpeaks].corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pe_df = PCR_df.loc[peaklist, eigenpeaks]\n",
    "pe_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mm = 1/25.4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = np.zeros((5,5)).astype(bool)\n",
    "for i in range(5):\n",
    "    for j in range(i+1):\n",
    "        mask[i,j] = True\n",
    "mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(2,2,figsize=(70*mm, 70*mm), dpi=300, gridspec_kw={'hspace':0.3})\n",
    "\n",
    "sns.heatmap(pplot_df[peaklist].corr(), ax=ax[0,0], cbar=None, vmin=-1.0, vmax=1.0, cmap='coolwarm',\n",
    "           mask=mask)\n",
    "\n",
    "for j,peak in enumerate(peaklist):\n",
    "    alpha = PCR_df.loc['CTLA4_Rsq',peak]\n",
    "    ax[0,0].add_patch(Rectangle((j+0.1,j+0.1),0.8,0.8, color='green', fill=True, clip_on=False,\n",
    "                               alpha=alpha))\n",
    "    ax[0,0].text(j+0.5, j+0.5, '{:.2f}'.format(alpha), ha='center', va='center')\n",
    "    \n",
    "sns.heatmap(eigen_df[eigenpeaks].corr(), ax=ax[1,1],cbar=None, vmin=-1.0, vmax=1.0, cmap='coolwarm',\n",
    "           mask=mask)\n",
    "\n",
    "for j,peak in enumerate(eigenpeaks):\n",
    "    alpha = PCR_df.loc['CTLA4_Rsq',peak]\n",
    "    ax[1,1].add_patch(Rectangle((j+0.1,j+0.1),0.8,0.8, color='green', fill=True, clip_on=False,\n",
    "                               alpha=alpha))\n",
    "    ax[1,1].text(j+0.5, j+0.5, '{:.2f}'.format(alpha), ha='center', va='center')\n",
    "\n",
    "sns.heatmap(-pe_df, ax=ax[0,1], vmin=-1, vmax=1, cmap='PuOr', cbar=None, annot=True, fmt='1.1f')\n",
    "\n",
    "peaklabels = ['Int. {}'.format(peak.split('_')[1]) for peak in peaklist]\n",
    "eigenlabels = ['Eig. {}'.format(peak[-1]) for peak in eigenpeaks]\n",
    "ax[0,0].set_yticklabels(peaklabels)\n",
    "ax[0,0].set_xticklabels([])\n",
    "ax[0,0].set_xlabel('')\n",
    "ax[0,0].set_ylabel('')\n",
    "ax[1,0].set_visible(False)\n",
    "ax[1,1].set_yticklabels(eigenlabels)\n",
    "ax[1,1].set_xticklabels(eigenlabels, rotation=90)\n",
    "ax[1,1].tick_params(labelbottom=False,labeltop=True, pad=-1)\n",
    "ax[0,1].set_yticklabels([])\n",
    "ax[0,1].set_ylabel('')\n",
    "ax[0,1].set_xticklabels([])\n",
    "ax[0,1].set_xlabel('')\n",
    "ax[1,1].set_ylabel('')\n",
    "ax[1,1].set_xlabel('')\n",
    "ax[0,0].set_title('Peak correlations')\n",
    "ax[0,1].set_title('Weights')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eigenpeaks.sort()\n",
    "eigenpeaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaklist.sort()\n",
    "peaklist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "super_df = pd.concat([pplot_df[peaklist], eigen_df[eigenpeaks]], axis=1)\n",
    "super_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "super_df.corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCR_df.loc[peaklist, eigenpeaks]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pgplot('CD28', cutoff=0.1, window=50000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('GTSF1', window=50000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('B2M', window=50000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('PLS3-1', cutoff=0.2, window=50000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:13: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  fig, ax = plt.subplots(num_rows,1,figsize=figsize, dpi=300, sharex=True,\n",
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('ATXN1', cutoff=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('PTPLA', window=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('CX3CR1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('PDCD1', protein='PD1 BV711')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('SELL', protein='CD62L BV670', cutoff=0.1, window=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('CD8A', protein='CD8 A7', cutoff=0.1, window=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('CTLA4', cutoff=0.1, window=200000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-72-3c8e666a6e43>:110: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "pgplot('FARP2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Interactive plots with plotly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "import plotly.express as px\n",
    "from plotly.subplots import make_subplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clone_order = list(gene_clonal['GNLY'].sort_values().index)\n",
    "GNLY_df = pd.DataFrame(adata[adata.obs.Clone_ID.isin(clone_order)].to_df()['GNLY'])\n",
    "GNLY_df['Clone'] = adata.obs.Clone_ID\n",
    "GNLY_df['CloneNo'] = GNLY_df['Clone'].apply(lambda cl : clone_order.index(cl))\n",
    "GNLY_df = GNLY_df.sort_values(by = 'CloneNo')\n",
    "GNLY_df['Obs'] = GNLY_df.index\n",
    "GNLY_df.to_csv('GNLY.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GNLY_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCR_df = PCR_analysis('GNLY', window=50000)\n",
    "peaklist = [peak for peak in PCR_df.index if peak[:8] == 'Interval']\n",
    "peaklist = [peak for peak in peaklist if PCR_df.loc['GNLY_Rsq',peak] > 0.2]\n",
    "pplot_df = ATAC_data.to_df()[peaklist].groupby(ATAC_data.obs.Clone_ID).mean()    \n",
    "pplot_df[peaklist] = (pplot_df[peaklist] - pplot_df[peaklist].min()) #/ pplot_df[peaklist].max()\n",
    "pplot_df[peaklist] = pplot_df[peaklist] / pplot_df[peaklist].max()\n",
    "pplot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pplot_df = pplot_df.melt(ignore_index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pplot_df['Clone'] = pd.Categorical(pplot_df.index, categories = clone_order, ordered = True)\n",
    "pplot_df = pplot_df[pplot_df.Clone.isin(pg_clones)]\n",
    "pplot_df['CloneNo'] = pplot_df['Clone'].apply(lambda cl : clone_order.index(cl))\n",
    "pplot_df.head()\n",
    "pplot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pplot_df['R'] = pplot_df['PeakID'].map(PCR_ans['GNLY'].loc['GNLY_R'])\n",
    "pplot_df['R_sq'] = pplot_df['R']**2\n",
    "pplot_df['Peak location'] = pplot_df['PeakID'].map(ATAC_data.var.loc[peaklist,'mid'])\n",
    "#pplot_df['end'] = pplot_df['PeakID'].map(ATAC_data.var.loc[peaklist,'end'])\n",
    "pplot_df = pplot_df.sort_values(by=['CloneNo','R_sq'])\n",
    "pplot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "gene_chart = alt.Chart(GNLY_df).mark_circle(size=100).encode(\n",
    "    x={\"field\": \"Clone\",\n",
    "      \"type\": \"ordinal\",\n",
    "       \"sort\" : clone_order,\n",
    "       \"axis\" : alt.Axis(labels=False, title=\" \")},\n",
    "    y='GNLY',\n",
    "    color='Clone',\n",
    "    tooltip=['Clone', 'Obs', 'GNLY']\n",
    ").interactive().properties(\n",
    "    width=400,\n",
    "    height=200\n",
    ")\n",
    "\n",
    "highlight = alt.selection(type='single', on='mouseover',\n",
    "                          fields=['PeakID'], nearest=True)\n",
    "\n",
    "TSS = adata.var.loc['GNLY','TSS']\n",
    "\n",
    "peak_base = alt.Chart(pplot_df).encode(\n",
    "    x={\"field\": \"Clone\",\n",
    "               \"type\": \"ordinal\",\n",
    "               \"sort\" : clone_order},\n",
    "    y={\"field\" : \"value\",\n",
    "       \"type\": \"quantitative\",\n",
    "       \"axis\" : alt.Axis(labels=False, title=\"Peak height\")},\n",
    "    color=alt.Color(\"PeakID\")\n",
    ")\n",
    "\n",
    "# interval selection in the bar chart\n",
    "pts = alt.selection_multi(encodings=[\"x\"])\n",
    "\n",
    "points = peak_base.mark_circle().encode(\n",
    "    opacity=alt.value(0)\n",
    ").add_selection(\n",
    "    highlight\n",
    ").properties(\n",
    "    width=400\n",
    ")\n",
    "\n",
    "peak_chart = peak_base.mark_line().encode(\n",
    "    size=alt.condition(~highlight, alt.value(1), alt.value(3)), #,\n",
    "    opacity=alt.Opacity(\"R_sq\", legend=None),\n",
    "    tooltip=['Clone', 'PeakID', 'value']#,\n",
    "    ).transform_filter(pts).interactive().properties(\n",
    "    width=400,\n",
    "    height=100\n",
    ")\n",
    "\n",
    "peak_chart = points + peak_chart\n",
    "\n",
    "\n",
    "peakR_chart = alt.Chart(pplot_df).mark_bar().encode(\n",
    "    x={\"field\":\"Peak location\", \n",
    "       \"type\":\"quantitative\",\n",
    "       \"axis\":alt.Axis(values=[TSS-50000, TSS, TSS+50000])},\n",
    "    y=\"R\",\n",
    "    text=\"PeakID\",\n",
    "    color=\"PeakID\",\n",
    "    opacity=alt.condition(~highlight, alt.value(0.1), alt.value(1.0)),\n",
    "    tooltip=['PeakID','R','R_sq']\n",
    "    #color=alt.condition(\n",
    "    #    alt.datum.R > 0,\n",
    "    #    alt.value(\"steelblue\"),  # The positive color\n",
    "    #    alt.value(\"orange\")  # The negative color\n",
    "    #)\n",
    ").interactive().properties(\n",
    "    width=300,\n",
    "    height=50).add_selection(pts)\n",
    "\n",
    "\n",
    "\n",
    "TSS_line = alt.Chart(pd.DataFrame({'TSS': [TSS]})).mark_rule().encode(\n",
    "    x=alt.X('TSS'))\n",
    "\n",
    "peakR_chart = peakR_chart + TSS_line\n",
    "\n",
    "#peakR_chart = peakR_chart + peakR_text\n",
    "\n",
    "#peaks_chart = alt.hconcat(peak_chart, peakR_chart)\n",
    "\n",
    "chart = alt.vconcat(gene_chart, peak_chart, peakR_chart).resolve_scale(\n",
    "    color='independent')\n",
    "\n",
    "chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chart.save('GNLY.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pgplotly(gene, protein = None, cutoff=0.2, window=50000):\n",
    "    if protein:\n",
    "        num_rows = 3\n",
    "        specs=[[{\"type\": \"scatter\"}],\n",
    "               [{\"type\": \"line\"}],\n",
    "               [{\"type\": \"line\"}]]\n",
    "    else:\n",
    "        num_rows = 2\n",
    "        specs=[[{\"type\": \"scatter\"}],\n",
    "               [{\"type\": \"scatter\"}]]\n",
    "        \n",
    "    #fig, ax = plt.subplots(num_rows,1,figsize=(2.5,0.75*num_rows), dpi=300, sharex=True,\n",
    "    #                       gridspec_kw = {'hspace':0.05, 'height_ratios':hr})            \n",
    "    #\n",
    "    #plt.rcParams['text.usetex'] = False\n",
    "    #setfontsize()\n",
    "\n",
    "    fig = make_subplots(\n",
    "        rows=num_rows, cols=1,\n",
    "        shared_xaxes=True,\n",
    "        vertical_spacing=0.03,\n",
    "        specs=specs\n",
    "    )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    PCR_df = PCR_analysis(gene, window=window)\n",
    "    clone_order = list(gene_clonal[gene].sort_values().index)\n",
    "    \n",
    "    gplot_df = pd.DataFrame(columns=[gene,'Clone'])\n",
    "    gplot_df[gene] = adata.to_df()[gene]\n",
    "    \n",
    "    \n",
    "    gplot_df['Clone'] = pd.Categorical(adata.obs.Clone_ID, categories = clone_order, ordered = True)\n",
    "    gplot_df = gplot_df[gplot_df.Clone.isin(pg_clones)]\n",
    "    clone_order = [clone for clone in clone_order if clone in pg_clones]\n",
    "    gplot_df['CloneNo'] = gplot_df['Clone'].apply(lambda cl : clone_order.index(cl))\n",
    "    gplot_df = gplot_df.sort_values(by='CloneNo')\n",
    "    gplot_df['Text'] = ''\n",
    "    for row in gplot_df.index:\n",
    "        expr = gplot_df.loc[row,gene]\n",
    "        cl = gplot_df.loc[row,'Clone']\n",
    "        gplot_df.loc[row,'Text'] = '{} expression {:1.3f} in Clone {}.'.format(gene,expr,cl)\n",
    "    col_range = list(range(len(gplot_df)))\n",
    "    #for clone in clone_order:\n",
    "    #    for row in range(num_rows):\n",
    "    #        ax[row].axvline(clone,linestyle=':',color='gray',linewidth=0.5, alpha=0.5, zorder=-5)\n",
    "    \n",
    "    # GENE EXPRESSION PLOT\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=gplot_df['Clone'],\n",
    "            y=gplot_df[gene],\n",
    "            mode=\"markers\",\n",
    "            marker=dict(size=16, \n",
    "                        color=gplot_df['CloneNo'], \n",
    "                        colorscale=px.colors.qualitative.Dark24),\n",
    "            text=gplot_df['Text'],\n",
    "            name=\"{} expression\".format(gene)\n",
    "        ),\n",
    "        row=1, col=1)\n",
    "    \n",
    "    #sns.scatterplot(x='Clone', y=gene, hue='Clone', data=gplot_df, ax=ax[0], legend=False)\n",
    "    #gene_ticks = [gplot_df[gene].min(), gplot_df[gene].mean(), gplot_df[gene].max()]\n",
    "    #ax[0].set_yticks(gene_ticks)\n",
    "    #ax[0].set_yticklabels(gene_ticks, x=0.03)\n",
    "    #ax[0].yaxis.set_major_formatter(FormatStrFormatter('%.1f'))\n",
    "    #ax[0].set_ylabel(gene)\n",
    "    #ax[0].yaxis.set_label_position(\"right\")\n",
    "    \n",
    "    # ATAC PLOT\n",
    "    peaklist = [peak for peak in PCR_df.index if peak[:8] == 'Interval']\n",
    "    peaklist = [peak for peak in peaklist if PCR_df.loc[gene+'_Rsq',peak] > cutoff]\n",
    "    pplot_df = ATAC_data.to_df()[peaklist].groupby(ATAC_data.obs.Clone_ID).mean()\n",
    "    pplot_df['Clone'] = pd.Categorical(pplot_df.index, categories = clone_order, ordered = True)\n",
    "    pplot_df = pplot_df[pplot_df.Clone.isin(pg_clones)]\n",
    "    \n",
    "    pplot_df[peaklist] = (pplot_df[peaklist] - pplot_df[peaklist].min()) #/ pplot_df[peaklist].max()\n",
    "    pplot_df[peaklist] = pplot_df[peaklist] / pplot_df[peaklist].max()\n",
    "    pplot_df['CloneNo'] = pplot_df['Clone'].apply(lambda cl : clone_order.index(cl))\n",
    "    pplot_df = pplot_df.sort_values(by='CloneNo')\n",
    "    #\n",
    "    no_peaks = 0\n",
    "    for peak in peaklist:\n",
    "        peak_Rsq = PCR_df.loc[gene+'_Rsq',peak]\n",
    "        if peak_Rsq > 0.05:\n",
    "            peak_label = '${:1.2f}$'.format(peak_Rsq)\n",
    "            no_peaks += 1\n",
    "        else:\n",
    "            peak_label = None\n",
    "        fig.add_trace(\n",
    "            go.Scatter(\n",
    "                x=pplot_df['Clone'],\n",
    "                y=pplot_df[peak],\n",
    "                mode=\"lines\",\n",
    "                #text=gplot_df['Text'],\n",
    "                name=peak\n",
    "            ),\n",
    "            row=2, col=1)\n",
    "        \n",
    "    \n",
    "    \n",
    "    # PROTEIN PLOT\n",
    "    #if protein:\n",
    "    #    proplot_df = pd.DataFrame(prot_clonal[protein], columns=[protein])\n",
    "    #    proplot_df[protein] = proplot_df[protein] - proplot_df[protein].min()\n",
    "    #    proplot_df[protein] = proplot_df[protein] / proplot_df[protein].max()\n",
    "    #    proplot_df['Clone'] = proplot_df.index\n",
    "    #    sns.lineplot(x='Clone', y=protein, data=proplot_df, ax=ax[1], linewidth=0.7)\n",
    "    #    ax[1].text(1,0.5,protein, ha='left')\n",
    "    #    ax[1].set_ylabel('Prot')\n",
    "    #    ax[1].yaxis.set_label_position(\"right\")\n",
    "    #    ax[1].set_yticks([])\n",
    "    #    ax[1].set_xlabel('')\n",
    "    #    ax[1].spines['top'].set_color('lightgray')\n",
    "    #    ax[1].spines['bottom'].set_color('lightgray')\n",
    "    \n",
    "\n",
    "    if protein:\n",
    "        title = '{}, {}, and {} peaks.'.format(gene, protein, len(peaklist))\n",
    "    else:\n",
    "        title = '{} and {} peaks.'.format(gene, len(peaklist))\n",
    "    #plt.suptitle(title, y=1.02)\n",
    "    \n",
    "    # Plot legend\n",
    "    #if no_peaks <= 6:\n",
    "    #    ncol = no_peaks\n",
    "    #else:\n",
    "    #    ncol = (no_peaks + 1)//2\n",
    "    #lgnd = plt.legend(bbox_to_anchor=(0.5, 0.05), title = 'Peak/Gene $R^2$',  \n",
    "    #                  loc=9, ncol=ncol, framealpha=0.5)\n",
    "    #lgnd.get_frame().set_linewidth(0.5)\n",
    "    \n",
    "    # Save and show figure.\n",
    "    #set_size(fig, (2.5, 1.5))\n",
    "    #plt.savefig(\"Panels/MultiPlots/{}_multi.pdf\".format(gene), bbox_inches='tight')\n",
    "    #plt.show()\n",
    "    fig.show()\n",
    "    fig.write_html(\"Plotly/{}_plotly.html\".format(gene), include_plotlyjs=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pgplotly('GNLY')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "setfontsize()\n",
    "fig,ax = plt.subplots(1,1,figsize=(3,3), dpi=300)\n",
    "\n",
    "n_max = PCR_summary.num_eig.max()\n",
    "plot_df = pd.DataFrame(index=PCR_summary.index)\n",
    "plot_df['Top eigenpeak R^2'] = PCR_summary['eigen0_R_sq'].fillna(0)\n",
    "plot_df['Number of peaks within 50kb'] = PCR_summary['num_peaks']\n",
    "plot_df['Number of eigenpeaks \\n with R^2 > 0.2'] = PCR_summary['num_eig'].astype('category')\n",
    "\n",
    "complex_genes = PCR_summary[PCR_summary.num_eig > 4].index\n",
    "peaky_genes = PCR_summary[PCR_summary.num_peaks > 50].index\n",
    "eigen_genes = PCR_summary[PCR_summary.eigen0_R_sq > 0.75].index\n",
    "ann_genes = complex_genes.union(peaky_genes).union(eigen_genes)\n",
    "\n",
    "ann_genes = ['KLRD1', 'GNLY', 'PLS3-1', 'B2M', 'CD28', 'CTLA4', 'ATXN1']\n",
    "for n in range(n_max+1):\n",
    "    p_df = plot_df[plot_df['Number of eigenpeaks \\n with R^2 > 0.2'] == n]\n",
    "    if n > 1:\n",
    "        ec = 'lightgreen'\n",
    "        lw = 0.4\n",
    "        s=20\n",
    "        if n > 2:\n",
    "            ec = 'red'\n",
    "    else:\n",
    "        ec = None\n",
    "        lw = 0\n",
    "        s=10\n",
    "    if n == 0:\n",
    "        leg = True\n",
    "    else:\n",
    "        leg = False\n",
    "    \n",
    "    if n == 2:\n",
    "        for gene in ann_genes:\n",
    "            x = plot_df.loc[gene, 'Top eigenpeak R^2']\n",
    "            y = plot_df.loc[gene, 'Number of peaks within 50kb']\n",
    "            if y < 20:\n",
    "                y_loc = y + 0.5\n",
    "            else:\n",
    "                y_loc = y*1.1\n",
    "            ax.text( x,y_loc, gene, color='black', ha='center',\n",
    "                    bbox = dict(facecolor='white', alpha=0.5, edgecolor='white', pad=0))\n",
    "    \n",
    "    sns.scatterplot(x='Top eigenpeak R^2', y='Number of peaks within 50kb', \n",
    "                    hue = 'Number of eigenpeaks \\n with R^2 > 0.2', \n",
    "                    edgecolor=ec, linewidth = lw, s=s, data=p_df, ax=ax, legend=leg)\n",
    "    \n",
    "    ag = [gene for gene in ann_genes if gene in p_df.index]\n",
    "    sns.scatterplot(x='Top eigenpeak R^2', y='Number of peaks within 50kb', \n",
    "                    hue = 'Number of eigenpeaks \\n with R^2 > 0.2', \n",
    "                    edgecolor='k', linewidth = 1.0, s=s, data=p_df.loc[ag], ax=ax, legend=False)\n",
    "\n",
    "\n",
    "    \n",
    "ax.set_yscale('symlog',linthreshy=20.0)\n",
    "ax.set_yticks([1,2,5,10,20,50,100])\n",
    "ax.set_yticklabels([1,2,5,10,20,50,100], x=0.02)\n",
    "ax.set_xticks([0,0.2,0.4,0.6,0.8,1.0])\n",
    "ax.set_xticklabels([0,0.2,0.4,0.6,0.8,1.0], y=0.02)\n",
    "lgnd = plt.legend(title='# Eigenpeaks \\n with $R^2 > 0.2$', bbox_to_anchor=(1.0, 1), loc='upper right')\n",
    "for j in range(4):\n",
    "    lgnd.legendHandles[j]._sizes = [25]\n",
    "set_size(fig, (3, 3))\n",
    "plt.savefig(\"Panels/PCR_allgenes.pdf\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of genes near variable peaks\n",
    "\n",
    "This part is a bit exploratory..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We might want to find peaks within a window around a gene too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_data.var.loc['Interval_1111']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genes_near(peak, window=50000):\n",
    "    if peak[:8] == 'Interval':\n",
    "        pk = peak\n",
    "    else:\n",
    "        pk = 'Interval_'+peak # No need to enter Interval_ in input.\n",
    "    pk_info = ATAC_data.var.loc[pk]\n",
    "    g_df = adata.var[adata.var['chr'] == pk_info['chr']]\n",
    "    g_df = g_df[np.abs(g_df.start + g_df.end - 2*pk_info['mid']) < window]\n",
    "    return g_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_near('11500', window=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks_clonal = ATAC_data.var[ATAC_data.var.quality_p > 0.1]\n",
    "peaks_clonal = peaks_clonal.clonal_p.sort_values()[:500]\n",
    "peaks_clonal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_peakclonal = [list(genes_near(peak)) for peak in peaks_clonal.index]\n",
    "genes_peakclonal = [item for itemlist in genes_peakclonal for item in itemlist]\n",
    "genes_peakclonal = set(genes_peakclonal)\n",
    "print('{} genes found within 50kb of a clonal peak.'.format(len(genes_peakclonal)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "desetfontsize()\n",
    "sns.histplot(adata.var.loc[genes_peakclonal].F_pval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(adata.var.F_pval)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
